---
Title: "Seasonal dynamics of microbial degradation"
Author: "Joeselle Serrana"
Date: "11/08/2023"
output: html_document
editor_options: 
  chunk_output_type: console
---

## Load required library and set working directory 

```{r}
library("phyloseq"); library("microeco") ## For data manipulation, analysis and visualization. Visit https://github.com/ChiLiubio/microeco for more details.
library("ggplot2"); library("ggradar"); library("ggnested"); library("ggcor"); library("ComplexHeatmap") ## For visualizations and plot.
library("ggh4x"); library("dplyr"); library("mice"); library("tibble"); library("magrittr"); library("ggpubr"); library("RColorBrewer"); library("aplot") ## Other packages needed.

#setwd("~/Library/Mobile Documents/com~apple~CloudDocs/Documents/(Draft) Spatiotemporal Biodegradation/MS-Spatiotemporal-BIE_Analysis")
setwd("C:/Users/JoeselleSerrana/iCloudDrive/Documents/(Draft) Spatiotemporal Biodegradation/MS-Spatiotemporal-BIE_Analysis")

## [START FROM STEP 4...] Read microtable object.
bac.mco <- readRDS("src/bac.mco.RDS"); euk.mco <- readRDS("src/euk.mco.RDS")

bac.mco.fld <- readRDS("src/bac.mco.fld.RDS"); bac.mco.btc <- readRDS("src/bac.mco.btc.RDS")
euk.mco.fld <- readRDS("src/euk.mco.fld.RDS"); euk.mco.btc <- readRDS("src/euk.mco.btc.RDS")

alg.mco.fld <- readRDS("src/alg.mco.fld.RDS")
fun.mco.fld <- readRDS("src/fun.mco.fld.RDS")
met.mco.fld <- readRDS("src/met.mco.fld.RDS")
pro.mco.fld <- readRDS("src/pro.mco.fld.RDS")

bdg.mco.fld <- readRDS("src/bdg.mco.fld.RDS")
keg.mco.fld <- readRDS("src/keg.mco.fld.RDS")
keg.mco.met <- readRDS("src/keg.mco.met.RDS")
keg.mco.xen <- readRDS("src/keg.mco.xen.RDS")

bac.mco.ssn <- readRDS("src/bac.mco.ssn.RDS")
alg.mco.ssn <- readRDS("src/alg.mco.ssn.RDS")
fun.mco.ssn <- readRDS("src/fun.mco.ssn.RDS")
pro.mco.ssn <- readRDS("src/pro.mco.ssn.RDS")
met.mco.ssn <- readRDS("src/met.mco.ssn.RDS")
```

## ########################################################## ##
##                                                            ##
##                   AMPLICON DATA ANALYSIS                   ##
##                                                            ##
## ########################################################## ##
### Amplicon sequence processing w/ DADA2 

```{r 16S, eval = FALSE}
library("dada2"); library("phangorn"); library("DECIPHER") ## For amplicon processing and phylogenetic analysis.

## For 16S rRNA Data

## STEP 0: Copy raw reads to ./data directory.

## Read processing for 16S amplicon sequence data.
#setwd("P:/2023_Seasonal-Biodegradation/18S-V4")
#flist <- list.files("P:/2023_Seasonal-Biodegradation/18S-V4/Result_X204SC23100432-Z01-F003_18SV4/result_X204SC23100432-Z01-F003/01.RawData/", "^.+[.]E*fastq.gz", full.names = TRUE, recursive = TRUE)
#file.copy(flist, "P:/2023_Seasonal-Biodegradation/18S-V4/data/raw-reads/"); setwd("P:/2023_Seasonal-Biodegradation/18S-V4/data/raw-reads/")
#remove.extendedFrags <- list.files("P:/2023_Seasonal-Biodegradation/18S-V4/data/raw-reads/", pattern = "extendedFrags"); file.remove(remove.extendedFrags)
#remove.raw <- list.files("P:/2023_Seasonal-Biodegradation/18S-V4/data/raw-reads/", pattern = "raw"); file.remove(remove.raw)

## STEP 1a: Read quality assessment: Plot read quality

## Sort the file names to ensure that the names for the corresponding forward/reverse reads are in same order and store them in two variables.
fnFs <- sort(list.files("data/16S-raw-reads", pattern = "_1.fastq.gz", full.names = TRUE)); fnRs <- sort(list.files("data/16S-raw-reads", pattern = "_2.fastq.gz", full.names = TRUE))

## Extract sample names from the filenames, assuming filenames have the format: SAMPLENAME_XXX.fastq.qz
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)

## Plot average quality scores for forward/reverse reads. The script below will plot samples 1 to 8 only. Change the number to include more files.
plotQualityProfile(fnFs[1:3]); plotQualityProfile(fnRs[1:3]) 

## STEP 1b: Quality filtering: Filter reads based on length and maxee

## Place filtered files in the "filtered-reads/" subdirectory.
filtFs <- file.path("data/16S-filtered-reads", paste0(sample.names, "_F_filt.fastq.gz")); filtRs <- file.path("data/16S-filtered-reads", paste0(sample.names, "_R_filt.fastq.gz"))

## No need for "trimLeft = c(19, 20)" since adapter and primer barcodes have been truncated from the raw reads.
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, 
                     maxN = 0,
                     maxEE = c(2,2), 
                     truncQ = 2,
                     #truncLen = c(220, 200),
                     minLen = 150,
                     rm.phix = TRUE,
                     compress = TRUE, 
                     multithread = TRUE)  ## On Windows set multithread = FALSE

head(out)

## STEP 1c: Quality filtering: Learn error rates

## Estimate the error model for DADA2 algorithm using forward/reverse reads.
errF <- learnErrors(filtFs, multithread = TRUE); errR <- learnErrors(filtRs, multithread = TRUE)

## Plot the error rates for all possible base transitions.
plotErrors(errF, nominalQ = TRUE)

## STEP 1d: Dereplicate and denoise reads

## Dereplicate FASTQ files for both forward and reverse reads to speed up computation.
exists <- file.exists(filtFs); derepFs <- derepFastq(filtFs[exists], verbose = TRUE); derepRs <- derepFastq(filtRs[exists], verbose = TRUE)

## Name the derep-class objects by the sample names.
names(derepFs) <- sample.names[exists]; names(derepRs) <- sample.names[exists]

## At this step, the core sample inference algorithm is applied to the dereplicated data.
dadaFs <- dada(derepFs, err = errF, multithread = TRUE); dadaRs <- dada(derepRs, err = errR, multithread = TRUE); dadaFs[[1]] # Inspect the returned dada-class object.

# STEP 1e: Merge paired-end reads and construct sequence table

## Merge denoised reads.
mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, minOverlap = 10, maxMismatch = 5, verbose = TRUE)
head(mergers[[1]]) # Inspect the merger data.frame from the first sample.

## Organize sequence variants into a sequence table.
seqtab <- makeSequenceTable(mergers); dim(seqtab) # The sequences being tabled vary in length.
table(nchar(getSequences(seqtab))) # Inspect distribution of sequence lengths.

# STEP 1f: Chimera filtering

## Perform de novo chimera sequence detection and removal.
seqtab.nochim <- removeBimeraDenovo(seqtab, method = "consensus", multithread = TRUE, verbose = TRUE) # Remove chimera.
dim(seqtab.nochim); sum(seqtab.nochim)/sum(seqtab)

## STEP 1g: Singleton filtering
## Remove all ASVs present only once (or less) across all samples:

is1 <- colSums(seqtab.nochim) <= 10
seqtab.nosing <- seqtab.nochim[,!is1]

## Extract read processing information (counts)

getN <- function(x) sum(getUniques(x)) # Track reads through the pipeline.
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim), rowSums(seqtab.nosing))

colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim", "nosing")
rownames(track) <- sample.names; head(track)

## Save non-chimeric ASV table.
saveRDS(seqtab.nochim, "src/dada2/16S_seqtab.nochim.RDS")
saveRDS(seqtab.sing, "src/dada2/16S_seqtab.nosing.RDS")

## Extract ASV-Table with sequence and abundance/sample information
write.table(track, "src/dada2/16S_read_counts.tsv") ## Read counts per processing step.
write.table(t(seqtab.nosing), "src/dada2/16S_asv_table.tsv") ## Transposed seqtab.nochim file.
#```

#```{r 18S, eval = FALSE}
## For 18S rRNA Data

## STEP 1a: Read quality assessmnet: Plot read quality

## Sort the file names to ensure that thenames for the corresponding forward/reverse reads are in same order and store them in two variables.
fnFs <- sort(list.files("data/18S-raw-reads", pattern = "_1.fastq.gz", full.names = TRUE))
fnRs <- sort(list.files("data/18S-raw-reads", pattern = "_2.fastq.gz", full.names = TRUE))

## Extract sample names from the filenames, assuming filenames have the format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)

## Plot average quality scores for forward/reverse reads. The script below will plot samples 1 to 8 only. Change the number to include more files.
plotQualityProfile(fnFs[1:4]); plotQualityProfile(fnRs[1:4]) 

## STEP 1b: Quality filtering: Filter reads based on length and maxee

## Place filtered files in the "filtered-reads/" subdirectory.
filtFs <- file.path("data/18S-filtered-reads", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path("data/18S-filtered-reads", paste0(sample.names, "_R_filt.fastq.gz"))

## No need for "trimLeft = c(19, 20)" since adapter and primer barcodes have been truncated from the raw reads.
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, 
                     maxN = 0,
                     maxEE = c(2,2), 
                     truncQ = 2,
                     #truncLen = c(220, 200), 
                     minLen = 150,
                     rm.phix = TRUE,
                     compress = TRUE, 
                     multithread = TRUE)  ## On Windows set multithread = FALSE

head(out)

## STEP 1c: Quality filtering: Learn error rates

## Estimate the error model for DADA2 algorithm using forward/reverse reads.
errF <- learnErrors(filtFs, multithread = TRUE); errR <- learnErrors(filtRs, multithread = TRUE)

## Plot the error rates for all possible base transitions.
plotErrors(errF, nominalQ = TRUE)

## STEP 1d: Dereplicate and denoise reads

## Dereplicate FASTQ files for both forward and reverse reads to speed up computation.
exists <- file.exists(filtFs)
derepFs <- derepFastq(filtFs[exists], verbose = TRUE); derepRs <- derepFastq(filtRs[exists], verbose = TRUE)

## Name the derep-class objects by the sample names.
names(derepFs) <- sample.names[exists]; names(derepRs) <- sample.names[exists]

## At this step, the core sample inference algorithm is applied to the dereplicated data.
dadaFs <- dada(derepFs, err = errF, multithread = TRUE); dadaRs <- dada(derepRs, err = errR, multithread = TRUE)
dadaFs[[1]] # Inspect the returned dada-class object.

# STEP 1e: Merge paired-end reads and construct sequence table

## Merge denoised reads.
mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, minOverlap = 10, maxMismatch = 5, verbose = TRUE)
head(mergers[[1]]) # Inspect the merger data.frame from the first sample.

## Organize sequence variants into a sequence table.
seqtab <- makeSequenceTable(mergers); dim(seqtab) # The sequences being tabled vary in length.
table(nchar(getSequences(seqtab))) # Inspect distribution of sequence lengths.

# STEP 1f: Chimera filtering

## Perform de novo chimera sequence detection and removal.
seqtab.nochim <- removeBimeraDenovo(seqtab, method = "consensus", multithread = TRUE, verbose = TRUE) # Remove chimera.
dim(seqtab.nochim); sum(seqtab.nochim)/sum(seqtab)

## STEP 1g: Singleton filtering
## Remove all ASVs present only once across all samples:

is1 <- colSums(seqtab.nochim) <= 1
seqtab.nosing <- seqtab.nochim[,!is1]

## Extract read processing information (counts)

getN <- function(x) sum(getUniques(x)) # Track reads through the pipeline.
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim), rowSums(seqtab.nosing))

colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim", "nosing")
rownames(track) <- sample.names; head(track)

## Save non-chimeric ASV table.
saveRDS(seqtab.nochim, "src/dada2/18S_seqtab.nochim.RDS")
saveRDS(seqtab.sing, "src/dada2/18S_seqtab.nosing.RDS")

## Extract ASV-Table with sequence and abundance/sample information
write.table(track, "src/dada2/18S_read_counts.tsv") ## Read counts per processing step.
write.table(t(seqtab.nosing), "src/dada2/18S_asv_table.tsv") ## Transposed seqtab.nochim file.
```

### Taxonomic assignment, and phylognetic analysis
 
```{r 16S, eval = FALSE}
## For 16S rRNA Data

## STEP 2a: Assign taxonomic with the GTDB database

seqtab.nosing <- readRDS("src/dada2/16S_seqtab.nosing.RDS") # Read saved RDS file if run not saved.

## Assign taxonomy using Silva (DADA2 formatted 16S rRNA gene sequences for both bacteria & archaea) from https://zenodo.org/records/14169026 (previously downloaded in the data/database/ directory).
taxa <- assignTaxonomy(seqtab.nosing, "data/database/silva_nr99_v138.2_toSpecies_trainset.fa.gz", multithread = TRUE)
#colnames(taxa) <- c("Kingdom", "Phylum","Class", "Order", "Family", "Genus", "Species"); unname(taxa)

## Save taxonomic assignment.
saveRDS(taxa, "src/dada2/16S-taxa.RDS")

## STEP 2c-1: Phylogenetic analysis

## Run sequence alignment (MSA) using DECIPHER.
sequences <- getSequences(seqtab.nosing); names(sequences) <- sequences ## Extract sequences from DADA2 output
alignment <- AlignSeqs(DNAStringSet(sequences), anchor = NA)

saveRDS(sequences, "src/dada2/16S-sequences.RDS") ## Save alignment.
saveRDS(alignment, "src/dada2/16S-alignment.RDS") ## Save alignment.

writeXStringSet(DNAStringSet(sequences), file = "src/dada2/16S-sequences.fasta") ## Save the alignment as FASTA file
writeXStringSet(alignment, file = "src/dada2/16S-alignment.fasta") ## Save the alignment as FASTA file

## STEP 2c: FastTree v2.1.11 (https://microbesonline.org/fasttree/)

#$ FastTree -gtr -nt src/dada2/16S-alignment.fasta > src/fasttree/16S.tree
## Save tree file at the "src/fasttree/" subdirectory.
```

```{r 18S, eval = FALSE}
## For 18S rRNA Data

## STEP 2a: Assign taxonomy with the PR2 database

seqtab.nosing <- readRDS("src/dada2/18S_seqtab.nosing.RDS") # Read saved RDS file if run not saved.
 
## Assign taxonomy using PR2 database version 4.41.0 (previously downloaded in the data/database/ directory).
taxa <- assignTaxonomy(seqtab.nosing, "data/database/pr2_version_5.0.0_SSU_dada2.fasta.gz", multithread = TRUE)
colnames(taxa) <- c("Domain", "Supergroup", "Division", "Subdivision", "Class", "Order", "Family", "Genus", "Species"); unname(taxa)

## Remove taxonomy suffix ":nucl" and ":plas" from PR2 database annotation because we only need the taxa and  not the organelle information.
taxa <- gsub(":nucl", "", taxa); taxa <- gsub(":plas", "", taxa)

## Save taxonomic assignment.
saveRDS(taxa, "src/dada2/18S-taxa.RDS")

## STEP 2b: Add ecological function and trophic group based Vaulot et al. (2022: https://doi.org/10.1111/1755-0998.13674).

taxa <- readRDS("src/dada2/18S-taxa.RDS") # Read saved RDS file if run not saved.

## Extract the taxonomic, ecological function and trophic group information from the asv.xlsx file downloaded from the metaPR2 dataset.
eco.function <- readxl::read_excel("data/database/metapr2_datasets_Eukaryota_2023-03-21/asv.xlsx")
eco.function <- unique(eco.function[,c(9,12:13)]); eco.function <- na.omit(eco.function)
colnames(eco.function) <- c("Genus","Function","Trophic") ## Extract unique species names and its corresponding ecological function and trophic group.

## Match ecological function data to taxa table, i.e., at the genus-level.
taxa.fun <- as.data.frame(taxa) %>% left_join(eco.function, by = "Genus"); rownames(taxa.fun) <- rownames(taxa)

## NOTE: Not all genera and species are represented in the metaPR2 ASV file, hence, taxa that have similar annotations at the division-level, i.e., (1) Chlorophyta, (2) Choanoflagellida, (3) Chrompodellids, (4) Ciliophora, (5) Opalozoa/Bigyra, (6) Metazoa, (7) Perkinsea, and (8) Streptophyta will be annotated with ecological function and trophic group information based on Table S2 of Vaulot et al. (2022) accordingly, other levels also assigned annotations based on subdivisions and class-level:

taxa.fun$Function <- ifelse(taxa.fun$Subdivision == "Dinoflagellata" & is.na(taxa.fun$Function), "dinoflagellates", taxa.fun$Function) 

taxa.fun$Function <- ifelse(taxa.fun$Class == "Eustigmatophyceae" | taxa.fun$Class == "Chrysophyceae" | taxa.fun$Class == "Bacillariophyceae" | taxa.fun$Division == "Chlorophyta" | taxa.fun$Division == "Streptophyta" & is.na(taxa.fun$Function), "phototrophs", taxa.fun$Function)   
taxa.fun$Trophic <- ifelse(taxa.fun$Class == "Eustigmatophyceae" | taxa.fun$Class == "Chrysophyceae" | taxa.fun$Class == "Bacillariophyceae" | taxa.fun$Subdivision == "Dinoflagellata" | taxa.fun$Division == "Chlorophyta" | taxa.fun$Division == "Streptophyta" & is.na(taxa.fun$Trophic), "phototrophic", taxa.fun$Trophic)   

taxa.fun$Function <- ifelse(taxa.fun$Class == "Ascomycota" | taxa.fun$Subdivision == "Choanoflagellata" | taxa.fun$Subdivision == "Chrompodellids" | taxa.fun$Subdivision == "Ciliophora" | taxa.fun$Subdivision == "Bigyra" & is.na(taxa.fun$Function), "phagotrophs", taxa.fun$Function)   
taxa.fun$Trophic <- ifelse(taxa.fun$Class == "Ascomycota" | taxa.fun$Subdivision == "Choanoflagellata" | taxa.fun$Subdivision == "Chrompodellids" | taxa.fun$Subdivision == "Ciliophora" | taxa.fun$Subdivision == "Bigyra" | taxa.fun$Division == "Tubulinea" & is.na(taxa.fun$Trophic), "consumer", taxa.fun$Trophic)

taxa.fun$Function <- ifelse(taxa.fun$Class == "Peronosporomycetes" | taxa.fun$Division == "Tubulinea" | taxa.fun$Division == "Evosea" | taxa.fun$Division == "Discosea" | taxa.fun$Subdivision == "Apicomplexa" | taxa.fun$Subdivision == "Perkinsea" & is.na(taxa.fun$Function), "parasites", taxa.fun$Function)   
taxa.fun$Trophic <- ifelse(taxa.fun$Class == "Peronosporomycetes" |taxa.fun$Division == "Evosea" | taxa.fun$Division == "Discosea" | taxa.fun$Subdivision == "Apicomplexa" | taxa.fun$Subdivision == "Perkinsea" & is.na(taxa.fun$Trophic), "parasitic", taxa.fun$Trophic)   
taxa.fun$Function <- ifelse(taxa.fun$Subdivision == "Metazoa", "metazoans", taxa.fun$Function)   
taxa.fun$Trophic <- ifelse(taxa.fun$Subdivision == "Metazoa", "metazoan", taxa.fun$Trophic)

taxa.fun$Trophic <- ifelse(taxa.fun$Subdivision == "Fungi" & is.na(taxa.fun$Trophic), "fungi", taxa.fun$Trophic)

## STEP 2b: Add grouping based on subdivision.

taxa.fun$Category <- NA

taxa.fun$Category <- ifelse(taxa.fun$Subdivision == "Ancyromonadida_X" | taxa.fun$Subdivision == "Apicomplexa" |taxa.fun$Subdivision == "Apusomonada_X" | taxa.fun$Subdivision == "Bigyra" |
                            taxa.fun$Subdivision == "Breviatea_X" | taxa.fun$Subdivision == "Centroplasthelida_X" | taxa.fun$Subdivision == "Cercozoa" | taxa.fun$Subdivision == "Choanoflagellata" |
                            taxa.fun$Subdivision == "Chrompodellids" | taxa.fun$Subdivision == "Ciliophora" | taxa.fun$Subdivision == "Dinoflagellata" | taxa.fun$Subdivision == "Discoba_X" |
                            taxa.fun$Subdivision == "Discosea_X" | taxa.fun$Subdivision == "Euglenozoa" | taxa.fun$Subdivision == "Eumycetozoa" | taxa.fun$Subdivision == "Evosea_X" |
                            taxa.fun$Subdivision == "Filasterea" | taxa.fun$Subdivision == "Nebulidia_X" | taxa.fun$Subdivision == "Nibbleridia_X" | taxa.fun$Subdivision == "Perkinsea" |
                            taxa.fun$Subdivision == "Pluriformea" | taxa.fun$Subdivision == "Preaxostyla" | taxa.fun$Subdivision == "Radiolaria" | taxa.fun$Subdivision == "Rigifilida_X" |
                            taxa.fun$Subdivision == "Rotosphaerida" | taxa.fun$Subdivision == "Telonemia_X" | taxa.fun$Subdivision == "Tubulinea_X" & is.na(taxa.fun$Category), "Protist", taxa.fun$Category)

taxa.fun$Category <- ifelse(taxa.fun$Subdivision == "Chlorophyta_X" | taxa.fun$Subdivision == "Cryptophyta_X" | taxa.fun$Subdivision == "Gyrista" | taxa.fun$Subdivision == "Haptophyta_X" |
                            taxa.fun$Subdivision == "Hemimastigophora_X" | taxa.fun$Subdivision == "Ichthyosporea" | taxa.fun$Subdivision == "Kathablepharida" | taxa.fun$Subdivision == "Picozoa_X" |
                            taxa.fun$Subdivision == "Rhizaria_X" | taxa.fun$Subdivision == "Rhodophyta_X" | taxa.fun$Subdivision == "Stramenopiles_X" | 
                            taxa.fun$Class == "Zygnemophyceae" & is.na(taxa.fun$Category), "Algae", taxa.fun$Category)

taxa.fun$Category <- ifelse(taxa.fun$Subdivision == "Fungi" & is.na(taxa.fun$Category), "Fungi", taxa.fun$Category)

taxa.fun$Category <- ifelse(taxa.fun$Subdivision == "Actinobacteria_X" | taxa.fun$Subdivision == "Firmicutes_X" | taxa.fun$Subdivision == "Proteobacteria_X" & is.na(taxa.fun$Category), "Bacteria", taxa.fun$Category)

taxa.fun$Category <- ifelse(taxa.fun$Subdivision == "Metazoa" & is.na(taxa.fun$Category), "Metazoa", taxa.fun$Category)

taxa.fun$Category <- ifelse(taxa.fun$Supergroup == "Archaeplastida" & is.na(taxa.fun$Category), "Algae", taxa.fun$Category)
taxa.fun$Category <- ifelse(taxa.fun$Supergroup == "Amoebozoa" & is.na(taxa.fun$Category), "Protist", taxa.fun$Category)

## Save taxonomic assignment with ecological function and trophic group information.
saveRDS(eco.function, "src/dada2/18S-eco.function.RDS"); saveRDS(taxa.fun, "src/dada2/18S-taxa-fun.RDS")

## STEP 2c: Phylogenetic analysis with phangorn

## Run sequence alignment (MSA) using DECIPHER.
sequences <- getSequences(seqtab.nosing); names(sequences) <- sequences ## Extract sequences from DADA2 output
alignment <- AlignSeqs(DNAStringSet(sequences), anchor = NA)

saveRDS(sequences, "src/dada2/18S-sequences.RDS") ## Save alignment.
saveRDS(alignment, "src/dada2/18S-alignment.RDS") ## Save alignment.

writeXStringSet(DNAStringSet(sequences), file = "src/dada2/18S-sequences.fasta") ## Save the alignment as FASTA file
writeXStringSet(alignment, file = "src/dada2/18S-alignment.fasta") ## Save the alignment as FASTA file

## STEP 2c: FastTree v2.1.11 (https://microbesonline.org/fasttree/)

#$ FastTree -gtr -nt src/dada2/18S-alignment.fasta > src/fasttree/18S.tree
## Save tree file at the "src/fasttree/" subdirectory.
```

### Generate ASV data for downstream analyses

```{r metadata, eval = FALSE}
## Import sample data information, i.e., metadata.txt.
metadata <- readxl::read_excel("../MS-Spatiotemporal-BIE_Supplementary-Tables.xlsx", sheet = "S1 Metadata")
metadata <- data.frame(metadata, row.names = 1)
metadata$River.Reach <- paste(metadata$River, metadata$Reach)
metadata$River.Season <- paste(metadata$River, metadata$Season)
metadata$River.Reach.Season <- paste(metadata$River, metadata$Reach, metadata$Season)
metadata$Type.River.Reach.Season <- paste(metadata$Type, metadata$River, metadata$Reach, metadata$Season)

## STEP 3b: Transformation and selection of environmental variables for downstream analysis.

## Environmental variables, except pH, were log10 (x + 1) transformed before analysis to meet the conditions of normality and homogeneity of variance in the residuals.
metadata.log <- metadata; metadata.log[,c(7:24)] <- log1p(abs(metadata.log[,c(7:24)]))

## For 16S rRNA Data
## STEP 3b: Import as phyloseq object for easier data processing.

## Create phyloseq object: contains OTU table (taxa abundances), sample metadata, taxonomy table (mapping between OTUs and higher-level taxonomic classifications), and phylogenetic tree (relations between the taxa).
bac.seqtab.nosing <- readRDS("src/dada2/16S_seqtab.nosing.RDS")
rownames(bac.seqtab.nosing) <- sub("B", "", rownames(bac.seqtab.nosing)) # Read otu table if run not saved.
bac.taxa <- readRDS("src/dada2/16S-taxa.RDS") # Read saved taxonomic annotation file if run not saved.
bac.tree <- read_tree("src/fasttree/16S.tree") ## Import tree file if fasttree was used.

## Create phloseq object.
bac.psq <- phyloseq(otu_table(bac.seqtab.nosing, taxa_are_rows = FALSE), 
                    phyloseq::tax_table(as.matrix(bac.taxa)), 
                    phy_tree(bac.tree), 
                    sample_data(metadata.log))
bac.psq ## phy_tree(tree)) if tree was generated with iqtree.

## ASVs renamed to more convenient shorter form (e.g., ASV100), and actual ASV sequences stored as a DNAStringSet (refseq slot).
bac.dna <- Biostrings::DNAStringSet(taxa_names(bac.psq))
names(bac.dna) <- taxa_names(bac.psq)
bac.psq <- merge_phyloseq(bac.psq, bac.dna)
taxa_names(bac.psq) <- paste0("ASV", seq(ntaxa(bac.psq)))
bac.psq

## Root the phylogenetic tree to calculate phylogeny based diversity metrics.
set.seed(0327)
phy_tree(bac.psq) <- ape::root(phy_tree(bac.psq), sample(taxa_names(bac.psq), 1), resolve.root = TRUE)
ape::is.rooted(phy_tree(bac.psq))
bac.psq

## Rarefaction visualization: MicrobiotaProcess provided ggrarecurve to plot the curves, based on rrarefy of vegan.
library("MicrobiotaProcess")
set.seed(0327) # For reproducibly random number
rareres <- get_rarecurve(obj = bac.psq, chunks = 200)

R1 <- ggrarecurve(obj = rareres, indexNames = c("Observe","Chao1","ACE")) + theme_bw() +
      scale_color_manual(values = colorRampPalette(paletteer::paletteer_d("rcartocolor::Antique"))(96)) + 
      theme(legend.spacing.y = unit(0.02,"cm"), legend.text = element_text(size = 6), legend.position = "none") + 
      guides(fill = guide_legend(ncol = 10)); print(R1)

## STEP 3c: Initial data filtering and transformation.

## Standardize abundances to the median sequencing depth.
bac.total <- median(sample_sums(bac.psq))
bac.standf <- function(x, t = bac.total) round(t * (x / sum(x)))
bac.psq.median <- transform_sample_counts(bac.psq, bac.standf)

## STEP 3d: Transform phyloseq object to a microtable for microeco R package analyses.

## Create a microtable object and perform additional filters, i.e., removing OTUs with the taxonomic assignments “mitochondria” or “chloroplast”, and making the ASV and sample information consistent across all files in the dataset object with the function tidy_dataset to trim the dataset.
bac.mco <- file2meco::phyloseq2meco(bac.psq.median)
bac.mco$filter_pollution(taxa = c("mitochondria", "chloroplast"))
bac.mco$tidy_dataset(); print(bac.mco)

## use R subset function to filter taxa in tax_table.
bac.mco$tax_table %<>% base::subset(Kingdom == "k__Archaea" | Kingdom == "k__Bacteria");
bac.mco$tidy_dataset()
print(bac.mco)

bac.mco$tax_table$Species <- ifelse(bac.mco$tax_table$Species == "s__" & grepl("^g__$", bac.mco$tax_table$Genus), paste0("s__", sub("g__", "", bac.mco$tax_table$Genus)), 
                                    ifelse(bac.mco$tax_table$Species == "s__", paste0("s__", sub("g__", "", bac.mco$tax_table$Genus), " sp."), 
                                           paste0("s__", sub("g__", "", bac.mco$tax_table$Genus), sub("s__", " ", bac.mco$tax_table$Species))))

## For the taxonomic dataset: Add "unclassified" name to filtered out annotations if higher taxa level is available (e.g. s__Unclassified Ruminococcaceae).
bac.mco$tax_table %<>% tidy_taxonomy
tmp <- bac.mco$tax_table; for(i in 1:nrow(tmp)){if(any(grepl("__$", tmp[i, ]))){matchall <- grep("__$", tmp[i, ])
tmp[i, matchall] <- paste0(tmp[i, matchall], "Unc. ", gsub(".__", "", tmp[i, matchall[1] - 1]))}}
bac.mco$tax_table <- tmp

bac.mco$tax_table$Species <- ifelse(grepl("^s__Unc. ", bac.mco$tax_table$Species), paste0(bac.mco$tax_table$Species, " sp."), bac.mco$tax_table$Species)
print(bac.mco)

## STEP 3e: Calculate abundance, alpha and beta diversity estimates.
bac.mco$cal_abund(); class(bac.mco$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
bac.mco$cal_alphadiv(PD = TRUE); class(bac.mco$alpha_diversity) ## Calculate the Alpha diversity w/ Faith's phylogenetic diversity.
bac.mco$cal_betadiv(unifrac = TRUE); class(bac.mco$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
print(bac.mco)

## Extract ASV-Table with sequence, abundance/sample and taxonomic information.
write.csv(cbind(bac.mco$otu_table, bac.mco$tax_table), "src/output/16S-asv_table.tsv", quote = FALSE)
saveRDS(bac.mco, "src/bac.mco.RDS") # Save microtable used for downstream analyses.

## For 18S rRNA Data
## STEP 3b: Import as phyloseq object for easier data processing.

## Create phyloseq object: contains OTU table (taxa abundances), sample metadata, taxonomy table (mapping between OTUs and higher-level taxonomic classifications), and phylogenetic tree (relations between the taxa).
euk.seqtab.nosing <- readRDS("src/dada2/18S_seqtab.nosing.RDS")
rownames(euk.seqtab.nosing) <- sub("E", "", rownames(euk.seqtab.nosing)) # Read otu table if run not saved.
euk.taxa <- readRDS("src/dada2/18S-taxa-fun.RDS") # Read saved taxonomic annotation file if run not saved.
euk.tree <- read_tree("src/fasttree/18S.tree") ## Import tree file if fasttree was used.

## Cresate phloseq object.
euk.psq <- phyloseq(otu_table(euk.seqtab.nosing, taxa_are_rows = FALSE),
                    phyloseq::tax_table(as.matrix(euk.taxa)), 
                    phy_tree(euk.tree), sample_data(metadata.log))
euk.psq ## phy_tree(tree)) if tree was generated with iqtree.

## ASVs renamed to more convenient shorter form (e.g., ASV100), and actual ASV sequences stored as a DNAStringSet (refseq slot).
euk.dna <- Biostrings::DNAStringSet(taxa_names(euk.psq))
names(euk.dna) <- taxa_names(euk.psq)
euk.psq <- merge_phyloseq(euk.psq, euk.dna)
taxa_names(euk.psq) <- paste0("ASV", seq(ntaxa(euk.psq)))
euk.psq

## Root the phylogenetic tree to calculate phylogeny based diversity metrics.
set.seed(0327)
phy_tree(euk.psq) <- ape::root(phy_tree(euk.psq), sample(taxa_names(euk.psq), 1), resolve.root = TRUE)
ape::is.rooted(phy_tree(euk.psq))
euk.psq

## Rarefaction visualization: MicrobiotaProcess provided ggrarecurve to plot the curves, based on rrarefy of vegan.
library("MicrobiotaProcess")
set.seed(0327) # For reproducibly random number
rareres <- get_rarecurve(obj = euk.psq, chunks = 200)

R2 <- ggrarecurve(obj = rareres, indexNames = c("Observe","Chao1","ACE")) + theme_bw() +
      scale_color_manual(values = colorRampPalette(paletteer::paletteer_d("rcartocolor::Antique"))(96)) + 
      theme(legend.spacing.y = unit(0.02,"cm"), legend.text = element_text(size = 6), legend.position = "none") + 
      guides(fill = guide_legend(ncol = 10)); print(R2)

## STEP 3c: Initial data filtering and transformation.

## Standardize abundances to the median sequencing depth.
euk.total <- median(sample_sums(euk.psq))
euk.standf <- function(x, t = euk.total) round(t * (x / sum(x)))
euk.psq.median <- transform_sample_counts(euk.psq, euk.standf)

## STEP 3d: Transform phyloseq object to a microtable for microeco R package analyses.

## Create a microtable object and perform additional filters, i.e., removing OTUs with the taxonomic assignments “mitochondria” or “chloroplast”, and making the ASV and sample information consistent across all files in the dataset object with the function tidy_dataset to trim the dataset.
euk.mco <- file2meco::phyloseq2meco(euk.psq.median)
euk.mco$filter_pollution(taxa = c("mitochondria", "chloroplast"))
euk.mco$tidy_dataset()
print(euk.mco)

## For the taxonomic dataset: Add "unclassified" name to filtered out annotations if higher taxa level is available (e.g. s__Unclassified Ruminococcaceae).
euk.mco$tax_table %<>% tidy_taxonomy
tmp <- euk.mco$tax_table
for (i in 1:nrow(tmp)) {
  if (any(grepl("__$", tmp[i, ]))) {
    matchall <- grep("__$", tmp[i, ])
    # Ensure columns "Function", "Trophic", and "Category" are not modified
    matchall <- matchall[!colnames(tmp)[matchall] %in% c("Function", "Trophic", "Category")]
    tmp[i, matchall] <- paste0(tmp[i, matchall], "Unc. ", gsub(".__", "", tmp[i, matchall[1] - 1]))
  }
}
euk.mco$tax_table <- tmp

euk.mco$tax_table$Species <- ifelse(grepl("^s__Unc. ", euk.mco$tax_table$Species), paste0(euk.mco$tax_table$Species, " sp."), euk.mco$tax_table$Species)
print(euk.mco)

## Use R subset function to filter taxa in tax_table.
euk.mco$tax_table %<>% base::subset(Domain == "d__Eukaryota"); euk.mco$tidy_dataset(); print(euk.mco)
euk.mco$tax_table %<>% base::subset(Class != "c__Embryophyceae"); euk.mco$tidy_dataset(); print(euk.mco)

colnames(euk.mco$tax_table)[4] <- "Subdivision"

## STEP 3e: Calculate abundance, alpha and beta diversity estimates.
euk.mco$cal_abund(); class(euk.mco$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
euk.mco$cal_alphadiv(PD = TRUE); class(euk.mco$alpha_diversity) ## Calculate the Alpha diversity w/ Faith's phylogenetic diversity.
euk.mco$cal_betadiv(unifrac = TRUE); class(euk.mco$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
print(euk.mco)

## Extract ASV-Table with sequence, abundance/sample and taxonomic information.
write.csv(cbind(euk.mco$otu_table, euk.mco$tax_table), "src/output/18S-asv_table.tsv", quote = FALSE)
saveRDS(euk.mco, "src/euk.mco.RDS") # Save microtable used for downstream analyses.

## Create Subsets of the microtables

bac.mco.fld <- clone(bac.mco); bac.mco.fld$sample_table %<>% base::subset(Type == "Field"); bac.mco.fld$tidy_dataset(); print(bac.mco.fld)
bac.mco.fld$cal_abund(); class(bac.mco.fld$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
bac.mco.fld$cal_alphadiv(PD = TRUE); class(bac.mco.fld$alpha_diversity) ## Calculate the Alpha diversity w/ Faith's phylogenetic diversity.
bac.mco.fld$cal_betadiv(unifrac = TRUE); class(bac.mco.fld$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
print(bac.mco.fld)

bac.mco.btc <- clone(bac.mco); bac.mco.btc$sample_table %<>% base::subset(Type == "Batch"); bac.mco.btc$tidy_dataset(); print(bac.mco.btc)
bac.mco.btc$cal_abund(); class(bac.mco.btc$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
bac.mco.btc$cal_alphadiv(PD = TRUE); class(bac.mco.btc$alpha_diversity) ## Calculate the Alpha diversity w/ Faith's phylogenetic diversity.
bac.mco.btc$cal_betadiv(unifrac = TRUE); class(bac.mco.btc$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
print(bac.mco.btc)

euk.mco.fld <- clone(euk.mco); euk.mco.fld$sample_table %<>% base::subset(Type == "Field"); euk.mco.fld$tidy_dataset(); print(euk.mco.fld)
euk.mco.fld$cal_abund(); class(euk.mco.fld$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
euk.mco.fld$cal_alphadiv(PD = TRUE); class(euk.mco.fld$alpha_diversity) ## Calculate the Alpha diversity w/ Faith's phylogenetic diversity.
euk.mco.fld$cal_betadiv(unifrac = TRUE); class(euk.mco.fld$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
print(euk.mco.fld)

euk.mco.btc <- clone(euk.mco); euk.mco.btc$sample_table %<>% base::subset(Type == "Batch"); euk.mco.btc$tidy_dataset(); print(euk.mco.btc)
euk.mco.btc$cal_abund(); class(euk.mco.btc$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
euk.mco.btc$cal_alphadiv(PD = TRUE); class(euk.mco.btc$alpha_diversity) ## Calculate the Alpha diversity w/ Faith's phylogenetic diversity.
euk.mco.btc$cal_betadiv(unifrac = TRUE); class(euk.mco.btc$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
print(euk.mco.btc)

## Create subsets based on ecological categories
alg.mco.fld <- clone(euk.mco.fld); alg.mco.fld$tax_table %<>% base::subset(Category == "c__Algae"); alg.mco.fld$tidy_dataset(); print(alg.mco.fld)
alg.mco.fld$cal_abund(); class(alg.mco.fld$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
alg.mco.fld$cal_alphadiv(PD = TRUE); class(alg.mco.fld$alpha_diversity) ## Calculate the Alpha diversity w/ Faith's phylogenetic diversity.
alg.mco.fld$cal_betadiv(unifrac = TRUE); class(alg.mco.fld$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
print(alg.mco.fld)

fun.mco.fld <- clone(euk.mco.fld); fun.mco.fld$tax_table %<>% base::subset(Category == "c__Fungi"); fun.mco.fld$tidy_dataset(); print(fun.mco.fld)
fun.mco.fld$cal_abund(); class(fun.mco.fld$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
fun.mco.fld$cal_alphadiv(PD = TRUE); class(fun.mco.fld$alpha_diversity) ## Calculate the Alpha diversity w/ Faith's phylogenetic diversity.
fun.mco.fld$cal_betadiv(unifrac = TRUE); class(fun.mco.fld$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
print(fun.mco.fld)

met.mco.fld <- clone(euk.mco.fld); met.mco.fld$tax_table %<>% base::subset(Category == "c__Metazoa"); met.mco.fld$tidy_dataset(); print(met.mco.fld)
met.mco.fld$cal_abund(); class(met.mco.fld$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
met.mco.fld$cal_alphadiv(PD = TRUE); class(met.mco.fld$alpha_diversity) ## Calculate the Alpha diversity w/ Faith's phylogenetic diversity.
met.mco.fld$cal_betadiv(unifrac = TRUE); class(met.mco.fld$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
print(met.mco.fld)

pro.mco.fld <- clone(euk.mco.fld); pro.mco.fld$tax_table %<>% base::subset(Category == "c__Protist"); pro.mco.fld$tidy_dataset(); print(pro.mco.fld)
pro.mco.fld$cal_abund(); class(pro.mco.fld$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
pro.mco.fld$cal_alphadiv(PD = TRUE); class(pro.mco.fld$alpha_diversity) ## Calculate the Alpha diversity w/ Faith's phylogenetic diversity.
pro.mco.fld$cal_betadiv(unifrac = TRUE); class(pro.mco.fld$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
print(pro.mco.fld)

alg.mco.btc <- clone(euk.mco.btc); alg.mco.btc$tax_table %<>% base::subset(Category == "c__Algae"); alg.mco.btc$tidy_dataset(); print(alg.mco.btc)
alg.mco.btc$cal_abund(); class(alg.mco.btc$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
alg.mco.btc$cal_alphadiv(PD = TRUE); class(alg.mco.btc$alpha_diversity) ## Calculate the Alpha diversity w/ Faith's phylogenetic diversity.
alg.mco.btc$cal_betadiv(unifrac = TRUE); class(alg.mco.btc$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
print(alg.mco.btc)

fun.mco.btc <- clone(euk.mco.btc); fun.mco.btc$tax_table %<>% base::subset(Category == "c__Fungi"); fun.mco.btc$tidy_dataset(); print(fun.mco.btc)
fun.mco.btc$cal_abund(); class(fun.mco.btc$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
fun.mco.btc$cal_alphadiv(PD = TRUE); class(fun.mco.btc$alpha_diversity) ## Calculate the Alpha diversity w/ Faith's phylogenetic diversity.
fun.mco.btc$cal_betadiv(unifrac = TRUE); class(fun.mco.btc$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
print(fun.mco.btc)

met.mco.btc <- clone(euk.mco.btc); met.mco.btc$tax_table %<>% base::subset(Category == "c__Metazoa"); met.mco.btc$tidy_dataset(); print(met.mco.btc)
met.mco.btc$cal_abund(); class(met.mco.btc$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
met.mco.btc$cal_alphadiv(PD = TRUE); class(met.mco.btc$alpha_diversity) ## Calculate the Alpha diversity w/ Faith's phylogenetic diversity.
met.mco.btc$cal_betadiv(unifrac = TRUE); class(met.mco.btc$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
print(met.mco.btc)

pro.mco.btc <- clone(euk.mco.btc); pro.mco.btc$tax_table %<>% base::subset(Category == "c__Protist"); pro.mco.btc$tidy_dataset(); print(pro.mco.btc)
pro.mco.btc$cal_abund(); class(pro.mco.btc$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
pro.mco.btc$cal_alphadiv(PD = TRUE); class(pro.mco.btc$alpha_diversity) ## Calculate the Alpha diversity w/ Faith's phylogenetic diversity.
pro.mco.btc$cal_betadiv(unifrac = TRUE); class(pro.mco.btc$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
print(pro.mco.btc)

## Save subsetted microtables
saveRDS(bac.mco.fld, "src/bac.mco.fld.RDS"); saveRDS(bac.mco.btc, "src/bac.mco.btc.RDS")
saveRDS(euk.mco.fld, "src/euk.mco.fld.RDS"); saveRDS(euk.mco.btc, "src/euk.mco.btc.RDS")

saveRDS(alg.mco.fld, "src/alg.mco.fld.RDS"); saveRDS(fun.mco.fld, "src/fun.mco.fld.RDS"); saveRDS(met.mco.fld, "src/met.mco.fld.RDS"); saveRDS(pro.mco.fld, "src/pro.mco.fld.RDS")
saveRDS(alg.mco.btc, "src/alg.mco.btc.RDS"); saveRDS(fun.mco.btc, "src/fun.mco.btc.RDS"); saveRDS(met.mco.btc, "src/met.mco.btc.RDS"); saveRDS(pro.mco.btc, "src/pro.mco.btc.RDS")
```

### Predict functional profile w/ Tax4Fun2

```{r, eval = FALSE}
# Tax4Fun2 (Wemheuer et al. 2020) is an R package for the prediction of functional profiles of prokaryotic communities from 16S rRNA gene sequences. It also provides two indexes for the evaluation of functional gene redundancies. If the user want to use Tax4Fun2 method, the representative fasta file is necessary to be added in the microtable object. https://chiliubio.github.io/microeco_tutorial/intro.html#tax4fun2 for installation

dataset <- clone(bac.mco.fld)
t1 <- trans_func$new(dataset) # create object of trans_func

dir.create("src/prediction") # create a directory for result and log files

t1$cal_tax4fun2(blast_tool_path = "data/database/ncbi-blast-2.5.0+/bin",
                path_to_reference_data = "data/database/Tax4Fun2_ReferenceData_v2", 
                database_mode = "Ref99NR", path_to_temp_folder = "src/prediction")

t1$cal_tax4fun2_FRI() # calculate functional redundancies
data(Tax4Fun2_KEGG) # prepare feature table and metadata

# create a microtable object for pathways
keg.mco.fld <- microtable$new(otu_table = t1$res_tax4fun2_pathway, tax_table = Tax4Fun2_KEGG$ptw_desc,
                              sample_table = dataset$sample_table)

keg.mco.fld$tidy_dataset()

# calculate relative abundances at three levels: Level.1, Level.2, Level.3
keg.mco.fld$cal_abund(); class(keg.mco.fld$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
keg.mco.fld$cal_alphadiv(PD = FALSE); class(keg.mco.fld$alpha_diversity) ## Calculate the Alpha diversity w/ Faith's phylogenetic diversity.
keg.mco.fld$cal_betadiv(unifrac = FALSE); class(keg.mco.fld$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
  
print(keg.mco.fld)

# save data
saveRDS(keg.mco.fld, "src/keg.mco.fld.RDS")

abundance <- as.data.frame(keg.mco.fld$otu_table); kegg <- as.data.frame(keg.mco.fld$tax_table); abundance$ID <- rownames(abundance); kegg$ID <- rownames(kegg)

merged_data <- merge(abundance, kegg, by = "ID")
openxlsx::write.xlsx(merged_data, file = "src/KEGG_pathways.xlsx", rowNames = FALSE)

# Subset
keg.mco.met <- clone(keg.mco.fld)
keg.mco.met$tax_table %<>% base::subset(Level.1 == "Metabolism"); keg.mco.met$tidy_dataset(); print(keg.mco.met)
keg.mco.met$cal_abund(); class(keg.mco.met$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
keg.mco.met$cal_betadiv(unifrac = FALSE); class(keg.mco.met$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
print(keg.mco.met)
saveRDS(keg.mco.met, "src/keg.mco.met.RDS")

keg.mco.xen <- clone(keg.mco.xen)
keg.mco.xen$tax_table %<>% base::subset(Level.2 == "Xenobiotics biodegradation and metabolism"); keg.mco.xen$tidy_dataset(); print(keg.mco.xen)
keg.mco.xen$cal_abund(); class(keg.mco.xen$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
keg.mco.xen$cal_betadiv(unifrac = FALSE); class(keg.mco.xen$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
print(keg.mco.xen)
saveRDS(keg.mco.xen, "src/keg.mco.xen.RDS")
```

### Generate bidegradation profile data

```{r, eval = FALSE}
## Import sample data information, i.e., metadata.txt.
metadata <- readxl::read_excel("../MS-Spatiotemporal-BIE_Supplementary-Tables.xlsx", sheet = "S1 Metadata")
metadata <- data.frame(metadata, row.names = 1)
metadata$River.Reach <- paste(metadata$River, metadata$Reach)
metadata$River.Season <- paste(metadata$River, metadata$Season)
metadata$River.Reach.Season <- paste(metadata$River, metadata$Reach, metadata$Season)
metadata$Type.River.Reach.Season <- paste(metadata$Type, metadata$River, metadata$Reach, metadata$Season)

biodeg.rates <- readxl::read_excel("../MS-Spatiotemporal-BIE_Supplementary-Tables.xlsx", sheet = "S2 kTCC,pH7")
biodeg.rates <- data.frame(biodeg.rates, row.names = 1)

# [NOT PEPRFROMED] Find the minimum value in the data frame and add the constant to every value in the data frame
#min_val <- min(biodeg.rates[c(1:96)], na.rm = TRUE); constant <- abs(min_val)
#biodeg.rates.transformed <- biodeg.rates[c(1:96)] + constant

biodeg.rates.transformed <- apply(biodeg.rates[c(1:96)], 2, function(x) ifelse(x < 0, 0, x)) # Transform negative biodegradation rates to 0.
biodeg.rates.transformed[is.na(biodeg.rates.transformed)] <- 0 # Replace all NA values with 0.

# Create phyloseq object
bdg.psq <- phyloseq(otu_table(as.matrix(biodeg.rates.transformed), taxa_are_rows = TRUE),
                    phyloseq::tax_table(as.matrix(biodeg.rates[c(98:104)])),
                    sample_data(metadata))

# Create microtable object for microeco downstream analyses
bdg.mco <- file2meco::phyloseq2meco(bdg.psq, auto_tidy = FALSE); bdg.mco$tidy_dataset()

bdg.mco.fld <- clone(bdg.mco); bdg.mco.fld$sample_table %<>% base::subset(Type == "Field"); bdg.mco.fld$tidy_dataset(); print(bdg.mco.fld)
bdg.mco.fld$cal_abund(); class(bdg.mco.fld$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
bdg.mco.fld$cal_alphadiv(PD = FALSE); class(bdg.mco.fld$alpha_diversity) ## Calculate the Alpha diversity w/ Faith's phylogenetic diversity.
bdg.mco.fld$cal_betadiv(unifrac = FALSE); class(bdg.mco.fld$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
print(bdg.mco.fld)

# Add abundance and beta diversity estimates
# Calculate the taxa abundance at each taxonomic rank using cal_abund()
bdg.mco.fld$cal_abund()
# Calculate alpha diversity
bdg.mco.fld$cal_alphadiv(PD = FALSE)
# Calculate beta diversity
bdg.mco.fld$cal_betadiv(unifrac = FALSE)

# Save files as RDS
saveRDS(bdg.mco.fld, "src/bdg.mco.fld.RDS")
```

## ########################################################## ##
##                                                            ##
##        DOWNSTREAM ANALYSES - BIODEGRADATION PROFILE        ##
##                                                            ##
## ########################################################## ##

### Visualizing the environmental parameters

```{r, correlation anaysis between environmental parameters}
## There may be some NA (missing value) in the user’s env data. If so, please add complete_na = TRUE for the interpolation when creating a trans_env object.

dataset <- clone(bac.mco.fld)
dataset$sample_table$River.Reach.Season %<>% factor(., levels = c("Knivstaån Up Summer", "Knivstaån Up Autumn", "Knivstaån Up Winter", "Knivstaån Up Spring", "Knivstaån Dn Summer", "Knivstaån Dn Autumn", "Knivstaån Dn Winter", "Knivstaån Dn Spring", "Vitsån Up Summer", "Vitsån Up Autumn", "Vitsån Up Winter", "Vitsån Up Spring", "Vitsån Dn Summer", "Vitsån Dn Autumn", "Vitsån Dn Winter", "Vitsån Dn Spring"))

# Use 'env_cols' to specify the columns containing environmental data
t1 <- trans_env$new(dataset = dataset, env_cols = c(6:24), complete_na = TRUE); t1$data_env

## Generally, it is beneficial to analyze environmental variables in order to better use more methods. The cal_diff function is used to test the significance of variables across groups like we have shown in trans_alpha and trans_diff class parts.

t1$cal_diff(method = "anova", group = "River.Reach.Season") # place all the plots into a list
tmp <- list(); 
for(i in colnames(t1$data_env)){ 
  tmp[[i]] <- t1$plot_diff(measure = i, add_sig_text_size = 2, group_order = "River.Reach.Season", color_values = "black") + 
              geom_boxplot(outlier.shape = NA) + geom_boxplot(fill = "black", color = NA, alpha = 0.5) + 
              stat_summary(fun = median, geom = "line", aes(group = 1), col = "red") + 
              ggtitle(i) + 
              theme(plot.title = element_text(size = 10, face = "bold"), 
                    axis.text.y = element_text(size = 6, angle = 90, hjust = 0.5),
                    axis.text.x = element_blank(),
                    #axis.text.x = element_text(size = 6, angle = 90, vjust = 0.5),
                    axis.title.x = element_blank(), axis.title.y = element_blank(),legend.position = "none") }

E1 <- plot(gridExtra::arrangeGrob(grobs = tmp[1:19], ncol = 4)); print(E1)

#library("GGally")
#t1$cal_autocor() + theme(axis.text.y = element_text(size = 5), axis.text.x = element_text(size = 5, angle = 90, vjust = 0.5),
#                         axis.title.y = element_text(size = 5), axis.title.x = element_text(size = 5),
#                         strip.text = element_text(size = 8))

library("ComplexHeatmap"); library("circlize"); library("Hmisc")
data_subset <- dataset$sample_table[6:24]
cor_results <- rcorr(as.matrix(data_subset), type = "pearson")
cor_matrix <- cor_results$r
p_matrix <- cor_results$P

# Define color function
col_fun <- colorRamp2(c(-1, 0, 1), c("#4169E1", "white", "#F08080"))

# Create heatmap
Heatmap(cor_matrix,
        name = "Pearson\nCorrelation",
        #clustering_distance_rows = "pearson",
        #clustering_method_rows = "complete", 
        col = col_fun, 
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(signif_matrix[i, j], x, y, gp = gpar(col = "black", fontsize = 10))
        })

openxlsx::write.xlsx(cor_results$P, file = "src/envi.correlation.xlsx", rowNames = FALSE)
```

### Field versus batch samples (Pre-assessment)

```{r, 16S alpha diversity}
# Plot alpha diversity

# Clone the dataset and redefine Region levels
dataset <- clone(bac.mco)
t1 <- trans_alpha$new(dataset = dataset, group = "Type"); t1$data_stat
t1$cal_diff(method = "t.test"); t1$res_diff

alpha.tbl <- as.data.frame(dataset$alpha_diversity)
alpha.tbl$Sample <- rownames(dataset$alpha_diversity)
metadata$Sample <- rownames(metadata)

alpha.tbl <- merge(metadata, alpha.tbl, by = "Sample", all.x = TRUE)

# Reorder the 'Type' factor to ensure Field comes before Batch
alpha.tbl$Type <- factor(alpha.tbl$Type, levels = c("Field", "Batch"))

# Plot

# Create the violin plot with specific colors, black borders, a box inside the violin, and no outliers
aa <- ggviolin(alpha.tbl, x = "Type", y = "Observed", color = "black", 
               palette = c("Field" = "#1B9E77", "Batch" = "#D95F02"), add = "boxplot", 
               outlier.shape = NA, fill = "Type", alpha = 0.5) +
               geom_jitter(aes(color = Type), alpha = 0.25, shape = 16, size = 1, stroke = 1, color = "gray50") +
               stat_compare_means(method = "t.test", label = "p.signif", label.x.npc = "middle", size = 5) +
               labs(x = NULL, y = "Observed") +
               theme(axis.text.x = element_blank(),
                     axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
                     legend.position = "none"); aa

ab <- ggviolin(alpha.tbl, x = "Type", y = "Chao1", color = "black", 
               palette = c("Field" = "#1B9E77", "Batch" = "#D95F02"), add = "boxplot", 
               outlier.shape = NA, fill = "Type", alpha = 0.5) +
               geom_jitter(aes(color = Type), alpha = 0.25, shape = 16, size = 1, stroke = 1, color = "gray50") +
               stat_compare_means(method = "t.test", label = "p.signif", label.x.npc = "middle", size = 5) +
               labs(x = NULL, y = "Chao1 Richness") +
               theme(axis.text.x = element_blank(),
                     axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
                     legend.position = "none"); ab

ac <- ggviolin(alpha.tbl, x = "Type", y = "Shannon", color = "black", 
               palette = c("Field" = "#1B9E77", "Batch" = "#D95F02"), add = "boxplot", 
               outlier.shape = NA, fill = "Type", alpha = 0.5) +
               geom_jitter(aes(color = Type), alpha = 0.25, shape = 16, size = 1, stroke = 1, color = "gray50") +
               stat_compare_means(method = "t.test", label = "p.signif", label.x.npc = "middle", size = 5) +
               labs(x = NULL, y = "Shannon Diversity") +
               theme(axis.text.x = element_blank(),
                     axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
                     legend.position = "none"); ac

ad <- ggviolin(alpha.tbl, x = "Type", y = "Simpson", color = "black", 
               palette = c("Field" = "#1B9E77", "Batch" = "#D95F02"), add = "boxplot", 
               outlier.shape = NA, fill = "Type", alpha = 0.5) +
               geom_jitter(aes(color = Type), alpha = 0.25, shape = 16, size = 1, stroke = 1, color = "gray50") +
               stat_compare_means(method = "t.test", label = "p.signif", label.x.npc = "middle", size = 5) +
               labs(x = NULL, y = "Simpson Diversity") +
               theme(axis.text.x = element_blank(),
                     axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
                     legend.position = "none"); ad

ae <- ggviolin(alpha.tbl, x = "Type", y = "Pielou", color = "black", 
               palette = c("Field" = "#1B9E77", "Batch" = "#D95F02"), add = "boxplot", 
               outlier.shape = NA, fill = "Type", alpha = 0.5) +
               geom_jitter(aes(color = Type), alpha = 0.25, shape = 16, size = 1, stroke = 1, color = "gray50") +
               stat_compare_means(method = "t.test", label = "p.signif", label.x.npc = "middle", size = 5) +
               labs(x = NULL, y = "Pielou's Evenness") +
               theme(axis.text.x = element_blank(),
                     axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
                     legend.position = "none"); ae

af <- ggviolin(alpha.tbl, x = "Type", y = "PD", color = "black", 
               palette = c("Field" = "#1B9E77", "Batch" = "#D95F02"), add = "boxplot", 
               outlier.shape = NA, fill = "Type", alpha = 0.5) +
               geom_jitter(aes(color = Type), alpha = 0.25, shape = 16, size = 1, stroke = 1, color = "gray50") +
               stat_compare_means(method = "t.test", label = "p.signif", label.x.npc = "middle", size = 5) +
               labs(x = NULL, y = "Faith's Phylogenetic Diversity") +
               theme(axis.text.x = element_blank(),
                     axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
                     legend.position = "none"); af

# Combine the plots in one row
a1 <- ( aa | ab | ac ) / ( ad  | ae  | af ) + patchwork::plot_layout(guides = 'collect') + patchwork::plot_annotation(); a1
```

```{r, 18S alpha diversity}
# Plot alpha diversity

# Clone the dataset and redefine Region levels
dataset <- clone(euk.mco)
t1 <- trans_alpha$new(dataset = dataset, group = "Type"); t1$data_stat
t1$cal_diff(method = "t.test"); t1$res_diff

alpha.tbl <- as.data.frame(dataset$alpha_diversity)
alpha.tbl$Sample <- rownames(dataset$alpha_diversity)
metadata$Sample <- rownames(metadata)

alpha.tbl <- merge(metadata, alpha.tbl, by = "Sample", all.x = TRUE)

# Reorder the 'Type' factor to ensure Field comes before Batch
alpha.tbl$Type <- factor(alpha.tbl$Type, levels = c("Field", "Batch"))

# Plot

# Create the violin plot with specific colors, black borders, a box inside the violin, and no outliers
aa <- ggviolin(alpha.tbl, x = "Type", y = "Observed", color = "black", 
               palette = c("Field" = "#1B9E77", "Batch" = "#D95F02"), add = "boxplot", 
               outlier.shape = NA, fill = "Type", alpha = 0.5) +
               geom_jitter(aes(color = Type), alpha = 0.25, shape = 16, size = 1, stroke = 1, color = "gray50") +
               stat_compare_means(method = "t.test", label = "p.signif", label.x.npc = "middle", size = 5) +
               labs(x = NULL, y = "Observed") +
               theme(axis.text.x = element_blank(),
                     axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
                     legend.position = "none"); aa

ab <- ggviolin(alpha.tbl, x = "Type", y = "Chao1", color = "black", 
               palette = c("Field" = "#1B9E77", "Batch" = "#D95F02"), add = "boxplot", 
               outlier.shape = NA, fill = "Type", alpha = 0.5) +
               geom_jitter(aes(color = Type), alpha = 0.25, shape = 16, size = 1, stroke = 1, color = "gray50") +
               stat_compare_means(method = "t.test", label = "p.signif", label.x.npc = "middle", size = 5) +
               labs(x = NULL, y = "Chao1 Richness") +
               theme(axis.text.x = element_blank(),
                     axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
                     legend.position = "none"); ab

ac <- ggviolin(alpha.tbl, x = "Type", y = "Shannon", color = "black", 
               palette = c("Field" = "#1B9E77", "Batch" = "#D95F02"), add = "boxplot", 
               outlier.shape = NA, fill = "Type", alpha = 0.5) +
               geom_jitter(aes(color = Type), alpha = 0.25, shape = 16, size = 1, stroke = 1, color = "gray50") +
               stat_compare_means(method = "t.test", label = "p.signif", label.x.npc = "middle", size = 5) +
               labs(x = NULL, y = "Shannon Diversity") +
               theme(axis.text.x = element_blank(),
                     axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
                     legend.position = "none"); ac

ad <- ggviolin(alpha.tbl, x = "Type", y = "Simpson", color = "black", 
               palette = c("Field" = "#1B9E77", "Batch" = "#D95F02"), add = "boxplot", 
               outlier.shape = NA, fill = "Type", alpha = 0.5) +
               geom_jitter(aes(color = Type), alpha = 0.25, shape = 16, size = 1, stroke = 1, color = "gray50") +
               stat_compare_means(method = "t.test", label = "p.signif", label.x.npc = "middle", size = 5) +
               labs(x = NULL, y = "Simpson Diversity") +
               theme(axis.text.x = element_blank(),
                     axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
                     legend.position = "none"); ad

ae <- ggviolin(alpha.tbl, x = "Type", y = "Pielou", color = "black", 
               palette = c("Field" = "#1B9E77", "Batch" = "#D95F02"), add = "boxplot", 
               outlier.shape = NA, fill = "Type", alpha = 0.5) +
               geom_jitter(aes(color = Type), alpha = 0.25, shape = 16, size = 1, stroke = 1, color = "gray50") +
               stat_compare_means(method = "t.test", label = "p.signif", label.x.npc = "middle", size = 5) +
               labs(x = NULL, y = "Pielou's Evenness") +
               theme(axis.text.x = element_blank(),
                     axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
                     legend.position = "none"); ae

af <- ggviolin(alpha.tbl, x = "Type", y = "PD", color = "black", 
               palette = c("Field" = "#1B9E77", "Batch" = "#D95F02"), add = "boxplot", 
               outlier.shape = NA, fill = "Type", alpha = 0.5) +
               geom_jitter(aes(color = Type), alpha = 0.25, shape = 16, size = 1, stroke = 1, color = "gray50") +
               stat_compare_means(method = "t.test", label = "p.signif", label.x.npc = "middle", size = 5) +
               labs(x = NULL, y = "Faith's Phylogenetic Diversity") +
               theme(axis.text.x = element_blank(),
                     axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
                     legend.position = "none"); af

# Combine the plots in one row
a2 <- ( aa | ab | ac ) / ( ad  | ae  | af ) + patchwork::plot_layout(guides = 'collect') + patchwork::plot_annotation(); a2
```

```{r, PCoA, CIA and Mantel test}
## ASV-level analysis
Field.psq <- file2meco::meco2phyloseq(bac.mco.fld)
Batch.psq <- file2meco::meco2phyloseq(bac.mco.btc)

## Create dataframe from otu table
Field.cia <- as.data.frame(otu_table(Field.psq)); Field.cia <- labdsv::hellinger(Field.cia); dim(Field.cia)
Batch.cia <- as.data.frame(otu_table(Batch.psq)); Batch.cia <- labdsv::hellinger(Batch.cia); dim(Batch.cia)

library("omicade4")
sp.ct <- list(Field.cia, Batch.cia) ## Merge datasets into one list
sapply(sp.ct, dim) ## Check the dimension of each dataset in the list

mcoin <- mcia(sp.ct, cia.nf = 10); class(mcoin)

type <- t(bac.mco.fld$sample_table$River.Season); type
plot(mcoin, axes = 1:2, phenovec = type, sample.lab = FALSE, df.color = c("#1B9E77","#D95F02"))
mcoin$mcoa$RV

## Beta diversity comparison
dataset <- clone(bac.mco)
dataset$sample_table$Type %<>% factor(., levels = c("Field", "Batch"))

# PCoA
t1 <- trans_beta$new(dataset = dataset, group = "Type", measure = "wei_unifrac")
t1$cal_ordination(method = "PCoA")

# extract the axis scores
tmp <- t1$res_ordination$scores
# differential test with trans_env class
t2 <- trans_env$new(dataset = dataset, add_data = tmp[, 1:2])
# 'KW_dunn' for non-parametric test
t2$cal_diff(group = "Type", method = "t.test")

p1 <- t1$plot_ordination(plot_color = "Type", plot_shape = "Type", ellipse_chull_alpha = 0.6,
                         #add_sample_label = "Reach", #plot_type = c("chull"), 
                         shape_values = c(19,17), plot_type = c("point")) + geom_vline(xintercept = 0, linetype = "dashed", color = "gray") + geom_hline(yintercept = 0, linetype = "dashed", color = "gray") + 
                         theme_bw() +
                         theme(panel.grid.minor = element_blank(), panel.grid.major = element_blank())

# groups order in p2 is same with p1; use legend.position = "none" to remove redundant legend
p2 <- t2$plot_diff(measure = "PCo1", add_sig = T) + coord_flip() + theme_bw() + #border() +
      #geom_boxplot(fill = col.river.season, color = "black", alpha = 0.6) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray") +
      theme(legend.position = "none", axis.title.x = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank())
p3 <- t2$plot_diff(measure = "PCo2", add_sig = T) + theme_bw() +
      #geom_boxplot(fill = col.river.season, color = "black", alpha = 0.6) + 
      theme(legend.position = "none", axis.title.y = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank())
# height of the upper figure and width of the right-hand figure are both 0.2-fold of the main figure
v1 <- p1 %>% aplot::insert_top(p2, height = 0.25) %>% aplot::insert_right(p3, width = 0.25)
v1

## PerMANOVA (Anderson 2001) for the differential test of distances among groups via the cal_manova function developed based on the adonis2 function of vegan package.
#t1$cal_manova(manova_all = TRUE); t1$res_manova
t1$cal_manova(manova_set = "Type"); t1$res_manova

## Beta diversity comparison

dataset <- clone(euk.mco)
dataset$sample_table$Type %<>% factor(., levels = c("Field", "Batch"))

# PCoA
t1 <- trans_beta$new(dataset = dataset, group = "Type", measure = "unwei_unifrac")
t1$cal_ordination(method = "PCoA")

# extract the axis scores
tmp <- t1$res_ordination$scores
# differential test with trans_env class
t2 <- trans_env$new(dataset = dataset, add_data = tmp[, 1:2])
# 'KW_dunn' for non-parametric test
t2$cal_diff(group = "Type", method = "t.test")

p1 <- t1$plot_ordination(plot_color = "Type", plot_shape = "Type", ellipse_chull_alpha = 0.6,
                         #add_sample_label = "Reach", #plot_type = c("chull"), 
                         shape_values = c(19,17), plot_type = c("point")) + geom_vline(xintercept = 0, linetype = "dashed", color = "gray") + geom_hline(yintercept = 0, linetype = "dashed", color = "gray") + 
                         theme_bw() +
                         theme(panel.grid.minor = element_blank(), panel.grid.major = element_blank())

# groups order in p2 is same with p1; use legend.position = "none" to remove redundant legend
p2 <- t2$plot_diff(measure = "PCo1", add_sig = T) + coord_flip() + theme_bw() + #border() +
      #geom_boxplot(fill = col.river.season, color = "black", alpha = 0.6) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray") +
      theme(legend.position = "none", axis.title.x = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank())
p3 <- t2$plot_diff(measure = "PCo2", add_sig = T) + theme_bw() +
      #geom_boxplot(fill = col.river.season, color = "black", alpha = 0.6) + 
      theme(legend.position = "none", axis.title.y = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank())
# height of the upper figure and width of the right-hand figure are both 0.2-fold of the main figure
v2 <- p1 %>% aplot::insert_top(p2, height = 0.25) %>% aplot::insert_right(p3, width = 0.25)
v2

## Mantel statistic based on Pearson's product-moment correlation 
bac.mco.fld.wunifrac <- bac.mco.fld$beta_diversity$wei_unifrac
bac.mco.btc.wunifrac <- bac.mco.btc$beta_diversity$wei_unifrac
bac.mco.fld.wunifrac <- vegan::decostand(bac.mco.fld.wunifrac, method = "hellinger")
bac.mco.btc.wunifrac <- vegan::decostand(bac.mco.btc.wunifrac, method = "hellinger")
vegan::mantel(bac.mco.fld.wunifrac, bac.mco.btc.wunifrac, method = "pearson", permutations = 999)

## Mantel statistic based on Pearson's product-moment correlation 
euk.mco.fld.wunifrac <- euk.mco.btc$beta_diversity$wei_unifrac
euk.mco.btc.wunifrac <- euk.mco.btc$beta_diversity$wei_unifrac
euk.mco.fld.wunifrac <- vegan::decostand(euk.mco.fld.wunifrac, method = "hellinger")
euk.mco.btc.wunifrac <- vegan::decostand(euk.mco.btc.wunifrac, method = "hellinger")
vegan::mantel(euk.mco.fld.wunifrac, euk.mco.btc.wunifrac, method = "pearson", permutations = 999)
```

### Assessment of biodegradation profile

```{r, Pie chart}
biodeg.rates <- readxl::read_excel("../MS-Spatiotemporal-BIE_Supplementary-Tables.xlsx", sheet = "S2 kTCC,pH7")
biodeg.rates <- data.frame(biodeg.rates, row.names = 1)

compound.class <- biodeg.rates[c(98:104)]

# Load required libraries
library("ggplot2"); library("dplyr")

# Prepare data: calculate percent abundance by Superclass
donut_data <- compound.class %>%
  count(Superclass) %>%
  mutate(percent = n / sum(n) * 100,
         label = paste0(round(percent, 1), "%"))

# Create donut chart with legend
ggplot(donut_data, aes(x = 2, y = percent, fill = Superclass)) +
  geom_bar(stat = "identity", width = 1, color = "black") +
  coord_polar(theta = "y") +
  xlim(0.5, 2.5) +
  theme_void() + scale_fill_manual(values = c("#8175AAFF","#31A1B3FF","#CCB22BFF","#94D0C0FF","#959C9EFF","#027B8EFF","#FFD700FF","#F3AE6DFF")) +
  #geom_text(aes(x = 2, label = label), position = position_stack(vjust = 0.5), size = 5) +
  theme(legend.title = element_text(size = 10, face = "bold"), legend.text = element_text(size = 9),
        legend.position = "right")
```

```{r, PCoA plot}
dataset <- clone(bdg.mco.fld)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# measure parameter must be one of names(dataset$beta_diversity)
t1 <- trans_beta$new(dataset = dataset, group = "River", measure = "jaccard")
# PCoA, PCA and NMDS are available
t1$cal_ordination(method = "PCoA"); class(t1$res_ordination) # t1$res_ordination is the ordination result list
# plot the PCoA result with confidence ellipse
ba <- t1$plot_ordination(plot_color = "Season", 
                         plot_shape = "River.Reach", 
                         plot_type = c("point"), 
                         ellipse_chull_alpha = 0,
                         #add_sample_label = "Reach",
                         color_values = RColorBrewer::brewer.pal(8, "Dark2")) + theme_bw() +
                         geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") +
                         geom_vline(xintercept = 0, linetype = "dashed", color = "gray95") +
                         scale_shape_manual(values = c(17, 2, 19, 1)) +
                         theme_bw() + 
                         theme(panel.grid = element_blank(), legend.position = "right", 
                               axis.text.y = element_text(angle = 90, hjust = 0.5), legend.title = element_text(face = "bold"),
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) +
                         ggalt::geom_encircle(aes(s_shape = 1, expand = 0, 
                                                  group = Season, fill = Season, colour = Season), alpha = 0.5) +
                         scale_fill_manual(values = RColorBrewer::brewer.pal(8, "Dark2"))
ba

tmp <- t1$res_ordination$scores
# differential test with trans_env class
t2 <- trans_env$new(dataset = dataset, add_data = tmp[, 1:2])
# 'KW_dunn' for non-parametric test
t2$cal_diff(group = "Season", method = "anova")

bb <- t2$plot_diff(measure = "PCo2", add_sig = T) + theme_bw() +
      geom_boxplot(fill = c("#1B9E77", "#D95F02", "#7570B3", "#E7298A"),
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none", 
            axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
            axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1))

# height of the upper figure and width of the right-hand figure are both 0.2-fold of the main figure
b1 <- ba %>% aplot::insert_right(bb, width = 0.25)
b1
```

```{r, Superclass biodegradation rate heatmap}
library("ComplexHeatmap"); library("microViz"); library("paletteer") 

dataset <- clone(bdg.mco.fld)

dataset$tax_table$Superclass <- gsub("s__", "", dataset$tax_table$Superclass)
psq <- file2meco::meco2phyloseq(dataset)

h1 <- psq %>% 
      ps_arrange(River.Reach.Season) %>% 
      tax_transform("compositional", rank = "Superclass") %>% 
      comp_heatmap(grid_col = NA, 
                   #colors = heat_palette(palette = "Cividis", rev = TRUE),
                   tax_anno = taxAnnotation(Prev. = anno_tax_prev(bar_width = 0.6, border = T), 
                                            Log10. = anno_tax_box(trans = "log10", zero_replace = "halfmin")),
                   #taxa_side = "bottom",
                   heatmap_legend_param = list(#at = 0:5 / 5, 
                                               direction = "horizontal", 
                                               title_position = "leftcenter", legend_width = grid::unit(4, "cm"),
                                               grid_height = grid::unit(5, "mm")),
                   sample_anno = sampleAnnotation(Season = anno_sample("Season"), 
                                                  col = list(Season = c("Winter" = "#7570B3", "Spring" = "#E7298A",
                                                                        "Summer" ="#1B9E77", "Autumn" = "#D95F02")),
                                                  border = TRUE,
                                                  River.Reach = anno_sample_cat(var = "River.Reach", 
                                                  box_col = NA, border_col = "black", legend_title = "River.Reach",)))

h1 %>% ComplexHeatmap::draw(annotation_legend_list = attr(h1, "AnnoLegends"), heatmap_legend_side = "bottom")
```

```{r, Class biodegradation rate heatmap}
library("ComplexHeatmap"); library("microViz"); library("paletteer")

dataset <- clone(bdg.mco.fld)

dataset$tax_table$Class <- gsub("c__", "", dataset$tax_table$Class)
psq <- file2meco::meco2phyloseq(dataset)

h1 <- psq %>% 
      ps_arrange(River.Reach.Season) %>% 
      tax_transform("compositional", rank = "Class") %>% 
      comp_heatmap(grid_col = NA, 
                   #colors = heat_palette(palette = "Cividis", rev = TRUE),
                   tax_anno = taxAnnotation(Prev. = anno_tax_prev(bar_width = 0.6, border = T), 
                                            Log10. = anno_tax_box(trans = "log10", zero_replace = "halfmin")),
                   #taxa_side = "bottom",
                   sample_anno = sampleAnnotation(Season = anno_sample("Season"), 
                                                  col = list(Season = c("Winter" = "#7570B3", "Spring" = "#E7298A", 
                                                                        "Summer" ="#1B9E77", "Autumn" = "#D95F02")),
                                                  border = TRUE,
                                                  River.Reach = anno_sample_cat(var = "River.Reach", 
                                                  box_col = NA, border_col = "black", legend_title = "River.Reach")))

h1 %>% ComplexHeatmap::draw(annotation_legend_list = attr(h1, "AnnoLegends"))
```

```{r, Parent Level 1 biodegradation rate heatmap}
library("ComplexHeatmap"); library("microViz"); library("paletteer")

dataset <- clone(bdg.mco.fld)

dataset$tax_table$Parent.Level.1 <- gsub("p__", "", dataset$tax_table$Parent.Level.1)
psq <- file2meco::meco2phyloseq(dataset)

h1 <- psq %>% 
      ps_arrange(River.Reach.Season) %>% 
      tax_transform("compositional", rank = "Parent.Level.1") %>% 
      comp_heatmap(grid_col = NA, 
                   #colors = heat_palette(palette = "Cividis", rev = TRUE),
                   tax_anno = taxAnnotation(Prev. = anno_tax_prev(bar_width = 0.6, border = T), 
                                            Log10. = anno_tax_box(trans = "log10", zero_replace = "halfmin")),
                   #taxa_side = "bottom",
                   sample_anno = sampleAnnotation(Season = anno_sample("Season"), 
                                                  col = list(Season = c("Winter" = "#7570B3", "Spring" = "#E7298A",
                                                                        "Summer" ="#1B9E77", "Autumn" = "#D95F02")),
                                                  border = TRUE,
                                                  River.Reach = anno_sample_cat(var = "River.Reach", 
                                                  box_col = NA, border_col = "black", legend_title = "River.Reach")))

h1 %>% ComplexHeatmap::draw(annotation_legend_list = attr(h1, "AnnoLegends"))
```

```{r, Differential analysis to identify seasoanal k rate}
dataset <- clone(bdg.mco.fld)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# use Genus level for parameter taxa_level, if you want to use all taxa, change to "all"
# nresam = 1 and boots = 1 represent no bootstrapping and use all samples directly
t1 <- trans_diff$new(dataset = dataset, method = "rf", group = "River.Reach.Season", taxa_level = "ASV")
# plot the MeanDecreaseGini bar
# group_order is designed to sort the groups
g1 <- t1$plot_diff_bar(use_number = 1:50)
# plot the abundance using same taxa in g1
g2 <- t1$plot_diff_abund(select_taxa = t1$plot_diff_bar_taxa, plot_type = "barerrorbar", 
                         add_sig = FALSE, errorbar_addpoint = FALSE, errorbar_color_black = TRUE)
# now the y axis in g1 and g2 is same, so we can merge them
# remove g1 legend; remove g2 y axis text and ticks
g1 <- g1 + theme(legend.position = "none")
g2 <- g2 + theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), panel.border = element_blank())
p <- g1 %>% aplot::insert_right(g2)
p

openxlsx::write.xlsx(t1$res_diff, file = "src/bdg.rf.xlsx", rowNames = FALSE)
```

```{r, plot rf result}
biodeg.rates <- readxl::read_excel("../MS-Spatiotemporal-BIE_Supplementary-Tables.xlsx", sheet = "S2 kTCC,pH7")
biodeg.rates <- data.frame(biodeg.rates, row.names = 1)

compound.class <- biodeg.rates

df <- data.frame(Indicator = compound.class$Indicator, Superclass = compound.class$Superclass)

# Count occurrences by Indicator and Superclass
df_counts <- as.data.frame(table(df$Indicator, df$Superclass))
colnames(df_counts) <- c("Indicator", "Superclass", "Count")

# Plot using ggplot2
ggplot(df_counts, aes(x = reorder(Indicator, -Count), y = Count, fill = Superclass)) +
  geom_bar(stat = "identity") + theme_classic() +
  geom_hline(yintercept = 6, linetype = "dashed", color = "gray85") +
  labs(y = "No. of Compounds") +
  theme(legend.position ="none", plot.title = element_text(size = 12, face = "bold"),
        axis.text.x = element_text(angle = 45, hjust = 1, size = 8), 
        panel.grid = element_blank(),
        axis.title.x = element_blank()) +
  scale_fill_manual(values = c("#8175AAFF","#31A1B3FF","#CCB22BFF","#94D0C0FF",
                               "#959C9EFF","#027B8EFF","#FFD700FF","#F3AE6DFF"))
``` 

```{r, plot dbRDA R2 results}
# Category <- Hydrological parameters = c(pH, DO, EC, Water temp., Flow rate, TSS, TOC), Nutrients = c(NH4-N, NO2-NO3-N, Total N, Total P), Major ions = c(Ca,Mg,Cl), Metals = c(Cr,Cu,Ni,Pb,Zn)

data <- data.frame(factor = c("pH", "DO", "EC", "Water temp.","Flow rate", "Ca", "Cl", "Cr", "Cu",
                              "Mg", "NH4-N", "NO2-NO3-N", "Ni","Pb", "TSS", "TOC", "Total N", "Total P", "Zn"),
                   r2 = c(0.7302,0.2553,0.4193,0.0801,0.1301,0.0663,0.41,0.2156,0.0374,0.07,
                          0.4043,0.294,0.163,0.3764,0.1141,0.0542,0.3495,0.2406,0.1441),
                   significance = c("***","**","***","","*","","***","**","","","***","**","*","***","*","","***","**","*"))

# Reorder factors by r2 value
data$factor <- factor(data$factor, levels = data$factor[order(data$r2, decreasing = TRUE)])

# Plot
f1 <- ggplot(data, aes(x = factor, y = r2)) +
      geom_bar(stat = "identity", fill = "gray75") + 
      geom_text(aes(label = significance), vjust = -0.5, size = 5) +
      labs(x = "", y = expression("Variance explained, " ~ R^2)) + theme_classic() +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 10), 
            panel.grid = element_blank(), axis.title.x = element_blank()) + ylim(0, 0.8) #+ coord_flip()

f1
```

```{r, eval = FALSE}
# Load required library
library(ggplot2)

# Create data frame
data <- data.frame(factor = c("pH", "DO", "EC", "Water temp.","Flow rate", "TSS", "TOC",
                              "NH4-N", "NO2-NO3-N", "Total N", "Total P",
                              "Ca", "Mg", "Cl",
                              "Cr", "Cu", "Ni", "Pb", "Zn"),
                   r2 = c(0.7302,0.2553,0.4193,0.0801,0.1301,0.1141,0.0542,0.4043,0.294,0.3495,
                          0.2406,0.0663,0.07,0.41,0.2156,0.0374,0.163,0.3764,0.1441),
                   significance = c("***","**","***","","*","*","","***","**","***","**","","", "***","**","","*","***","*"))

# Define categories
data$category <- c(rep("Hydrological parameters", 7), rep("Nutrients", 4), rep("Major ions", 3), rep("Metals", 5))

# Set factor levels to arrange by category
data$factor <- factor(data$factor, levels = data$factor)

# Plot
f1 <- ggplot(data, aes(x = factor, y = r2, fill = category)) + geom_bar(stat = "identity") + 
      geom_text(aes(label = significance), vjust = -0.5, size = 5) +
      geom_bar(stat = "identity", fill = "gray75") +
      labs(x = "", y = expression("Variance explained, " ~ R^2), fill = "Category") + theme_classic() +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 10), legend.position = "none",
            panel.grid = element_blank(), axis.title.x = element_blank()) + ylim(0, 0.8)

# Display plot
print(f1)
```

## ########################################################## ##
##                                                            ##
##         DOWNSTREAM ANALYSES - MICROBIAL COMMUNITY          ##
##                                                            ##
## ########################################################## ##

### Visualize community composition

```{r, Prokaryotes and eukaryotes relative abundance plot}
dataset <- clone(bac.mco.fld)
dataset$sample_table$River.Reach.Season %<>% factor(., levels = c("Knivstaån Up Winter", "Knivstaån Up Spring", 
                                                                  "Knivstaån Up Summer", "Knivstaån Up Autumn", 
                                                                  "Knivstaån Dn Winter", "Knivstaån Dn Spring", 
                                                                  "Knivstaån Dn Summer", "Knivstaån Dn Autumn", 
                                                                  "Vitsån Up Winter", "Vitsån Up Spring", "Vitsån Up Summer", "Vitsån Up Autumn", 
                                                                  "Vitsån Dn Winter", "Vitsån Dn Spring", "Vitsån Dn Summer", "Vitsån Dn Autumn"))

t1 <- trans_abund$new(dataset = dataset, taxrank = "Phylum", ntaxa = 10, groupmean = "River.Reach.Season")
g1 <- t1$plot_bar(coord_flip = FALSE, clustering_plot = FALSE, color_values = paletteer::paletteer_d("ggthemes::Miller_Stone")) +
      theme(axis.text.x = element_blank(), axis.ticks.x =  element_blank(), panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5),
            axis.title.y = element_text(size = 12), axis.text.y = element_text(size = 10), legend.title = element_text(size = 10, face = "bold"), legend.text = element_text(size = 9),
            legend.position = "right") #+ ggtitle("Prokaryotes")
g1

dataset <- clone(euk.mco.fld)
dataset$sample_table$River.Reach.Season %<>% factor(., levels = c("Knivstaån Up Winter", "Knivstaån Up Spring", "Knivstaån Up Summer", "Knivstaån Up Autumn", 
                                                                  "Knivstaån Dn Winter", "Knivstaån Dn Spring", "Knivstaån Dn Summer", "Knivstaån Dn Autumn", 
                                                                  "Vitsån Up Winter", "Vitsån Up Spring", "Vitsån Up Summer", "Vitsån Up Autumn", 
                                                                  "Vitsån Dn Winter", "Vitsån Dn Spring", "Vitsån Dn Summer", "Vitsån Dn Autumn"))

t1 <- trans_abund$new(dataset = dataset, taxrank = "Division", ntaxa = 10, groupmean = "River.Reach.Season")
g2 <- t1$plot_bar(coord_flip = FALSE, clustering_plot = FALSE, color_values = paletteer::paletteer_d("rcartocolor::Antique")) +
      theme(axis.text.x = element_blank(), axis.ticks.x =  element_blank(), panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5),
            axis.title.y = element_text(size = 12), axis.text.y = element_text(size = 10), legend.title = element_text(size = 10, face = "bold"), legend.text = element_text(size = 9),
            legend.position = "right") #+ ggtitle("Eukaryotes")
g2

g1 / g2 + patchwork::plot_layout(widths = unit(c(10,10), c("null","null")))
```

```{r, Different eukaryotes relative abundance plot}
dataset <- clone(alg.mco.fld)
dataset$sample_table$River.Reach.Season %<>% factor(., levels = c("Knivstaån Up Winter", "Knivstaån Up Spring", "Knivstaån Up Summer", "Knivstaån Up Autumn", 
                                                                  "Knivstaån Dn Winter", "Knivstaån Dn Spring", "Knivstaån Dn Summer", "Knivstaån Dn Autumn", 
                                                                  "Vitsån Up Winter", "Vitsån Up Spring", "Vitsån Up Summer", "Vitsån Up Autumn", 
                                                                  "Vitsån Dn Winter", "Vitsån Dn Spring", "Vitsån Dn Summer", "Vitsån Dn Autumn"))

t1 <- trans_abund$new(dataset = dataset, taxrank = "Class", ntaxa = 10, groupmean = "River.Reach.Season")
g3 <- t1$plot_bar(coord_flip = FALSE, clustering_plot = FALSE, color_values = paletteer::paletteer_d("ggsci::green_material")) +
      theme(axis.text.x = element_text(size = 8, hjust = 1, angle = 45), panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5),
            axis.title.y = element_text(size = 12), axis.text.y = element_text(size = 10), legend.title = element_text(size = 10, face = "bold"), legend.text = element_text(size = 9),
            legend.position = "right") + 
            ggtitle("Algae")
g3

dataset <- clone(fun.mco.fld)
dataset$sample_table$River.Reach.Season %<>% factor(., levels = c("Knivstaån Up Winter", "Knivstaån Up Spring", "Knivstaån Up Summer", "Knivstaån Up Autumn", 
                                                                  "Knivstaån Dn Winter", "Knivstaån Dn Spring", "Knivstaån Dn Summer", "Knivstaån Dn Autumn", 
                                                                  "Vitsån Up Winter", "Vitsån Up Spring", "Vitsån Up Summer", "Vitsån Up Autumn", 
                                                                  "Vitsån Dn Winter", "Vitsån Dn Spring", "Vitsån Dn Summer", "Vitsån Dn Autumn"))

t1 <- trans_abund$new(dataset = dataset, taxrank = "Class", ntaxa = 10, groupmean = "River.Reach.Season")
g4 <- t1$plot_bar(coord_flip = FALSE, clustering_plot = FALSE, color_values = paletteer::paletteer_d("ggsci::deep_purple_material")) +
      theme(axis.text.x = element_text(size = 8, hjust = 1, angle = 45), panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5),
            axis.title.y = element_text(size = 12), axis.text.y = element_text(size = 10), legend.title = element_text(size = 10, face = "bold"), legend.text = element_text(size = 9),
            legend.position = "right") + 
            ggtitle("Fungi")
g4

dataset <- clone(pro.mco.fld)
dataset$sample_table$River.Reach.Season %<>% factor(., levels = c("Knivstaån Up Winter", "Knivstaån Up Spring", "Knivstaån Up Summer", "Knivstaån Up Autumn", 
                                                                  "Knivstaån Dn Winter", "Knivstaån Dn Spring", "Knivstaån Dn Summer", "Knivstaån Dn Autumn", 
                                                                  "Vitsån Up Winter", "Vitsån Up Spring", "Vitsån Up Summer", "Vitsån Up Autumn", 
                                                                  "Vitsån Dn Winter", "Vitsån Dn Spring", "Vitsån Dn Summer", "Vitsån Dn Autumn"))

t1 <- trans_abund$new(dataset = dataset, taxrank = "Class", ntaxa = 10, groupmean = "River.Reach.Season")
g5 <- t1$plot_bar(coord_flip = FALSE, clustering_plot = FALSE, color_values = paletteer::paletteer_d("ggsci::yellow_material")) +
      theme(axis.text.x = element_text(size = 8, hjust = 1, angle = 45), panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5),
            axis.title.y = element_text(size = 12), axis.text.y = element_text(size = 10), legend.title = element_text(size = 10, face = "bold"), legend.text = element_text(size = 9),
            legend.position = "right") + 
            ggtitle("Protists")
g5

dataset <- clone(met.mco.fld)
dataset$sample_table$River.Reach.Season %<>% factor(., levels = c("Knivstaån Up Winter", "Knivstaån Up Spring", "Knivstaån Up Summer", "Knivstaån Up Autumn", 
                                                                  "Knivstaån Dn Winter", "Knivstaån Dn Spring", "Knivstaån Dn Summer", "Knivstaån Dn Autumn", 
                                                                  "Vitsån Up Winter", "Vitsån Up Spring", "Vitsån Up Summer", "Vitsån Up Autumn", 
                                                                  "Vitsån Dn Winter", "Vitsån Dn Spring", "Vitsån Dn Summer", "Vitsån Dn Autumn"))

t1 <- trans_abund$new(dataset = dataset, taxrank = "Class", ntaxa = 10, groupmean = "River.Reach.Season")
g6 <- t1$plot_bar(coord_flip = FALSE, clustering_plot = FALSE, color_values = paletteer::paletteer_d("ggsci::red_material")) +
      theme(axis.text.x = element_text(size = 8, hjust = 1, angle = 45), panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5),
            axis.title.y = element_text(size = 12), axis.text.y = element_text(size = 10), legend.title = element_text(size = 10, face = "bold"), legend.text = element_text(size = 9),
            legend.position = "right") + 
            ggtitle("Metazoan")
g6

( g3 + g4 ) / ( g5 + g6 ) + patchwork::plot_layout(widths = unit(c(10,10,10,10), c("null","null","null","null")))
```

### Alpha diversity analysis 

```{r, Prokaryotes alpha diversity, replace mco data to generate other groups}
dataset <- clone(bac.mco.fld)
dataset$sample_table$River.Reach.Season %<>% factor(., levels = c("Knivstaån Up Winter", "Knivstaån Up Spring", 
                                                                  "Knivstaån Up Summer", "Knivstaån Up Autumn", 
                                                                  "Knivstaån Dn Winter", "Knivstaån Dn Spring", 
                                                                  "Knivstaån Dn Summer", "Knivstaån Dn Autumn", 
                                                                  "Vitsån Up Winter", "Vitsån Up Spring", "Vitsån Up Summer", "Vitsån Up Autumn", 
                                                                  "Vitsån Dn Winter", "Vitsån Dn Spring", "Vitsån Dn Summer", "Vitsån Dn Autumn"))

col.rrs <- c("#7570B3", "#E7298A", "#1B9E77", "#D95F02", "#7570B3", "#E7298A", "#1B9E77", "#D95F02",
             "#7570B3", "#E7298A", "#1B9E77", "#D95F02", "#7570B3", "#E7298A", "#1B9E77", "#D95F02")

t1 <- trans_alpha$new(dataset = dataset, group = "River.Reach.Season"); t1$data_stat
t1$cal_diff(method = "anova"); t1$res_diff

# y_increase can adjust the distance from the letters to the highest point
aa <- t1$plot_alpha(measure = "Observed", add_sig_text_size = 4, order_x_mean = FALSE,
                    boxplot_add = "dotplot", 
                    color_values = rep("black", 16)) + #scale_color_grey() +
                    geom_boxplot(fill = col.rrs, color = "black", alpha = 0.85, outlier.shape = NA) + theme_bw() +
                    #geom_rect(data = shading, aes(xmin = min, xmax = max, ymin = -Inf, ymax = Inf, fill = factor(col), alpha = 0.1)) +
                    theme(legend.position = "none", axis.text.y = element_text(size = 8, angle = 90), 
                          panel.border = element_rect(colour = "black", fill = NA, linewidth = 1), panel.grid.major.x = element_blank(),
                          #axis.text.x = element_text(size = 8, angle = 45, hjust = 1)) + labs(y = "Observed")
                          axis.text.x = element_blank()) + labs(y = "Observed Richness")

ab <- t1$plot_alpha(measure = "Shannon", add_sig_text_size = 4, order_x_mean = FALSE,
                    boxplot_add = "dotplot", 
                    color_values = rep("black", 16)) + #scale_color_grey() +
                    geom_boxplot(fill = col.rrs, color = "black", alpha = 0.85, outlier.shape = NA) + theme_bw() +
                    #geom_rect(data = shading, aes(xmin = min, xmax = max, ymin = -Inf, ymax = Inf, fill = factor(col), alpha = 0.1)) +
                    theme(legend.position = "none", axis.text.y = element_text(size = 8, angle = 90), 
                          panel.border = element_rect(colour = "black", fill = NA, linewidth = 1), panel.grid.major.x = element_blank(),
                          #axis.text.x = element_text(size = 8, angle = 45, hjust = 1)) + labs(y = "Observed")
                          axis.text.x = element_blank()) + labs(y = "Shannon Diversity")

ac <- t1$plot_alpha(measure = "Pielou", add_sig_text_size = 4, order_x_mean = FALSE,
                    boxplot_add = "dotplot", 
                    color_values = rep("black", 16)) + #scale_color_grey() +
                    geom_boxplot(fill = col.rrs, color = "black", alpha = 0.85, outlier.shape = NA) + theme_bw() +
                    #geom_rect(data = shading, aes(xmin = min, xmax = max, ymin = -Inf, ymax = Inf, fill = factor(col), alpha = 0.1)) +
                    theme(legend.position = "none", axis.text.y = element_text(size = 8, angle = 90), 
                          panel.border = element_rect(colour = "black", fill = NA, linewidth = 1), panel.grid.major.x = element_blank(),
                          #axis.text.x = element_text(size = 8, angle = 45, hjust = 1)) + labs(y = "Observed")
                          axis.text.x = element_blank()) + labs(y = "Pielou's Evenness")

ad <- t1$plot_alpha(measure = "PD", add_sig_text_size = 4, order_x_mean = FALSE,
                    boxplot_add = "dotplot", 
                    color_values = rep("black", 16)) + #scale_color_grey() +
                    geom_boxplot(fill = col.rrs, color = "black", alpha = 0.85, outlier.shape = NA) + theme_bw() +
                    theme(legend.position = "none", axis.text.y = element_text(size = 8, angle = 90), 
                          panel.border = element_rect(colour = "black", fill = NA, linewidth = 1), panel.grid.major.x = element_blank(),
                          #axis.text.x = element_text(size = 8, angle = 45, hjust = 1)) + labs(y = "Observed")
                          axis.text.x = element_blank()) + labs(y = "Faith's PD")

# Combine the plots in one row
a1 <- (aa / ab / ac / ad) + patchwork::plot_layout(guides = 'collect') + patchwork::plot_annotation(); a1
```

```{r, Prokaryotes alpha diversity with grouping, replace mco data to generate other groups}
dataset <- clone(met.mco.fld)
dataset$sample_table$River.Reach.Season %<>% factor(., levels = c("Winter", "Spring", "Summer", "Autumn"))

t1 <- trans_alpha$new(dataset = dataset, group = "Season", by_group = "River.Reach")
t1$cal_diff(method = "anova")

aa <- t1$plot_alpha(measure = "Observed", add_sig_text_size = 3, order_x_mean = FALSE, boxplot_add = "dotplot", 
                    color_values = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3")) + theme_light() +
                    stat_summary(fun = median, geom = "line", aes(group = 1), col = "red") + 
                    theme(legend.position = "right", 
                          axis.text.x = element_text(size = 8, angle = 45, hjust = 1),
                          axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
                          panel.border = element_rect(colour = "black", fill = NA, linewidth = 1), panel.grid.major.x = element_blank())+ 
                          labs(y = "Observed") #+ coord_flip()
aa

ab <- t1$plot_alpha(measure = "Simpson", add_sig_text_size = 3, order_x_mean = FALSE, boxplot_add = "dotplot", 
                    color_values = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3")) + theme_light() +
                    stat_summary(fun = median, geom = "line", aes(group = 1), col = "red") + 
                    theme(legend.position = "none", 
                          axis.text.x = element_text(size = 8, angle = 45, hjust = 1),
                          axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
                          panel.border = element_rect(colour = "black", fill = NA, linewidth = 1), panel.grid.major.x = element_blank())+ 
                          labs(y = "Simpson") #+ coord_flip()
ab

ac <- t1$plot_alpha(measure = "Pielou", add_sig_text_size = 3, order_x_mean = FALSE, boxplot_add = "dotplot", 
                    color_values = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3")) + theme_light() +
                    stat_summary(fun = median, geom = "line", aes(group = 1), col = "red") + 
                    theme(legend.position = "none", 
                          axis.text.x = element_text(size = 8, angle = 45, hjust = 1),
                          axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
                          panel.border = element_rect(colour = "black", fill = NA, linewidth = 1), panel.grid.major.x = element_blank())+ 
                          labs(y = "Pielou") #+ coord_flip()
ac

ad <- t1$plot_alpha(measure = "PD", add_sig_text_size = 3, order_x_mean = FALSE, boxplot_add = "dotplot", 
                    color_values = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3")) + theme_light() +
                    stat_summary(fun = median, geom = "line", aes(group = 1), col = "red") + 
                    theme(legend.position = "none", 
                          axis.text.x = element_text(size = 8, angle = 45, hjust = 1),
                          axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
                          panel.border = element_rect(colour = "black", fill = NA, linewidth = 1), panel.grid.major.x = element_blank())+ 
                          labs(y = "Faith's PD") #+ coord_flip()
ad

a2 <- ( aa | ab | ac | ad ) + patchwork::plot_annotation(title = "Metazoan")
a2
```

```{r, Linear regression analysis between alpha diversity of trophic groups}
# Create a data frame for plotting
data <- data.frame(bac = log(bac.mco.fld$alpha_diversity$Observed), euk = log(euk.mco.fld$alpha_diversity$Observed), river = bac.mco.fld$sample_table$River, season = bac.mco.fld$sample_table$Season)
# Plot the data and add a regression line with R² and p-value annotations
ca <- ggplot(data, aes(x = bac, y = euk, shape = river, color = season)) +
      geom_point(color = "gray", alpha = 0.5) + 
      scale_shape_manual(values = c(17, 19)) +
      geom_smooth(method = "lm", col = "gray", fill = "gray") + # Scatter plot
      labs(title = "Observed Richness", x = "Prokaryotes", y = "Eukaryotes") + ylim(6.3,7.4) + 
      theme_classic2() + theme(legend.position = "none", axis.title.x = element_text(size = 12), 
                               axis.title.y = element_text(size = 12, angle = 90, vjust = 0.5, hjust = 0.5), 
                               axis.text.x = element_text(size = 10), 
                               axis.text.y = element_text(size = 10, angle = 90, vjust = 0.5, hjust = 0.5), 
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
                               plot.title = element_text(size = 13)) # Rotate y-axis text
ca

# Create a data frame for plotting
data <- data.frame(bac = log(bac.mco.fld$alpha_diversity$Shannon), euk = log(euk.mco.fld$alpha_diversity$Shannon), river = bac.mco.fld$sample_table$River, season = bac.mco.fld$sample_table$Season)
# Plot the data and add a regression line with R² and p-value annotations
cb <- ggplot(data, aes(x = bac, y = euk, shape = river, color = season)) +
      geom_point(color = "gray", alpha = 0.5) + 
      scale_shape_manual(values = c(17, 19)) +
      geom_smooth(method = "lm", col = "gray", fill = "gray") + # Scatter plot
      labs(title = "Shannon Diversity", x = "Prokaryotes", y = "Eukaryotes") + ylim(1.4,1.75) + 
      theme_classic2() + theme(legend.position = "none", axis.title.x = element_text(size = 12), 
                               axis.title.y = element_text(size = 12, angle = 90, vjust = 0.5, hjust = 0.5), 
                               axis.text.x = element_text(size = 10), 
                               axis.text.y = element_text(size = 10, angle = 90, vjust = 0.5, hjust = 0.5), 
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
                               plot.title = element_text(size = 13)) # Rotate y-axis text
cb

# Create a data frame for plotting
data <- data.frame(bac = log(bac.mco.fld$alpha_diversity$Pielou), euk = log(euk.mco.fld$alpha_diversity$Pielou), river = bac.mco.fld$sample_table$River, season = bac.mco.fld$sample_table$Season)
# Plot the data and add a regression line with R² and p-value annotations
cc <- ggplot(data, aes(x = bac, y = euk, shape = river, color = season)) +
      geom_point(color = "gray", alpha = 0.5) + 
      scale_shape_manual(values = c(17, 19)) +
      geom_smooth(method = "lm", col = "gray", fill = "gray") + # Scatter plot
      labs(title = "Pielou's Evenness", x = "Prokaryotes", y = "Eukaryotes") + ylim(-0.50,-0.25) + 
      theme_classic2() + theme(legend.position = "none", axis.title.x = element_text(size = 12), 
                               axis.title.y = element_text(size = 12, angle = 90, vjust = 0.5, hjust = 0.5), 
                               axis.text.x = element_text(size = 10), 
                               axis.text.y = element_text(size = 10, angle = 90, vjust = 0.5, hjust = 0.5), 
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
                               plot.title = element_text(size = 13)) # Rotate y-axis text
cc

# Create a data frame for plotting
data <- data.frame(bac = log(bac.mco.fld$alpha_diversity$PD), euk = log(euk.mco.fld$alpha_diversity$PD), river = bac.mco.fld$sample_table$River, season = bac.mco.fld$sample_table$Season)
# Plot the data and add a regression line with R² and p-value annotations
cd <- ggplot(data, aes(x = bac, y = euk, shape = river, color = season)) +
      geom_point(color = "gray", alpha = 0.5) + 
      scale_shape_manual(values = c(17, 19)) +
      geom_smooth(method = "lm", col = "gray", fill = "gray") + # Scatter plot
      labs(title = "Faith's PD", x = "Prokaryotes", y = "Eukaryotes") + ylim(5,5.8) + 
      theme_classic2() + theme(legend.position = "none", axis.title.x = element_text(size = 12), 
                               axis.title.y = element_text(size = 12, angle = 90, vjust = 0.5, hjust = 0.5), 
                               axis.text.x = element_text(size = 10), 
                               axis.text.y = element_text(size = 10, angle = 90, vjust = 0.5, hjust = 0.5), 
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
                               plot.title = element_text(size = 13)) # Rotate y-axis text
cd

c2 <- cb + cd; c2
c3 <- (ca | cb | cc | cd) + patchwork::plot_layout(axis_titles = "collect"); c3

# Perform correlation test
correlation <- cor.test(bac.mco.fld$alpha_diversity$Observed, euk.mco.fld$alpha_diversity$Observed)
p_value <- signif(correlation$p.value, 4); r_value <- correlation$estimate; r_squared <- r_value^2 # Calculate R-squared
p_value; r_value; r_squared

# Perform correlation test
correlation <- cor.test(bac.mco.fld$alpha_diversity$Shannon, euk.mco.fld$alpha_diversity$Shannon)
p_value <- signif(correlation$p.value, 4); r_value <- correlation$estimate; r_squared <- r_value^2 # Calculate R-squared
p_value; r_value; r_squared

# Perform correlation test
correlation <- cor.test(bac.mco.fld$alpha_diversity$Pielou, euk.mco.fld$alpha_diversity$Pielou)
p_value <- signif(correlation$p.value, 4); r_value <- correlation$estimate; r_squared <- r_value^2 # Calculate R-squared
p_value; r_value; r_squared

# Perform correlation test
correlation <- cor.test(bac.mco.fld$alpha_diversity$PD, euk.mco.fld$alpha_diversity$PD)
p_value <- signif(correlation$p.value, 4); r_value <- correlation$estimate; r_squared <- r_value^2 # Calculate R-squared
p_value; r_value; r_squared

``` 

### Beta diversity analysis

```{r, Prokaryotes PCoA Plot}
dataset <- clone(bac.mco.fld)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# measure parameter must be one of names(dataset$beta_diversity)
t1 <- trans_beta$new(dataset = dataset, group = "River.Reach.Season", measure = "wei_unifrac")
# PCoA, PCA and NMDS are available
t1$cal_ordination(method = "PCoA"); class(t1$res_ordination) # t1$res_ordination is the ordination result list
# plot the PCoA result with confidence ellipse
ba <- t1$plot_ordination(plot_color = "Season", 
                         plot_shape = "River.Reach", 
                         plot_type = c("point"), 
                         ellipse_chull_alpha = 0,
                         #add_sample_label = "Reach",
                         color_values = RColorBrewer::brewer.pal(8, "Dark2")) + theme_bw() +
                         geom_hline(yintercept = 0, linetype = "dashed", color = "gray85") +
                         geom_vline(xintercept = 0, linetype = "dashed", color = "gray85") +
                         scale_shape_manual(values = c(17, 2, 19, 1)) +
                         theme_bw() + 
                         theme(panel.grid = element_blank(), legend.position = "right", #axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
                               axis.text.y = element_text(angle = 90, hjust = 0.5), legend.title = element_text(face = "bold"),
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) +
                         ggalt::geom_encircle(aes(s_shape = 1, expand = 0, group = River.Reach.Season, 
                                                  fill = Season, colour = Season), alpha = 0.5) +
                         scale_fill_manual(values = RColorBrewer::brewer.pal(8, "Dark2")) + 
                         ggtitle("Prokaryotes")
ba

tmp <- t1$res_ordination$scores
# differential test with trans_env class
t2 <- trans_env$new(dataset = dataset, add_data = tmp[, 1:2])
# 'KW_dunn' for non-parametric test
t2$cal_diff(group = "River.Season", method = "anova")

bb <- t2$plot_diff(measure = "PCo2", add_sig = T, color = rep("white", 8)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), 
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none", 
            axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
            axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1))
bb

# height of the upper figure and width of the right-hand figure are both 0.2-fold of the main figure
b1 <- ba #%>% aplot::insert_right(bb, width = 0.5)
b1

# manova for specified group set: such as "Region"
t1$cal_manova(manova_set = "River*Reach*Season"); t1$res_manova

t1$cal_anosim(group = "River.Reach.Season"); t1$res_anosim
#t1$cal_betadisper(); t1$res_betadisper

t1$cal_group_distance(within_group = TRUE) # return t1$res_group_distance
t1$cal_group_distance_diff(method = "anova") # perform Wilcoxon Rank Sum and Signed Rank Tests
c1 <- t1$plot_group_distance(color = rep("white", 16)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3",
                            "#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), 
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none",
            axis.title.y = element_text(size = 12),
            axis.text.x = element_text(size = 8, angle = 45, hjust = 1), 
            axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) #+ ylim(0.5, 0.6) + 
      ggtitle("Prokaryotes")
```

```{r, Predicted Functions PCoA Plot}
dataset <- clone(keg.mco.fld)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# measure parameter must be one of names(dataset$beta_diversity)
t1 <- trans_beta$new(dataset = dataset, group = "River.Reach.Season", measure = "jaccard")
# PCoA, PCA and NMDS are available
t1$cal_ordination(method = "PCoA"); class(t1$res_ordination) # t1$res_ordination is the ordination result list
# plot the PCoA result with confidence ellipse
ba <- t1$plot_ordination(plot_color = "Season", 
                         plot_shape = "River.Reach", 
                         plot_type = c("point"), 
                         ellipse_chull_alpha = 0,
                         #add_sample_label = "Reach",
                         color_values = RColorBrewer::brewer.pal(8, "Dark2")) + theme_bw() +
                         geom_hline(yintercept = 0, linetype = "dashed", color = "gray85") +
                         geom_vline(xintercept = 0, linetype = "dashed", color = "gray85") +
                         scale_shape_manual(values = c(17, 2, 19, 1)) +
                         theme_bw() + 
                         theme(panel.grid = element_blank(), legend.position = "none", #axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
                               axis.text.y = element_text(angle = 90, hjust = 0.5), legend.title = element_text(face = "bold"),
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) +
                         ggalt::geom_encircle(aes(s_shape = 1, expand = 0, group = River.Reach.Season, fill = Season, colour = Season), alpha = 0.5) +
                         scale_fill_manual(values = RColorBrewer::brewer.pal(8, "Dark2")) + 
                         ggtitle("Predicted Functions")
ba

tmp <- t1$res_ordination$scores
# differential test with trans_env class
t2 <- trans_env$new(dataset = dataset, add_data = tmp[, 1:2])
# 'KW_dunn' for non-parametric test
t2$cal_diff(group = "River.Season", method = "anova")

bb <- t2$plot_diff(measure = "PCo2", add_sig = T, color = rep("white", 8)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), 
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none", 
            axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
            axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1))
bb

# height of the upper figure and width of the right-hand figure are both 0.2-fold of the main figure
b2 <- ba #%>% aplot::insert_right(bb, width = 0.5)
b2

# manova for specified group set: such as "Region"
t1$cal_manova(manova_set = "River*Reach*Season"); t1$res_manova

t1$cal_anosim(group = "River.Reach.Season"); t1$res_anosim
#t1$cal_betadisper(); t1$res_betadisper

t1$cal_group_distance(within_group = TRUE) # return t1$res_group_distance
t1$cal_group_distance_diff(method = "anova") # perform Wilcoxon Rank Sum and Signed Rank Tests
c2 <- t1$plot_group_distance(color = rep("white", 16)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3",
                            "#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), 
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none",
            axis.title.y = element_text(size = 12),
            axis.text.x = element_text(size = 8, angle = 45, hjust = 1), 
            axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) #+ ylim(0.5, 0.6) + 
      ggtitle("Predicted Functions")
```

```{r, Algae PCoA Plot}
dataset <- clone(alg.mco.fld)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# measure parameter must be one of names(dataset$beta_diversity)
t1 <- trans_beta$new(dataset = dataset, group = "River.Reach.Season", measure = "wei_unifrac")
# PCoA, PCA and NMDS are available
t1$cal_ordination(method = "PCoA"); class(t1$res_ordination) # t1$res_ordination is the ordination result list
# plot the PCoA result with confidence ellipse
ba <- t1$plot_ordination(plot_color = "Season", 
                         plot_shape = "River.Reach", 
                         plot_type = c("point"), 
                         ellipse_chull_alpha = 0,
                         #add_sample_label = "Reach",
                         color_values = RColorBrewer::brewer.pal(8, "Dark2")) + theme_bw() +
                         geom_hline(yintercept = 0, linetype = "dashed", color = "gray85") +
                         geom_vline(xintercept = 0, linetype = "dashed", color = "gray85") +
                         scale_shape_manual(values = c(17, 2, 19, 1)) +
                         theme_bw() + 
                         theme(panel.grid = element_blank(), legend.position = "none", #axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
                               axis.text.y = element_text(angle = 90, hjust = 0.5), legend.title = element_text(face = "bold"),
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) +
                         ggalt::geom_encircle(aes(s_shape = 1, expand = 0, group = River.Reach.Season, fill = Season, colour = Season), alpha = 0.5) +
                         scale_fill_manual(values = RColorBrewer::brewer.pal(8, "Dark2")) + 
                         ggtitle("Algae")
ba

tmp <- t1$res_ordination$scores
# differential test with trans_env class
t2 <- trans_env$new(dataset = dataset, add_data = tmp[, 1:2])
# 'KW_dunn' for non-parametric test
t2$cal_diff(group = "River.Season", method = "anova")

bb <- t2$plot_diff(measure = "PCo2", add_sig = T, color = rep("white",8)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), 
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none", 
            axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
            axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1))
bb

# height of the upper figure and width of the right-hand figure are both 0.2-fold of the main figure
b3 <- ba #%>% aplot::insert_right(bb, width = 0.5)
b3

# manova for specified group set: such as "Region"
t1$cal_manova(manova_set = "River*Reach*Season"); t1$res_manova

t1$cal_anosim(group = "River.Reach.Season"); t1$res_anosim
#t1$cal_betadisper(); t1$res_betadisper

t1$cal_group_distance(within_group = TRUE) # return t1$res_group_distance
t1$cal_group_distance_diff(method = "anova") # perform Wilcoxon Rank Sum and Signed Rank Tests
c3 <- t1$plot_group_distance(color = rep("white", 16)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3",
                            "#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), 
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none",
            axis.title.y = element_text(size = 12),
            axis.text.x = element_text(size = 8, angle = 45, hjust = 1), 
            axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) + #ylim(0.025, 0.06) + 
      ggtitle("Algae")
```

```{r, Fungi PCoA Plot}
dataset <- clone(fun.mco.fld)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# measure parameter must be one of names(dataset$beta_diversity)
t1 <- trans_beta$new(dataset = dataset, group = "River.Reach.Season", measure = "wei_unifrac")
# PCoA, PCA and NMDS are available
t1$cal_ordination(method = "PCoA"); class(t1$res_ordination) # t1$res_ordination is the ordination result list
# plot the PCoA result with confidence ellipse
ba <- t1$plot_ordination(plot_color = "Season", 
                         plot_shape = "River.Reach", 
                         plot_type = c("point"), 
                         ellipse_chull_alpha = 0,
                         #add_sample_label = "Reach",
                         color_values = RColorBrewer::brewer.pal(8, "Dark2")) + theme_bw() +
                         geom_hline(yintercept = 0, linetype = "dashed", color = "gray85") +
                         geom_vline(xintercept = 0, linetype = "dashed", color = "gray85") +
                         scale_shape_manual(values = c(17, 2, 19, 1)) +
                         theme_bw() + 
                         theme(panel.grid = element_blank(), legend.position = "none", #axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
                               axis.text.y = element_text(angle = 90, hjust = 0.5), legend.title = element_text(face = "bold"),
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) +
                         ggalt::geom_encircle(aes(s_shape = 1, expand = 0, group = River.Reach.Season, fill = Season, colour = Season), alpha = 0.5) +
                         scale_fill_manual(values = RColorBrewer::brewer.pal(8, "Dark2")) + 
                         ggtitle("Fungi")
ba

tmp <- t1$res_ordination$scores
# differential test with trans_env class
t2 <- trans_env$new(dataset = dataset, add_data = tmp[, 1:2])
# 'KW_dunn' for non-parametric test
t2$cal_diff(group = "River.Season", method = "anova")

bb <- t2$plot_diff(measure = "PCo2", add_sig = T, color = rep("white",8)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), 
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none", 
            axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
            axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1))
bb

# height of the upper figure and width of the right-hand figure are both 0.2-fold of the main figure
b4 <- ba #%>% aplot::insert_right(bb, width = 0.5)
b4

# manova for specified group set: such as "Region"
t1$cal_manova(manova_set = "River*Reach*Season"); t1$res_manova

t1$cal_anosim(group = "River.Reach.Season"); t1$res_anosim
#t1$cal_betadisper(); t1$res_betadisper

t1$cal_group_distance(within_group = TRUE) # return t1$res_group_distance
t1$cal_group_distance_diff(method = "anova") # perform Wilcoxon Rank Sum and Signed Rank Tests
c4 <- t1$plot_group_distance(color = rep("white", 16)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3",
                            "#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), 
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none",
            axis.title.y = element_text(size = 12),
            axis.text.x = element_text(size = 8, angle = 45, hjust = 1), 
            axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) + ylim(0.02, 0.105) + 
      ggtitle("Fungi")
```

```{r, Protists PCoA Plot}
dataset <- clone(pro.mco.fld)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# measure parameter must be one of names(dataset$beta_diversity)
t1 <- trans_beta$new(dataset = dataset, group = "River.Reach.Season", measure = "wei_unifrac")
# PCoA, PCA and NMDS are available
t1$cal_ordination(method = "PCoA"); class(t1$res_ordination) # t1$res_ordination is the ordination result list
# plot the PCoA result with confidence ellipse
ba <- t1$plot_ordination(plot_color = "Season", 
                         plot_shape = "River.Reach", 
                         plot_type = c("point"), 
                         ellipse_chull_alpha = 0,
                         #add_sample_label = "Reach",
                         color_values = RColorBrewer::brewer.pal(8, "Dark2")) + theme_bw() +
                         geom_hline(yintercept = 0, linetype = "dashed", color = "gray85") +
                         geom_vline(xintercept = 0, linetype = "dashed", color = "gray85") +
                         scale_shape_manual(values = c(17, 2, 19, 1)) +
                         theme_bw() + 
                         theme(panel.grid = element_blank(), legend.position = "none", #axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
                               axis.text.y = element_text(angle = 90, hjust = 0.5), legend.title = element_text(face = "bold"),
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) +
                         ggalt::geom_encircle(aes(s_shape = 1, expand = 0, group = River.Reach.Season, fill = Season, colour = Season), alpha = 0.5) +
                         scale_fill_manual(values = RColorBrewer::brewer.pal(8, "Dark2")) + 
                         ggtitle("Protists")
ba

tmp <- t1$res_ordination$scores
# differential test with trans_env class
t2 <- trans_env$new(dataset = dataset, add_data = tmp[, 1:2])
# 'KW_dunn' for non-parametric test
t2$cal_diff(group = "River.Season", method = "anova")

bb <- t2$plot_diff(measure = "PCo2", add_sig = T, color = rep("white",8)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"),
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none", 
            axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
            axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1))
bb

# height of the upper figure and width of the right-hand figure are both 0.2-fold of the main figure
b5 <- ba #%>% aplot::insert_right(bb, width = 0.5)
b5

# manova for specified group set: such as "Region"
t1$cal_manova(manova_set = "River*Reach*Season"); t1$res_manova

t1$cal_anosim(group = "River.Reach.Season"); t1$res_anosim
#t1$cal_betadisper(); t1$res_betadisper

t1$cal_group_distance(within_group = TRUE) # return t1$res_group_distance
t1$cal_group_distance_diff(method = "anova") # perform Wilcoxon Rank Sum and Signed Rank Tests
c5 <- t1$plot_group_distance(color = rep("white", 16)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3",
                            "#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"),
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none",
            axis.title.y = element_text(size = 12),
            axis.text.x = element_text(size = 8, angle = 45, hjust = 1), 
            axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) + ylim(0.05, 0.31) + 
      ggtitle("Protists")
```

```{r, Metazoan PCoA Plot}
dataset <- clone(met.mco.fld)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# measure parameter must be one of names(dataset$beta_diversity)
t1 <- trans_beta$new(dataset = dataset, group = "River.Reach.Season", measure = "wei_unifrac")
# PCoA, PCA and NMDS are available
t1$cal_ordination(method = "PCoA"); class(t1$res_ordination) # t1$res_ordination is the ordination result list
# plot the PCoA result with confidence ellipse
ba <- t1$plot_ordination(plot_color = "Season", 
                         plot_shape = "River.Reach", 
                         plot_type = c("point"), 
                         ellipse_chull_alpha = 0,
                         #add_sample_label = "Reach",
                         color_values = RColorBrewer::brewer.pal(8, "Dark2")) + theme_bw() +
                         geom_hline(yintercept = 0, linetype = "dashed", color = "gray85") +
                         geom_vline(xintercept = 0, linetype = "dashed", color = "gray85") +
                         scale_shape_manual(values = c(17, 2, 19, 1)) +
                         theme_bw() + 
                         theme(panel.grid = element_blank(), legend.position = "none", #axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
                               axis.text.y = element_text(angle = 90, hjust = 0.5), legend.title = element_text(face = "bold"),
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) +
                         ggalt::geom_encircle(aes(s_shape = 1, expand = 0, group = River.Reach.Season, fill = Season, colour = Season), alpha = 0.5) +
                         scale_fill_manual(values = RColorBrewer::brewer.pal(8, "Dark2")) + 
                         ggtitle("Metazoan")
ba

tmp <- t1$res_ordination$scores
# differential test with trans_env class
t2 <- trans_env$new(dataset = dataset, add_data = tmp[, 1:2])
# 'KW_dunn' for non-parametric test
t2$cal_diff(group = "River.Season", method = "anova")

bb <- t2$plot_diff(measure = "PCo2", add_sig = T, color = rep("white",8)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none", 
            axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
            axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1))
bb

# height of the upper figure and width of the right-hand figure are both 0.2-fold of the main figure
b6 <- ba #%>% aplot::insert_right(bb, width = 0.5)
b6

# manova for specified group set: such as "Region"
t1$cal_manova(manova_set = "River*Reach*Season"); t1$res_manova

t1$cal_anosim(group = "River.Reach.Season"); t1$res_anosim
#t1$cal_betadisper(); t1$res_betadisper

t1$cal_group_distance(within_group = TRUE) # return t1$res_group_distance
t1$cal_group_distance_diff(method = "anova") # perform Wilcoxon Rank Sum and Signed Rank Tests
c6 <- t1$plot_group_distance(color = rep("white", 16)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3",
                            "#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none",
            axis.title.y = element_text(size = 12),
            axis.text.x = element_text(size = 8, angle = 45, hjust = 1), 
            axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) + #ylim(0.09, 0.5) + 
      ggtitle("Metazoan")
```

```{r, Function PCoA Plot}
dataset <- clone(keg.mco.fld)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# measure parameter must be one of names(dataset$beta_diversity)
t1 <- trans_beta$new(dataset = dataset, group = "River.Reach.Season", measure = "bray")
# PCoA, PCA and NMDS are available
t1$cal_ordination(method = "PCoA"); class(t1$res_ordination) # t1$res_ordination is the ordination result list
# plot the PCoA result with confidence ellipse
ba <- t1$plot_ordination(plot_color = "Season", 
                         plot_shape = "River.Reach", 
                         plot_type = c("point"), 
                         ellipse_chull_alpha = 0,
                         #add_sample_label = "Reach",
                         color_values = RColorBrewer::brewer.pal(8, "Dark2")) + theme_bw() +
                         geom_hline(yintercept = 0, linetype = "dashed", color = "gray85") +
                         geom_vline(xintercept = 0, linetype = "dashed", color = "gray85") +
                         scale_shape_manual(values = c(17, 2, 19, 1)) +
                         theme_bw() + 
                         theme(panel.grid = element_blank(), legend.position = "none", #axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
                               axis.text.y = element_text(angle = 90, hjust = 0.5), legend.title = element_text(face = "bold"),
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) +
                         ggalt::geom_encircle(aes(s_shape = 1, expand = 0, group = River.Reach.Season, fill = Season, colour = Season), alpha = 0.5) +
                         scale_fill_manual(values = RColorBrewer::brewer.pal(8, "Dark2")) + 
                         ggtitle("All Predicted Functions")
ba

tmp <- t1$res_ordination$scores
# differential test with trans_env class
t2 <- trans_env$new(dataset = dataset, add_data = tmp[, 1:2])
# 'KW_dunn' for non-parametric test
t2$cal_diff(group = "River.Season", method = "anova")

bb <- t2$plot_diff(measure = "PCo2", add_sig = T, color = rep("white",8)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none", 
            axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
            axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1))
bb

# height of the upper figure and width of the right-hand figure are both 0.2-fold of the main figure
b7 <- ba #%>% aplot::insert_right(bb, width = 0.5) 
b7

# manova for specified group set: such as "Region"
t1$cal_manova(manova_set = "River*Reach*Season"); t1$res_manova

t1$cal_anosim(group = "River.Reach.Season"); t1$res_anosim
#t1$cal_betadisper(); t1$res_betadisper

t1$cal_group_distance(within_group = TRUE) # return t1$res_group_distance
t1$cal_group_distance_diff(method = "anova") # perform Wilcoxon Rank Sum and Signed Rank Tests
c7 <- t1$plot_group_distance(color = rep("white", 16)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3",
                            "#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none",
            axis.title.y = element_text(size = 12),
            axis.text.x = element_text(size = 8, angle = 45, hjust = 1), 
            axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) + #ylim(0.09, 0.5) + 
      ggtitle("Metazoan")
```

```{r, Merge PCoA plots}
( b1 + b2) / ( b3 + b4 ) / ( b5 + b6 ) + 
patchwork::plot_layout(widths = unit(c(10,10,10,10,10,10),c("null","null","null","null","null","null"))) 
```

```{r, Merge PCoA plots}
( c2 + c2 ) / ( c3 + c4 ) / ( c5 + c6) + patchwork::plot_layout(widths = unit(c(10,10,10,10,10,10),c("null","null","null","null","null","null"))) 
```

```{r, PERMANOVA plot of proportion of variance}
# Load libraries
library("tidyr"); library("ggplot2"); library("paletteer")
manova.df <- data.frame(Group = c("Prokaryotes", "Algae", "Fungi", "Protists", "Metazoan"),
                        River = c(0.14912752,0.41561909,0.28704619,0.20866742,0.0524531),
                        Reach = c(0.13890579,0.04008244,0.0767617,0.08450973,0.04881504),
                        Season = c(0.09343573,0.17543133,0.11964821,0.10060574,0.22927824),
                        River.Reach = c(0.13182928,0.07450847,0.09087205,0.08979094,0.03054893),
                        River.Season = c(0.12058207,0.06173121,0.14127383,0.1158869,0.16063719),
                        Reach.Season = c(0.10172173,0.05446374,0.07372515,0.09273264,0.10588322),
                        River.Reach.Season = c(0.09969194,0.03724603,0.05478977,0.06254156,0.090277),
                        Residual = c(0.16470593,0.1409177,0.15588309,0.24526505,0.28210729))

# Multiply all numeric columns by 100
manova.df[,-1] <- manova.df[,-1] * 100 

manova_long <- pivot_longer(manova.df, cols = -Group, names_to = "Effect", values_to = "Value")
manova_long$Group <- factor(manova_long$Group, levels = c("Metazoan", "Protists", "Fungi", "Algae", "Prokaryotes"))
manova_long$Effect <- factor(manova_long$Effect, levels = rev(c("Season", "River", "Reach", "River.Reach", "River.Season",
                                                                "Reach.Season", "River.Reach.Season", "Residual")))
 
# Create transposed stacked bar plot with custom colors
r1 <- ggplot(manova_long, aes(x = Group, y = Value, fill = Effect)) + 
      geom_bar(stat = "identity", alpha = 0.9) +
      geom_text(aes(label = paste0(round(Value, 1), "%")), position = position_stack(vjust = 0.5), size = 3) +
      scale_fill_paletteer_d("MetBrewer::Monet", direction = -1) + theme_classic2() +
      labs(x = NULL, y = expression("Proportion of variance explained, R"^2*" (%)")) + 
      theme(legend.position = "top", legend.title = element_text(size = 10, face = "bold"), 
            legend.text = element_text(size = 9),
            axis.title.y = element_blank(), axis.text.y = element_text(size = 10),
            axis.text.x = element_text(size = 10)) +
      guides(fill = guide_legend(title = "Category", ncol = 4, reverse = TRUE)) + coord_flip() 
r1
```

### Phylogenetic and null model analysis (NOT USED)

```{r, beta nearest taxon index (betaNTI)}
# In recent decades, the integration of phylogenetic analysis and null model promotes the inference of niche and neutral influences on community assembly more powerfully by adding a phylogeny dimension (Webb et al. 2002; Kembel et al. 2010; Stegen et al. 2013). The trans_nullmodel class provides an encapsulation, including the calculation of the phylogenetic signal, beta mean pairwise phylogenetic distance (betaMPD), beta mean nearest taxon distance (betaMNTD), beta nearest taxon index (betaNTI), beta net relatedness index (betaNRI) and Bray-Curtis-based Raup-Crick (RCbray). The approach for phylogenetic signal analysis is based on the mantel correlogram (C. Liu et al. 2017), in which the change of phylogenetic signal is intuitional and clear compared to other approaches.

# The combinations between RCbray and betaNTI can be used to infer the strength of each ecological process dominating the community assembly under the specific hypothesis (Stegen et al. 2013).

dataset <- clone(bac.mco.fld)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# generate trans_nullmodel object
# as an example, we only use high abundance OTU with mean relative abundance > 0.0005
t1 <- trans_nullmodel$new(dataset, filter_thres = 0.0005, add_data = dataset$sample_table[, c(6:24)])

# use pH as the test variable
t1$cal_mantel_corr(use_env = "pH") # return t1$res_mantel_corr
t1$plot_mantel_corr() # plot the mantel correlogram

set.seed(0327)

# null model run 500 times
t1$cal_ses_betamntd(runs = 1000, abundance.weighted = TRUE, null.model = "taxa.labels") # return t1$res_ses_betamntd

## Plot the betaNRI.
dataset$beta_diversity[["betaNTI"]] <- dataset # add betaNRI matrix to beta_diversity list.
dataset$beta_diversity[["betaNTI"]] <- t1$res_ses_betamntd
t2 <- trans_beta$new(dataset = dataset, group = "River.Reach.Season", measure = "betaNTI") ## create trans_beta class, use measure "betaNTI".

t2$cal_ordination(method = "PCoA"); class(t2$res_ordination) ## t2$res_ordination is the ordination result list.
t2$cal_group_distance(); t2$cal_group_distance_diff(method = "anova") ## Transform the distance for each group.

## Plot the PCoA result with confidence ellipse.
na <- t2$plot_ordination(plot_color = "Season", 
                         plot_shape = "River.Reach", 
                         plot_type = c("point"), 
                         ellipse_chull_alpha = 0,
                         #add_sample_label = "Reach",
                         color_values = RColorBrewer::brewer.pal(8, "Dark2")) + theme_bw() +
                         geom_hline(yintercept = 0, linetype = "dashed", color = "gray85") +
                         geom_vline(xintercept = 0, linetype = "dashed", color = "gray85") +
                         scale_shape_manual(values = c(17, 2, 19, 1)) +
                         theme_bw() + 
                         theme(panel.grid = element_blank(), legend.position = "right", 
                               #axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
                               axis.text.y = element_text(angle = 90, hjust = 0.5), legend.title = element_text(face = "bold"),
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) +
                         ggalt::geom_encircle(aes(s_shape = 1, expand = 0, group = River.Reach.Season, 
                                                  fill = Season, colour = Season), alpha = 0.5) +
                         scale_fill_manual(values = RColorBrewer::brewer.pal(8, "Dark2")) + 
                         ggtitle("Prokaryotes")
na

t2$cal_group_distance(within_group = TRUE) # return t1$res_group_distance
t2$cal_group_distance_diff(method = "anova") # perform Wilcoxon Rank Sum and Signed Rank Tests

nb <- t2$plot_group_distance(color = rep("white", 16)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3",
                            "#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), 
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = -2, linetype = 2, color = "gray50") + geom_hline(yintercept = 2, linetype = 2, color = "gray50") +
      theme(legend.position = "none",
            axis.title.y = element_text(size = 12),
            axis.text.x = element_text(size = 8, angle = 45, hjust = 1), 
            axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) #+ ylim(0.5, 0.6)

nb

# Combine the plots in one row
#n1 <- (na / nb ) + patchwork::plot_layout(guides = 'collect') + patchwork::plot_annotation(); n1
#n1
```

```{r, RCbray (Bray-Curtis-based Raup-Crick)}
# RCbray (Bray-Curtis-based Raup-Crick) can be calculated using function cal_rcbray() to assess whether the compositional turnover was governed primarily by drift (Chase et al. 2011). We applied null model to simulate species distribution by randomly sampling individuals from each species pool with preserving species occurrence frequency and sample species richness (C. Liu et al. 2017).
t1$cal_rcbray(runs = 1000); t1$res_rcbray

## Plot the betaNRI.
dataset$beta_diversity[["RCbray"]] <- dataset # add betaNRI matrix to beta_diversity list.
dataset$beta_diversity[["RCbray"]] <- t1$res_rcbray
t2 <- trans_beta$new(dataset = dataset, group = "River.Reach.Season", measure = "RCbray") ## create trans_beta class, use measure "RCbray".

t2$cal_ordination(method = "PCoA"); class(t2$res_ordination) ## t1$res_ordination is the ordination result list.
t2$cal_group_distance(); t2$cal_group_distance_diff(method = "anova") ## Transform the distance for each group.

## Plot the PCoA result with confidence ellipse.
nc <- t2$plot_ordination(plot_color = "Season", 
                         plot_shape = "River.Reach", 
                         plot_type = c("point"), 
                         ellipse_chull_alpha = 0,
                         #add_sample_label = "Reach",
                         color_values = RColorBrewer::brewer.pal(8, "Dark2")) + theme_bw() +
                         geom_hline(yintercept = 0, linetype = "dashed", color = "gray85") +
                         geom_vline(xintercept = 0, linetype = "dashed", color = "gray85") +
                         scale_shape_manual(values = c(17, 2, 19, 1)) +
                         theme_bw() + 
                         theme(panel.grid = element_blank(), legend.position = "right", 
                               #axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
                               axis.text.y = element_text(angle = 90, hjust = 0.5), legend.title = element_text(face = "bold"),
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) +
                         ggalt::geom_encircle(aes(s_shape = 1, expand = 0, group = River.Reach.Season, 
                                                  fill = Season, colour = Season), alpha = 0.5) +
                         scale_fill_manual(values = RColorBrewer::brewer.pal(8, "Dark2")) + 
                         ggtitle("Prokaryotes")
nc

t2$cal_group_distance(within_group = TRUE) # return t2$res_group_distance
t2$cal_group_distance_diff(method = "anova") # perform Wilcoxon Rank Sum and Signed Rank Tests

nd <- t2$plot_group_distance(color = rep("white", 16)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3",
                            "#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), 
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = -2, linetype = 2, color = "gray50") + geom_hline(yintercept = 2, linetype = 2, color = "gray50") +
      theme(legend.position = "none",
            axis.title.y = element_text(size = 12),
            axis.text.x = element_text(size = 8, angle = 45, hjust = 1), 
            axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) #+ ylim(0.5, 0.6)

nd
```

```{r, Calculate the proportion of the inferred processes on the community assembly}
# use betaNTI and rcbray to evaluate processes
t1$cal_process(use_betamntd = TRUE, group = "River.Season"); t1$res_process
#t1$cal_process(use_betamntd = TRUE); t1$res_process

library(ggplot2)
library(dplyr)

# Assuming your data is in t1$res_process
df <- t1$res_process

# Prepare data: calculate fraction and label
df <- df %>% group_by(River.Season) %>% mutate(fraction = percentage / sum(percentage), 
                                               label = paste0(process, "\n", round(percentage, 1), "%"))

# Plot pie charts
p1 <- ggplot(df, aes(x = "", y = fraction, fill = process)) +
      geom_bar(stat = "identity", width = 1, color = "white") +
      coord_polar("y") +
      facet_wrap(~ River.Season, ncol = 4) +
      theme_void() +
      theme(strip.text = element_text(size = 12, face = "bold")) +
      labs(title = "Ecological Process Distribution by River and Season", fill = "Process")

p1
```

### Seasonal indicator species analysis

```{r, differential analysis to generate spatiotemporal communities}
## STEP 9a: Differential abundance test: RF (random forest + differential test).
dataset <- clone(bac.mco.fld)
dataset$add_rownames2taxonomy("ASV"); dataset$cal_abund(); dataset$cal_abund(select_cols = "ASV")

set.seed(0327)

## The ‘rf’ method depends on the random forest(Beck and Foster 2014; Yatsunenko et al. 2012) and the non-parametric test. Calculate random forest by bootstrapping only using significant features. MeanDecreaseGini is selected as the indicator value in the analysis.

## Use Genus level for parameter taxa_level. Using only high abundance ASV with mean relative abundance > 0.005.
r1 <- trans_diff$new(dataset = dataset, method = "rf", group = "River.Reach.Season", 
                     taxa_level = "ASV", 
                     filter_thres = 0.001); r1$res_diff

r2 <- r1$plot_diff_bar(use_number = 1:50, width = 0.8,
                       color_values = c("#D95F02", "#1B9E77", "#7570B3", "#E7298A",
                                        "#D95F02", "#1B9E77", "#7570B3", "#E7298A",
                                        "#D95F02", "#1B9E77", "#7570B3", "#E7298A",
                                        "#D95F02", "#1B9E77", "#7570B3", "#E7298A")) + theme_classic2()
r2

#openxlsx::write.xlsx(r1$res_diff, file = "src/bac.rf.xlsx", rowNames = FALSE)

seasonal.asv.list <- c(r1$res_diff$Taxa)
tmp <- file2meco::meco2phyloseq(dataset)

# Subset the phyloseq object to keep only these ASVs
tmp.psq.ssn <- prune_taxa(seasonal.asv.list, tmp)
tmp.mco.ssn <- file2meco::phyloseq2meco(tmp.psq.ssn); tmp.mco.ssn$tidy_dataset()
tmp.mco.ssn

tmp.mco.ssn$cal_abund(); class(tmp.mco.ssn$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
tmp.mco.ssn$cal_alphadiv(PD = TRUE); class(tmp.mco.ssn$alpha_diversity) ## Calculate the Alpha diversity w/ Faith's phylogenetic diversity.
tmp.mco.ssn$cal_betadiv(unifrac = TRUE); class(tmp.mco.ssn$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
print(tmp.mco.ssn)

saveRDS(tmp.mco.ssn, "src/bac.mco.ssn.RDS")
```

```{r, differential analysis to generate season}
## STEP 9a: Differential abundance test: RF (random forest + differential test).
dataset <- clone(bac.mco.fld)
dataset$add_rownames2taxonomy("ASV"); dataset$cal_abund(); dataset$cal_abund(select_cols = "ASV")

set.seed(0327)

## The ‘rf’ method depends on the random forest(Beck and Foster 2014; Yatsunenko et al. 2012) and the non-parametric test. Calculate random forest by bootstrapping only using significant features. MeanDecreaseGini is selected as the indicator value in the analysis.

## Use Genus level for parameter taxa_level. Using only high abundance ASV with mean relative abundance > 0.005.
r1 <- trans_diff$new(dataset = dataset, method = "rf", group = "Season", 
                     taxa_level = "ASV", 
                     filter_thres = 0.001); r1$res_diff

r2 <- r1$plot_diff_bar(use_number = 1:50, width = 0.8,
                       color_values = c("#D95F02", "#E7298A")) + theme_classic2()
r2

#openxlsx::write.xlsx(r1$res_diff, file = "src/bac.rf.xlsx", rowNames = FALSE)

seasonal.asv.list <- c(r1$res_diff$Taxa)
tmp <- file2meco::meco2phyloseq(dataset)

# Subset the phyloseq object to keep only these ASVs
tmp.psq.ssn <- prune_taxa(seasonal.asv.list, tmp)
tmp.mco.ssn <- file2meco::phyloseq2meco(tmp.psq.ssn); tmp.mco.ssn$tidy_dataset()
tmp.mco.ssn

tmp.mco.ssn$cal_abund(); class(tmp.mco.ssn$taxa_abund) ## Calculate the taxa abundance at each taxonomic rank using cal_abund().
tmp.mco.ssn$cal_alphadiv(PD = TRUE); class(tmp.mco.ssn$alpha_diversity) ## Calculate the Alpha diversity w/ Faith's phylogenetic diversity.
tmp.mco.ssn$cal_betadiv(unifrac = TRUE); class(tmp.mco.ssn$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity
print(tmp.mco.ssn)

saveRDS(tmp.mco.ssn, "src/bac.mco.ssn.grp.RDS")
```


```{r, Seasonal Prokaryotes PCoA Plot}
dataset <- clone(bac.mco.ssn)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# measure parameter must be one of names(dataset$beta_diversity)
t1 <- trans_beta$new(dataset = dataset, group = "River.Reach.Season", measure = "wei_unifrac")
# PCoA, PCA and NMDS are available
t1$cal_ordination(method = "PCoA"); class(t1$res_ordination) # t1$res_ordination is the ordination result list
# plot the PCoA result with confidence ellipse
ba <- t1$plot_ordination(plot_color = "Season", 
                         plot_shape = "River.Reach", 
                         plot_type = c("point"), 
                         ellipse_chull_alpha = 0,
                         #add_sample_label = "Reach",
                         color_values = RColorBrewer::brewer.pal(8, "Dark2")) + theme_bw() +
                         geom_hline(yintercept = 0, linetype = "dashed", color = "gray85") +
                         geom_vline(xintercept = 0, linetype = "dashed", color = "gray85") +
                         scale_shape_manual(values = c(17, 2, 19, 1)) +
                         theme_bw() + 
                         theme(panel.grid = element_blank(), legend.position = "none", #axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
                               axis.text.y = element_text(angle = 90, hjust = 0.5), legend.title = element_text(face = "bold"),
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) +
                         ggalt::geom_encircle(aes(s_shape = 1, expand = 0, group = River.Reach.Season, fill = Season, colour = Season), alpha = 0.5) +
                         scale_fill_manual(values = RColorBrewer::brewer.pal(8, "Dark2")) + 
                         ggtitle("Prokaryotes")
ba

tmp <- t1$res_ordination$scores
# differential test with trans_env class
t2 <- trans_env$new(dataset = dataset, add_data = tmp[, 1:2])
# 'KW_dunn' for non-parametric test
t2$cal_diff(group = "River.Season", method = "anova")

bb <- t2$plot_diff(measure = "PCo2", add_sig = T, color = rep("white", 8)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), 
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none", 
            axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
            axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1))
bb

# height of the upper figure and width of the right-hand figure are both 0.2-fold of the main figure
b1 <- ba #%>% aplot::insert_right(bb, width = 0.5)
b1

# manova for specified group set: such as "Region"
t1$cal_manova(manova_set = "River*Reach*Season"); t1$res_manova

t1$cal_anosim(group = "River.Reach.Season"); t1$res_anosim
#t1$cal_betadisper(); t1$res_betadisper

t1$cal_group_distance(within_group = TRUE) # return t1$res_group_distance
t1$cal_group_distance_diff(method = "anova") # perform Wilcoxon Rank Sum and Signed Rank Tests
c1 <- t1$plot_group_distance(color = rep("white", 16)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3",
                            "#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), 
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none",
            axis.title.y = element_text(size = 12),
            axis.text.x = element_text(size = 8, angle = 45, hjust = 1), 
            axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) #+ ylim(0.5, 0.6) + 
      ggtitle("Prokaryotes")
```

```{r, Seasonal Algae PCoA Plot}
dataset <- clone(alg.mco.ssn)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# measure parameter must be one of names(dataset$beta_diversity)
t1 <- trans_beta$new(dataset = dataset, group = "River.Reach.Season", measure = "wei_unifrac")
# PCoA, PCA and NMDS are available
t1$cal_ordination(method = "PCoA"); class(t1$res_ordination) # t1$res_ordination is the ordination result list
# plot the PCoA result with confidence ellipse
ba <- t1$plot_ordination(plot_color = "Season", 
                         plot_shape = "River.Reach", 
                         plot_type = c("point"), 
                         ellipse_chull_alpha = 0,
                         #add_sample_label = "Reach",
                         color_values = RColorBrewer::brewer.pal(8, "Dark2")) + theme_bw() +
                         geom_hline(yintercept = 0, linetype = "dashed", color = "gray85") +
                         geom_vline(xintercept = 0, linetype = "dashed", color = "gray85") +
                         scale_shape_manual(values = c(17, 2, 19, 1)) +
                         theme_bw() + 
                         theme(panel.grid = element_blank(), legend.position = "none", #axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
                               axis.text.y = element_text(angle = 90, hjust = 0.5), legend.title = element_text(face = "bold"),
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) +
                         ggalt::geom_encircle(aes(s_shape = 1, expand = 0, group = River.Reach.Season, fill = Season, colour = Season), alpha = 0.5) +
                         scale_fill_manual(values = RColorBrewer::brewer.pal(8, "Dark2")) + 
                         ggtitle("Algae")
ba

tmp <- t1$res_ordination$scores
# differential test with trans_env class
t2 <- trans_env$new(dataset = dataset, add_data = tmp[, 1:2])
# 'KW_dunn' for non-parametric test
t2$cal_diff(group = "River.Season", method = "anova")

bb <- t2$plot_diff(measure = "PCo2", add_sig = T, color = rep("white",8)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), 
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none", 
            axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
            axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1))
bb

# height of the upper figure and width of the right-hand figure are both 0.2-fold of the main figure
b3 <- ba #%>% aplot::insert_right(bb, width = 0.5)
b3

# manova for specified group set: such as "Region"
t1$cal_manova(manova_set = "River*Reach*Season"); t1$res_manova

t1$cal_anosim(group = "River.Reach.Season"); t1$res_anosim
#t1$cal_betadisper(); t1$res_betadisper

t1$cal_group_distance(within_group = TRUE) # return t1$res_group_distance
t1$cal_group_distance_diff(method = "anova") # perform Wilcoxon Rank Sum and Signed Rank Tests
c3 <- t1$plot_group_distance(color = rep("white", 16)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3",
                            "#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), 
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none",
            axis.title.y = element_text(size = 12),
            axis.text.x = element_text(size = 8, angle = 45, hjust = 1), 
            axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) + #ylim(0.025, 0.06) + 
      ggtitle("Algae")
```

```{r, Seasonal Fungi PCoA Plot}
dataset <- clone(fun.mco.ssn)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# measure parameter must be one of names(dataset$beta_diversity)
t1 <- trans_beta$new(dataset = dataset, group = "River.Reach.Season", measure = "wei_unifrac")
# PCoA, PCA and NMDS are available
t1$cal_ordination(method = "PCoA"); class(t1$res_ordination) # t1$res_ordination is the ordination result list
# plot the PCoA result with confidence ellipse
ba <- t1$plot_ordination(plot_color = "Season", 
                         plot_shape = "River.Reach", 
                         plot_type = c("point"), 
                         ellipse_chull_alpha = 0,
                         #add_sample_label = "Reach",
                         color_values = RColorBrewer::brewer.pal(8, "Dark2")) + theme_bw() +
                         geom_hline(yintercept = 0, linetype = "dashed", color = "gray85") +
                         geom_vline(xintercept = 0, linetype = "dashed", color = "gray85") +
                         scale_shape_manual(values = c(17, 2, 19, 1)) +
                         theme_bw() + 
                         theme(panel.grid = element_blank(), legend.position = "none", #axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
                               axis.text.y = element_text(angle = 90, hjust = 0.5), legend.title = element_text(face = "bold"),
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) +
                         ggalt::geom_encircle(aes(s_shape = 1, expand = 0, group = River.Reach.Season, fill = Season, colour = Season), alpha = 0.5) +
                         scale_fill_manual(values = RColorBrewer::brewer.pal(8, "Dark2")) + 
                         ggtitle("Fungi")
ba

tmp <- t1$res_ordination$scores
# differential test with trans_env class
t2 <- trans_env$new(dataset = dataset, add_data = tmp[, 1:2])
# 'KW_dunn' for non-parametric test
t2$cal_diff(group = "River.Season", method = "anova")

bb <- t2$plot_diff(measure = "PCo2", add_sig = T, color = rep("white",8)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), 
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none", 
            axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
            axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1))
bb

# height of the upper figure and width of the right-hand figure are both 0.2-fold of the main figure
b4 <- ba #%>% aplot::insert_right(bb, width = 0.5)
b4

# manova for specified group set: such as "Region"
t1$cal_manova(manova_set = "River*Reach*Season"); t1$res_manova

t1$cal_anosim(group = "River.Reach.Season"); t1$res_anosim
#t1$cal_betadisper(); t1$res_betadisper

t1$cal_group_distance(within_group = TRUE) # return t1$res_group_distance
t1$cal_group_distance_diff(method = "anova") # perform Wilcoxon Rank Sum and Signed Rank Tests
c4 <- t1$plot_group_distance(color = rep("white", 16)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3",
                            "#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), 
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none",
            axis.title.y = element_text(size = 12),
            axis.text.x = element_text(size = 8, angle = 45, hjust = 1), 
            axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) + ylim(0.02, 0.105) + 
      ggtitle("Fungi")
```

```{r, Seasonal Protists PCoA Plot}
dataset <- clone(pro.mco.ssn)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# measure parameter must be one of names(dataset$beta_diversity)
t1 <- trans_beta$new(dataset = dataset, group = "River.Reach.Season", measure = "wei_unifrac")
# PCoA, PCA and NMDS are available
t1$cal_ordination(method = "PCoA"); class(t1$res_ordination) # t1$res_ordination is the ordination result list
# plot the PCoA result with confidence ellipse
ba <- t1$plot_ordination(plot_color = "Season", 
                         plot_shape = "River.Reach", 
                         plot_type = c("point"), 
                         ellipse_chull_alpha = 0,
                         #add_sample_label = "Reach",
                         color_values = RColorBrewer::brewer.pal(8, "Dark2")) + theme_bw() +
                         geom_hline(yintercept = 0, linetype = "dashed", color = "gray85") +
                         geom_vline(xintercept = 0, linetype = "dashed", color = "gray85") +
                         scale_shape_manual(values = c(17, 2, 19, 1)) +
                         theme_bw() + 
                         theme(panel.grid = element_blank(), legend.position = "none", #axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
                               axis.text.y = element_text(angle = 90, hjust = 0.5), legend.title = element_text(face = "bold"),
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) +
                         ggalt::geom_encircle(aes(s_shape = 1, expand = 0, group = River.Reach.Season, fill = Season, colour = Season), alpha = 0.5) +
                         scale_fill_manual(values = RColorBrewer::brewer.pal(8, "Dark2")) + 
                         ggtitle("Protists")
ba

tmp <- t1$res_ordination$scores
# differential test with trans_env class
t2 <- trans_env$new(dataset = dataset, add_data = tmp[, 1:2])
# 'KW_dunn' for non-parametric test
t2$cal_diff(group = "River.Season", method = "anova")

bb <- t2$plot_diff(measure = "PCo2", add_sig = T, color = rep("white",8)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"),
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none", 
            axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
            axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1))
bb

# height of the upper figure and width of the right-hand figure are both 0.2-fold of the main figure
b5 <- ba #%>% aplot::insert_right(bb, width = 0.5)
b5

# manova for specified group set: such as "Region"
t1$cal_manova(manova_set = "River*Reach*Season"); t1$res_manova

t1$cal_anosim(group = "River.Reach.Season"); t1$res_anosim
#t1$cal_betadisper(); t1$res_betadisper

t1$cal_group_distance(within_group = TRUE) # return t1$res_group_distance
t1$cal_group_distance_diff(method = "anova") # perform Wilcoxon Rank Sum and Signed Rank Tests
c5 <- t1$plot_group_distance(color = rep("white", 16)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3",
                            "#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"),
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none",
            axis.title.y = element_text(size = 12),
            axis.text.x = element_text(size = 8, angle = 45, hjust = 1), 
            axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) + ylim(0.05, 0.31) + 
      ggtitle("Protists")
```

```{r, Seasonal Metazoan PCoA Plot}
dataset <- clone(met.mco.ssn)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# measure parameter must be one of names(dataset$beta_diversity)
t1 <- trans_beta$new(dataset = dataset, group = "River.Reach.Season", measure = "wei_unifrac")
# PCoA, PCA and NMDS are available
t1$cal_ordination(method = "PCoA"); class(t1$res_ordination) # t1$res_ordination is the ordination result list
# plot the PCoA result with confidence ellipse
ba <- t1$plot_ordination(plot_color = "Season", 
                         plot_shape = "River.Reach", 
                         plot_type = c("point"), 
                         ellipse_chull_alpha = 0,
                         #add_sample_label = "Reach",
                         color_values = RColorBrewer::brewer.pal(8, "Dark2")) + theme_bw() +
                         geom_hline(yintercept = 0, linetype = "dashed", color = "gray85") +
                         geom_vline(xintercept = 0, linetype = "dashed", color = "gray85") +
                         scale_shape_manual(values = c(17, 2, 19, 1)) +
                         theme_bw() + 
                         theme(panel.grid = element_blank(), legend.position = "none", #axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
                               axis.text.y = element_text(angle = 90, hjust = 0.5), legend.title = element_text(face = "bold"),
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) +
                         ggalt::geom_encircle(aes(s_shape = 1, expand = 0, group = River.Reach.Season, fill = Season, colour = Season), alpha = 0.5) +
                         scale_fill_manual(values = RColorBrewer::brewer.pal(8, "Dark2")) + 
                         ggtitle("Metazoan")
ba

tmp <- t1$res_ordination$scores
# differential test with trans_env class
t2 <- trans_env$new(dataset = dataset, add_data = tmp[, 1:2])
# 'KW_dunn' for non-parametric test
t2$cal_diff(group = "River.Season", method = "anova")

bb <- t2$plot_diff(measure = "PCo2", add_sig = T, color = rep("white",8)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none", 
            axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
            axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1))
bb

# height of the upper figure and width of the right-hand figure are both 0.2-fold of the main figure
b6 <- ba #%>% aplot::insert_right(bb, width = 0.5)
b6

# manova for specified group set: such as "Region"
t1$cal_manova(manova_set = "River*Reach*Season"); t1$res_manova

t1$cal_anosim(group = "River.Reach.Season"); t1$res_anosim
#t1$cal_betadisper(); t1$res_betadisper

t1$cal_group_distance(within_group = TRUE) # return t1$res_group_distance
t1$cal_group_distance_diff(method = "anova") # perform Wilcoxon Rank Sum and Signed Rank Tests
c6 <- t1$plot_group_distance(color = rep("white", 16)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3",
                            "#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none",
            axis.title.y = element_text(size = 12),
            axis.text.x = element_text(size = 8, angle = 45, hjust = 1), 
            axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) + #ylim(0.09, 0.5) + 
      ggtitle("Metazoan")
```

```{r, Merge Seasonal PCoA plots}
( b1 + b1) / ( b3 + b4 ) / ( b5 + b6 ) + 
patchwork::plot_layout(widths = unit(c(10,10,10,10,10,10),c("null","null","null","null","null","null"))) 
```

### Functional analysis

```{r, Functional Profile PCoA Plot}
dataset <- clone(keg.mco.xen)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# measure parameter must be one of names(dataset$beta_diversity)
t1 <- trans_beta$new(dataset = dataset, group = "River.Reach.Season", measure = "bray")
# PCoA, PCA and NMDS are available
t1$cal_ordination(method = "PCoA"); class(t1$res_ordination) # t1$res_ordination is the ordination result list
# plot the PCoA result with confidence ellipse
ba <- t1$plot_ordination(plot_color = "Season", 
                         plot_shape = "River.Reach", 
                         plot_type = c("point"), 
                         ellipse_chull_alpha = 0,
                         #add_sample_label = "Reach",
                         color_values = RColorBrewer::brewer.pal(8, "Dark2")) + theme_bw() +
                         geom_hline(yintercept = 0, linetype = "dashed", color = "gray85") +
                         geom_vline(xintercept = 0, linetype = "dashed", color = "gray85") +
                         scale_shape_manual(values = c(17, 2, 19, 1)) +
                         theme_bw() + 
                         theme(panel.grid = element_blank(), legend.position = "none", #axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
                               axis.text.y = element_text(angle = 90, hjust = 0.5), legend.title = element_text(face = "bold"),
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) +
                         ggalt::geom_encircle(aes(s_shape = 1, expand = 0, group = River.Reach.Season, fill = Season, colour = Season), alpha = 0.5) +
                         scale_fill_manual(values = RColorBrewer::brewer.pal(8, "Dark2")) + 
                         ggtitle("Predicted Functions")
ba

tmp <- t1$res_ordination$scores
# differential test with trans_env class
t2 <- trans_env$new(dataset = dataset, add_data = tmp[, 1:2])
# 'KW_dunn' for non-parametric test
t2$cal_diff(group = "River.Season", method = "anova")

bb <- t2$plot_diff(measure = "PCo2", add_sig = T, color = rep("white", 8)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), 
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none", 
            axis.text.x = element_blank(), axis.ticks.x = element_blank(), 
            axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1))
bb

# height of the upper figure and width of the right-hand figure are both 0.2-fold of the main figure
b1 <- ba #%>% aplot::insert_right(bb, width = 0.5)
b1

# manova for specified group set: such as "Region"
t1$cal_manova(manova_set = "River*Reach*Season"); t1$res_manova

t1$cal_anosim(group = "River.Reach.Season"); t1$res_anosim
#t1$cal_betadisper(); t1$res_betadisper

t1$cal_group_distance(within_group = TRUE) # return t1$res_group_distance
t1$cal_group_distance_diff(method = "anova") # perform Wilcoxon Rank Sum and Signed Rank Tests
c1 <- t1$plot_group_distance(color = rep("white", 16)) + theme_bw() +
      geom_boxplot(fill = c("#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3",
                            "#D95F02", "#E7298A", "#1B9E77", "#7570B3", "#D95F02", "#E7298A", "#1B9E77", "#7570B3"), 
                   color = "black", alpha = 0.5) + 
      geom_hline(yintercept = 0, linetype = "dashed", color = "gray95") + theme_bw() +
      theme(legend.position = "none",
            axis.title.y = element_text(size = 12),
            axis.text.x = element_text(size = 8, angle = 45, hjust = 1), 
            axis.text.y = element_text(size = 8, angle = 90, hjust = 0.5),
            panel.grid.minor = element_blank(), panel.grid.major = element_blank(),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) #+ ylim(0.5, 0.6)
```

```{r, Level.1 predicted functions heatmap}
library("ComplexHeatmap"); library("microViz"); library("paletteer") 

dataset <- clone(keg.mco.xen)

#dataset$tax_table$Level.1 <- gsub("l__", "", dataset$tax_table$Superclass)
psq <- file2meco::meco2phyloseq(dataset)

h1 <- psq %>% 
      ps_arrange(River.Reach.Season) %>% 
      tax_transform("identity", rank = "Level.3") %>% 
      comp_heatmap(grid_col = NA, 
                   #colors = heat_palette(palette = "Viridis", rev = TRUE),
                   colors = heat_palette(breaks = c(0.0001, 0.0005, 0.001, 0.005, 0.01), rev = T),
                   tax_anno = taxAnnotation(Prev. = anno_tax_prev(bar_width = 0.6, border = T), 
                                            Log10. = anno_tax_box(trans = "log10", zero_replace = "halfmin")),
                   #taxa_side = "bottom",
                   #heatmap_legend_param = list(#at = 0:5 / 5, 
                   #                            direction = "horizontal", 
                   #                            title_position = "leftcenter", legend_width = grid::unit(4, "cm"),
                   #                            grid_height = grid::unit(5, "mm")),
                   sample_anno = sampleAnnotation(Season = anno_sample("Season"), 
                                                  col = list(Season = c("Winter" = "#7570B3", "Spring" = "#E7298A",
                                                                        "Summer" ="#1B9E77", "Autumn" = "#D95F02")),
                                                  border = TRUE,
                                                  River.Reach = anno_sample_cat(var = "River.Reach", 
                                                  box_col = NA, border_col = "black", legend_title = "River.Reach")),  sample_seriation = "Identity")

h1 %>% ComplexHeatmap::draw(annotation_legend_list = attr(h1, "AnnoLegends"), heatmap_legend_side = "right")
```

## ########################################################## ##
##                                                            ##
##       DOWNSTREAM ANALYSES - CORRELATION ANALYSES           ##
##                                                            ##
## ########################################################## ##

### Correlation between alpha diversity of trophic groups, functions, and biodegradation potential

```{r, Linear regression analysis between alpha diversity of trophic groups, functions, and biodegradation potential}
# Create a data frame for plotting
data <- data.frame(alpha.1 = log(bac.mco.fld$alpha_diversity$Shannon), alpha.2 = log(bdg.mco.fld$alpha_diversity$Shannon), river = bdg.mco.fld$sample_table$River, season = bdg.mco.fld$sample_table$Season)
# Plot the data and add a regression line with R² and p-value annotations
ca <- ggplot(data, aes(x = alpha.1, y = alpha.2, shape = river, color = season)) + geom_point(color = "#4682B4", alpha = 0.5) + scale_shape_manual(values = c(17, 19)) +
      geom_smooth(method = "lm", col = "#4682B4", fill = "#4682B4") + # Linear regression line
      labs(x = "", title = "Prokaryotes", y = "Biodegradation Profile") + 
      theme_classic2() + theme(axis.title.x = element_text(size = 12), legend.position = "none",
                               axis.title.y = element_text(size = 12, angle = 90, vjust = 0.5, hjust = 0.5), 
                               axis.text.x = element_text(size = 9), 
                               axis.text.y = element_text(size = 9, angle = 90, vjust = 0.5, hjust = 0.5), 
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
                               plot.title = element_text(size = 13))
ca

# Create a data frame for plotting
data <- data.frame(alpha.1 = log(keg.mco.fld$alpha_diversity$Shannon), alpha.2 = log(bdg.mco.fld$alpha_diversity$Shannon), river = bdg.mco.fld$sample_table$River, season = bdg.mco.fld$sample_table$Season)
# Plot the data and add a regression line with R² and p-value annotations
cb <- ggplot(data, aes(x = alpha.1, y = alpha.2, shape = river, color = season)) + geom_point(color = "#008B8B", alpha = 0.5) + scale_shape_manual(values = c(17, 19)) +
      geom_smooth(method = "lm", col = "#008B8B", fill = "#008B8B") + # Linear regression line
      labs(x = "", title = "All Predicted Functions", y = "Biodegradation") + 
      theme_classic2() + theme(axis.title.x = element_text(size = 9), axis.text.x = element_text(size = 9),
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
                               plot.title = element_text(size = 13))
cb

# Create a data frame for plotting
data <- data.frame(alpha.1 = log(alg.mco.fld$alpha_diversity$Shannon), alpha.2 = log(bdg.mco.fld$alpha_diversity$Shannon), river = bdg.mco.fld$sample_table$River, season = bdg.mco.fld$sample_table$Season)
# Plot the data and add a regression line with R² and p-value annotations
cc <- ggplot(data, aes(x = alpha.1, y = alpha.2, shape = river, color = season)) + geom_point(color = "#2E8B57", alpha = 0.5) + scale_shape_manual(values = c(17, 19)) +
      geom_smooth(method = "lm", col = "#2E8B57", fill = "#2E8B57") + # Linear regression line
      labs(x = "", title = "Algae", y = "Biodegradation") + 
      theme_classic2() + theme(axis.title.x = element_text(size = 9), axis.text.x = element_text(size = 9),
                               axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(), legend.position = "none",
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
                               plot.title = element_text(size = 13))
cc

# Create a data frame for plotting
data <- data.frame(alpha.1 = log(fun.mco.fld$alpha_diversity$Shannon), alpha.2 = log(bdg.mco.fld$alpha_diversity$Shannon), river = bdg.mco.fld$sample_table$River, season = bdg.mco.fld$sample_table$Season)
# Plot the data and add a regression line with R² and p-value annotations
cd <- ggplot(data, aes(x = alpha.1, y = alpha.2, shape = river, color = season)) + geom_point(color = "#6A5ACD", alpha = 0.5) + scale_shape_manual(values = c(17, 19)) +
      geom_smooth(method = "lm", col = "#6A5ACD", fill = "#6A5ACD") + # Linear regression line
      labs(x = "", title = "Fungi", y = "Biodegradation") + 
      theme_classic2() + theme(axis.title.x = element_text(size = 9), axis.text.x = element_text(size = 9),
                               axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(), legend.position = "none",
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
                               plot.title = element_text(size = 13))
cd

# Create a data frame for plotting
data <- data.frame(alpha.1 = log(pro.mco.fld$alpha_diversity$Shannon), alpha.2 = log(bdg.mco.fld$alpha_diversity$Shannon), river = bdg.mco.fld$sample_table$River, season = bdg.mco.fld$sample_table$Season)
# Plot the data and add a regression line with R² and p-value annotations
ce <- ggplot(data, aes(x = alpha.1, y = alpha.2, shape = river, color = season)) + geom_point(color = "#DAA520", alpha = 0.5) + scale_shape_manual(values = c(17, 19)) +
      geom_smooth(method = "lm", col = "#DAA520", fill = "#DAA520") + # Linear regression line
      labs(x = "", title = "Protists", y = "Biodegradation") + 
      theme_classic2() + theme(axis.title.x = element_text(size = 9), axis.text.x = element_text(size = 9),
                               axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(), legend.position = "none",
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
                               plot.title = element_text(size = 14))
ce

# Create a data frame for plotting
data <- data.frame(alpha.1 = log(met.mco.fld$alpha_diversity$Shannon), alpha.2 = log(bdg.mco.fld$alpha_diversity$Shannon), river = bdg.mco.fld$sample_table$River, season = bdg.mco.fld$sample_table$Season)
# Plot the data and add a regression line with R² and p-value annotations
cf <- ggplot(data, aes(x = alpha.1, y = alpha.2, shape = river, color = season)) + geom_point(color = "#DC143C", alpha = 0.5) + scale_shape_manual(values = c(17, 19)) +
      geom_smooth(method = "lm", col = "#DC143C", fill = "#DC143C") + # Linear regression line
      labs(x = "", title = "Metazoan", y = "Biodegradation") + 
      theme_classic2() + theme(axis.title.x = element_text(size = 9), axis.text.x = element_text(size = 9),
                               axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(), legend.position = "none",
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
                               plot.title = element_text(size = 13))
cf

c4 <- ( ca | cb  | cc | cd | ce | cf )
c4

# Perform correlation test: Prokaryotes
correlation <- cor.test(log(bac.mco.fld$alpha_diversity$Shannon), log(bdg.mco.fld$alpha_diversity$Shannon))
p_value <- signif(correlation$p.value, 4); r_value <- correlation$estimate; r_squared <- r_value^2 # Calculate R-squared
p_value; r_value; r_squared
# Perform correlation test: Functions
correlation <- cor.test(log(keg.mco.fld$alpha_diversity$Shannon), log(bdg.mco.fld$alpha_diversity$Shannon))
p_value <- signif(correlation$p.value, 4); r_value <- correlation$estimate; r_squared <- r_value^2 # Calculate R-squared
p_value; r_value; r_squared
# Perform correlation test: Algae
correlation <- cor.test(log(alg.mco.fld$alpha_diversity$Shannon), log(bdg.mco.fld$alpha_diversity$Shannon))
p_value <- signif(correlation$p.value, 4); r_value <- correlation$estimate; r_squared <- r_value^2 # Calculate R-squared
p_value; r_value; r_squared
# Perform correlation test: Fungi
correlation <- cor.test(log(fun.mco.fld$alpha_diversity$Shannon), log(bdg.mco.fld$alpha_diversity$Shannon))
p_value <- signif(correlation$p.value, 4); r_value <- correlation$estimate; r_squared <- r_value^2 # Calculate R-squared
p_value; r_value; r_squared
# Perform correlation test: Protists
correlation <- cor.test(log(pro.mco.fld$alpha_diversity$Shannon), log(bdg.mco.fld$alpha_diversity$Shannon))
p_value <- signif(correlation$p.value, 4); r_value <- correlation$estimate; r_squared <- r_value^2 # Calculate R-squared
p_value; r_value; r_squared
# Perform correlation test: Metazoan
correlation <- cor.test(log(met.mco.fld$alpha_diversity$Shannon), log(bdg.mco.fld$alpha_diversity$Shannon))
p_value <- signif(correlation$p.value, 4); r_value <- correlation$estimate; r_squared <- r_value^2 # Calculate R-squared
p_value; r_value; r_squared
``` 

### Correlation between beta diversity of trophic groups, functions, and biodegradation potential

```{r, Linear regression analysis between beta diversity of trophic groups, functions, and biodegradation potential}
dataset <- clone(bac.mco.fld)

t1 <- trans_env$new(dataset = dataset, env_cols = c(6:24), complete_na = TRUE) ## env_cols is used to add the environmental data
t2 <- trans_env$new(dataset = bdg.mco.fld, env_cols = c(6:24), complete_na = TRUE) ## env_cols is used to add the environmental data

da <- t1$plot_scatterfit(x = dataset$beta_diversity$bray[rownames(t1$data_env), rownames(t1$data_env)], y = bdg.mco.fld$beta_diversity$bray[rownames(t2$data_env), rownames(t2$data_env)], 
                         type = "lm", point_size = 2, point_alpha = 0, label.x.npc = "left", label.y.npc = "top", line_alpha = 0.25, line_se_color = "#4682B4", lm_equation = FALSE, line_color = "#4682B4",
                         x_axis_title = "Bray-Curtis", y_axis_title = "Biodegradation profile (Bray-Curtis)") + #ylim(0.4,0.9) +
      geom_point(colour = "#4682B4", alpha = 0.05) + theme_classic2() + theme(axis.title.x = element_text(size = 12), legend.position = "bottom",
                                                                             axis.title.y = element_text(size = 12, angle = 90, vjust = 0.5, hjust = 0.5), 
                                                                             axis.text.x = element_text(size = 9), 
                                                                             axis.text.y = element_text(size = 9, angle = 90, vjust = 0.5, hjust = 0.5), 
                                                                             panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) + # Rotate y-axis text
      ggtitle("Prokaryotes")
da

dataset <- clone(keg.mco.fld)

t1 <- trans_env$new(dataset = dataset, env_cols = c(6:24), complete_na = TRUE) ## env_cols is used to add the environmental data
t2 <- trans_env$new(dataset = bdg.mco.fld, env_cols = c(6:24), complete_na = TRUE) ## env_cols is used to add the environmental data

db <- t1$plot_scatterfit(x = dataset$beta_diversity$jaccard[rownames(t1$data_env), rownames(t1$data_env)], y = bdg.mco.fld$beta_diversity$bray[rownames(t2$data_env), rownames(t2$data_env)], 
                         type = "lm", point_size = 2, point_alpha = 0, label.x.npc = "left", label.y.npc = "top", line_alpha = 0.25, line_se_color = "#008B8B", lm_equation = FALSE, line_color = "#008B8B",
                         x_axis_title = "Jaccard", y_axis_title = "") + 
      geom_point(colour = "#008B8B", alpha = 0.05) + theme_classic2() + theme(axis.title.x = element_text(size = 12), axis.text.x = element_text(size = 9),
                               axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(), legend.position = "none",
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) + # Rotate y-axis text
      ggtitle("All Predicted Functions")
db

dataset <- clone(alg.mco.fld)

t1 <- trans_env$new(dataset = dataset, env_cols = c(6:24), complete_na = TRUE) ## env_cols is used to add the environmental data
t2 <- trans_env$new(dataset = bdg.mco.fld, env_cols = c(6:24), complete_na = TRUE) ## env_cols is used to add the environmental data

dc <- t1$plot_scatterfit(x = dataset$beta_diversity$bray[rownames(t1$data_env), rownames(t1$data_env)], y = bdg.mco.fld$beta_diversity$bray[rownames(t2$data_env), rownames(t2$data_env)], 
                         type = "lm", point_size = 2, point_alpha = 0, label.x.npc = "left", label.y.npc = "top", line_alpha = 0.25, line_se_color = "#2E8B57", lm_equation = FALSE, line_color = "#2E8B57",
                         x_axis_title = "Bray-Curtis", y_axis_title = "") + 
      geom_point(colour = "#2E8B57", alpha = 0.05) + theme_classic2() + theme(axis.title.x = element_text(size = 12), axis.text.x = element_text(size = 9),
                               axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(), legend.position = "none",
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) + # Rotate y-axis text
      ggtitle("Algae")
dc

dataset <- clone(fun.mco.fld)

t1 <- trans_env$new(dataset = dataset, env_cols = c(6:24), complete_na = TRUE) ## env_cols is used to add the environmental data
t2 <- trans_env$new(dataset = bdg.mco.fld, env_cols = c(6:24), complete_na = TRUE) ## env_cols is used to add the environmental data

dd <- t1$plot_scatterfit(x = dataset$beta_diversity$bray[rownames(t1$data_env), rownames(t1$data_env)], y = bdg.mco.fld$beta_diversity$bray[rownames(t2$data_env), rownames(t2$data_env)], 
                         type = "lm", point_size = 2, point_alpha = 0, label.x.npc = "left", label.y.npc = "top", line_alpha = 0.25, line_se_color = "#6A5ACD", lm_equation = FALSE, line_color = "#6A5ACD",
                         x_axis_title = "Bray-Curtis", y_axis_title = "Biodegradation profile (Bray-Curtis)") + 
      geom_point(colour = "#6A5ACD", alpha = 0.05) + theme_classic2() + theme(axis.title.x = element_text(size = 12), legend.position = "bottom",
                                                                             axis.title.y = element_text(size = 12, angle = 90, vjust = 0.5, hjust = 0.5), 
                                                                             axis.text.x = element_text(size = 9), 
                                                                             axis.text.y = element_text(size = 9, angle = 90, vjust = 0.5, hjust = 0.5), 
                                                                             panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) + # Rotate y-axis text
      ggtitle("Fungi")
dd

dataset <- clone(pro.mco.fld)

t1 <- trans_env$new(dataset = dataset, env_cols = c(6:24), complete_na = TRUE) ## env_cols is used to add the environmental data
t2 <- trans_env$new(dataset = bdg.mco.fld, env_cols = c(6:24), complete_na = TRUE) ## env_cols is used to add the environmental data

de <- t1$plot_scatterfit(x = dataset$beta_diversity$bray[rownames(t1$data_env), rownames(t1$data_env)], y = bdg.mco.fld$beta_diversity$bray[rownames(t2$data_env), rownames(t2$data_env)], 
                         type = "lm", point_size = 2, point_alpha = 0, label.x.npc = "left", label.y.npc = "top", line_alpha = 0.25, line_se_color = "#DAA520", lm_equation = FALSE, line_color = "#DAA520",
                         x_axis_title = "Bray-Curtis", y_axis_title = "") + 
      geom_point(colour = "#DAA520", alpha = 0.05) + theme_classic2() + theme(axis.title.x = element_text(size = 12), axis.text.x = element_text(size = 9),
                               axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(), legend.position = "none",
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) + # Rotate y-axis text
      ggtitle("Protists")
de

dataset <- clone(met.mco.fld)

t1 <- trans_env$new(dataset = dataset, env_cols = c(6:24), complete_na = TRUE) ## env_cols is used to add the environmental data
t2 <- trans_env$new(dataset = bdg.mco.fld, env_cols = c(6:24), complete_na = TRUE) ## env_cols is used to add the environmental data

df <- t1$plot_scatterfit(x = dataset$beta_diversity$bray[rownames(t1$data_env), rownames(t1$data_env)], y = bdg.mco.fld$beta_diversity$bray[rownames(t2$data_env), rownames(t2$data_env)], 
                         type = "lm", point_size = 2, point_alpha = 0, label.x.npc = "left", label.y.npc = "top", line_alpha = 0.25, line_se_color = "#DC143C", lm_equation = FALSE, line_color = "#DC143C",
                         x_axis_title = "Bray-Curtis", y_axis_title = "") + 
      geom_point(colour = "#DC143C", alpha = 0.05) + theme_classic2() + theme(axis.title.x = element_text(size = 12), axis.text.x = element_text(size = 9),
                               axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(), legend.position = "none",
                               panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) + # Rotate y-axis text
      ggtitle("Metazoan")
df

d4 <- ( da | db  | dc ) / ( dd | de | df )
d4
```

### Correlation with biodegradation profile and the environmnetal variables

```{r RDA}
dataset <- clone(bdg.mco.fld)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# Initialize trans_env object and add environmental data
t1 <- trans_env$new(dataset = dataset, add_data = dataset$sample_table[, c(6:24)])

t1$cal_ordination(method = "RDA", taxa_level = "Superclass")
# select 10 features and adjust the arrow length
t1$trans_ordination(show_taxa = 10, adjust_arrow_length = TRUE, max_perc_env = 1, max_perc_tax = 1, min_perc_env = 1, min_perc_tax = 1)
# t1$res_rda_trans is the transformed result for plot
d1 <- t1$plot_ordination(plot_color = "Season", 
                         plot_type = c("point"), 
                         plot_shape = "River.Reach", 
                         shape_values = c(17, 2, 19, 1), 
                         color_values = c("#1B9E77", "#D95F02", "#7570B3", "#E7298A")) + theme_bw() + 
      theme(panel.grid = element_blank(), legend.position = "right", 
            legend.title = element_text(face = "bold"),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) #+ ggtitle("Biodegradation")
d1

t1$cal_mantel(use_measure = "bray"); t1$res_mantel 
```

```{r}
dataset <- clone(bdg.mco.fld)

ps <- file2meco::meco2phyloseq(dataset)

# Load required libraries
library(phyloseq)
library(vegan)
library(ggplot2)

# Assume your phyloseq object is named 'ps'
# Step 1: Extract OTU table and metadata
otu <- as.data.frame(otu_table(ps))
if (taxa_are_rows(ps)) otu <- t(otu)
meta <- as.data.frame(sample_data(ps))

# Step 2: Hellinger transformation
otu_hell <- decostand(otu, method = "hellinger")

# Step 3: Run RDA
rda_model <- rda(otu_hell ~ ., data = dataset$sample_table[, c(6:24)])

# Step 4: Extract sample scores
scores_samples <- scores(rda_model, display = "sites")

# Step 5: Mahalanobis distance
mahal_dist <- mahalanobis(scores_samples, colMeans(scores_samples), cov(scores_samples))
threshold <- qchisq(0.99, df = ncol(scores_samples))
outliers <- which(mahal_dist > threshold)

# Step 6: Plot with outliers highlighted
scores_df <- as.data.frame(scores_samples)
scores_df$outlier <- FALSE
scores_df$outlier[outliers] <- TRUE

ggplot(scores_df, aes(x = RDA1, y = RDA2, color = outlier)) +
  geom_point(size = 3) +
  scale_color_manual(values = c("black", "red")) +
  theme_minimal() +
  labs(title = "RDA Outlier Detection", color = "Outlier")
```

```{r, dbRDA}
# Clone the dataset and redefine Region levels
dataset <- clone(bdg.mco.fld)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# Initialize trans_env object and add environmental data
t1 <- trans_env$new(dataset = dataset, add_data = dataset$sample_table[, c(6:24)])

# Perform dbRDA analysis using Bray-Curtis distance
t1$cal_ordination(method = "dbRDA", use_measure = "bray")

# Transform ordination results for visualization
t1$trans_ordination(adjust_arrow_length = TRUE, max_perc_env = 1.5)

# Plot ordination with environmental variables and color by Region
e1 <- t1$plot_ordination(plot_color = "Season", plot_type = c("point"), 
                         plot_shape = "River.Reach", 
                         shape_values = c(17, 2, 19, 1),
                         color_values = c("#1B9E77", "#D95F02", "#7570B3", "#E7298A")) + theme_bw() + 
      theme(panel.grid = element_blank(), legend.position = "none", legend.title = element_text(face = "bold"),
            panel.border = element_rect(colour = "black", fill = NA, linewidth = 1)) + ggtitle("Biodegradation")
e1

#From microeco v0.14.0, the function cal_ordination_anova is implemented to check the significance of the ordination model instead of the encapsulation in cal_ordination. Furthermore, the function cal_ordination_envfit can be used to get the contribution of each variables to the model.
t1$cal_ordination_anova(); t1$res_ordination_axis

# Perform envfit analysis
t1$cal_ordination_envfit(); t1$res_ordination_envfit
```

```{r, plot db-RDA R2}
# Load libraries
library(ggplot2)
library(dplyr)
library(tidyr)
library(pheatmap)

# Define original order
factor_levels <- c("Ca","Cl","Cr","Cu","Dissolved oxygen","Electrical cond.","Flow rate",
                   "Mg","NH4-N","Ni","NO2+NO3-N","Pb","pH","TOC","Total N","Total P","TSS",
                   "Water temperature","Zn")
group_levels <- c("Biodegradation", "Prokaryotes", "Algae", "Fungi", "Protists", "Metazoan")

# R² values
r2_data <- data.frame(row.names = factor_levels,
  Biodegradation = c(0.0663,0.41,0.2156,0.0374,0.2553,0.4193,0.1301,0.07,0.4043,0.163,0.294,0.3764,0.7302,0.0542,0.3495,0.2406,0.1141,0.0801,0.1441),
  Prokaryotes = c(0.0131,0.3853,0.1026,0.0224,0.0103,0.1835,0.0978,0.0068,0.3429,0.0048,0.6537,0.2923,0.1956,0.0239,0.5799,0.1473,0.6169,0.0731,0.0139),
  Algae = c(0.4748,0.1415,0.3708,0.4134,0.355,0.1724,0.3203,0.4509,0.1819,0.5246,0.0635,0.4125,0.1025,0.5249,0.1664,0.3879,0.144,0.3865,0.0498),
  Fungi = c(0.3611,0.1617,0.0519,0.2982,0.3047,0.0546,0.3855,0.3536,0.356,0.2455,0.2037,0.3269,0.031,0.2467,0.2863,0.323,0.4883,0.0863,0.0987),
  Protists = c(0.5513,0.2197,0.0888,0.1248,0.3745,0.1928,0.443,0.5473,0.1889,0.3319,0.3026,0.4034,0.1528,0.3223,0.3287,0.4998,0.2166,0.1365,0.0666),
  Metazoan = c(0.0565,0.0667,0.2785,0.083,0.1036,0.1185,0.1047,0.054,0.0059,0.0501,0.0138,0.1001,0.0273,0.0809,0.0064,0.0071,0.1021,0.3258,0.0099))

# Significance symbols
significance_data <- data.frame(row.names = factor_levels,
  Biodegradation = c("","***","**","","**","***","*","","***","*","**","***","***","","***","**","*","","*"),
  Prokaryotes = c("","***","","","","**","","","***","","***","**","**","","***","*","***","",""),
  Algae = c("***","*","***","**","***","*","***","***","**","***","","***","","***","*","***","*","***",""),
  Fungi = c("***","*","","***","***","","***","***","***","**","*","***","","**","***","***","**","",""),
  Protists = c("***","**","","*","***","**","***","***","*","***","***","***","*","***","***","***","**","*",""),
  Metazoan = c("","","***","","","","","","","","","","","","","","","***",""))

# Convert significance to matrix of labels
labels_matrix <- as.matrix(significance_data)

# Create heatmap with clustering on rows (environmental factors)
p1 <- pheatmap(as.matrix(r2_data), cutree_col = 1, cutree_rows = 1,
               cluster_rows = T, 
               cluster_cols = T,
               display_numbers = labels_matrix, number_color = "black",
               fontsize_number = 10,
               fontsize_row = 8, 
               fontsize_col = 10, 
               border_color = NA, clustering_method = "ward.D2",
               color = colorRampPalette(c("white", "gray50"))(10),
               angle_col = 45)
p1
```

```{r, random forest (NOT USED)}
# Use trans_classifier class (https://chiliubio.github.io/microeco_tutorial/model-based-class.html#trans_classifier-class) to perform machine learning to find important environmental variables in classifying groups.

dataset <- clone(bdg.mco.fld)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# Initialize trans_env object and add environmental data
t1 <- trans_env$new(dataset = dataset, add_data = dataset$sample_table[, c(6:24)])
tmp <- t1$data_env %>% t %>% as.data.frame

# caret package should be installed
t2 <- trans_classifier$new(dataset = dataset, x.predictors = tmp, y.response = "pH")
t2$cal_preProcess(method = c("center", "scale", "nzv"))

# All samples are used in training if cal_split function is not performed
t2$cal_train(method = "rf") # default method in caret package without significance
t2$cal_feature_imp(rf_feature_sig = TRUE, num.rep = 1000); t2$res_feature_imp

# Rename the row name
rownames(t2$res_feature_imp)[rownames(t2$res_feature_imp) == "Dissolved.O2"] <- "Dissolved Oygen"
rownames(t2$res_feature_imp)[rownames(t2$res_feature_imp) == "Electrical.conductivity"] <- "Electrical conductivity"
rownames(t2$res_feature_imp)[rownames(t2$res_feature_imp) == "Water.temperature"] <- "Water temperature"
rownames(t2$res_feature_imp)[rownames(t2$res_feature_imp) == "Flow.rate"] <- "Flow rate"
rownames(t2$res_feature_imp)[rownames(t2$res_feature_imp) == "NH4.N"] <- "NH4-N"
rownames(t2$res_feature_imp)[rownames(t2$res_feature_imp) == "NO2.NO3.N"] <- "NO2-NO3-N"
rownames(t2$res_feature_imp)[rownames(t2$res_feature_imp) == "NO2.NO3.N"] <- "Total N"
rownames(t2$res_feature_imp)[rownames(t2$res_feature_imp) == "Total.P"] <- "Total P"

r1 <- t2$plot_feature_imp(coord_flip = TRUE, fill = "gray", width = 0.6, add_sig = TRUE) + theme_classic2() +
      labs(y = "Mean Decrease Accuracy (%IncMSE)") +
      theme(axis.text.x = element_text(angle = 05, hjust = 1, size = 10), 
            panel.grid = element_blank())

r1
```

### Correlation with functional profile and the environmnetal variables

```{r, Correlation heatmap of functions}
library("microViz")

# Clone the dataset and redefine Region levels
dataset <- clone(keg.mco.xen)
dataset$sample_table$TSS <- dataset$sample_table$Turbidity

#dataset$tax_table$Phylum <- gsub("p__", "", dataset$tax_table$Phylum)

psq <- file2meco::meco2phyloseq(dataset)
psq %>%  tax_transform("log", rank = "Level.3") %>% 
         cor_heatmap(taxa_side = "left",
                     vars = c("pH","Dissolved.O2","Electrical.conductivity","Water.temperature",
                              "Flow.rate","Ca","Cl","Cr","Cu","Mg","NH4.N","NO2.NO3.N","Ni","Pb",
                              "TSS","TOC","Total.N","Total.P","Zn"), 
                     cor = "spearman",
                     numbers = NULL, border = T,
                     #var_anno = varAnnotation(log10p = anno_var_box(function(x) log10(x + 1))),
                     #colors = heat_palette("Green-Orange", rev = TRUE, sym = TRUE))
                     colors = heat_palette(c("#F0E68C", "white", "#9370DB"), rev = TRUE, sym = TRUE))

# compute correlations, with p values, and store in a dataframe
correlations_df <- psq %>% 
  tax_model(trans = "log", 
            rank = "Level.3",
            variables = list("pH","Dissolved.O2","Electrical.conductivity","Water.temperature",
                             "Flow.rate","Ca","Cl","Cr","Cu","Mg","NH4.N","NO2.NO3.N","Ni","Pb",
                             "TSS","TOC","Total.N","Total.P","Zn"),
            type = microViz::cor_test, method = "spearman",
            return_psx = FALSE, verbose = FALSE) %>% tax_models2stats(rank = "Level.3")

# get nice looking ordering of correlation estimates using hclust
taxa_hclust <- correlations_df %>%  
               dplyr::select(term, taxon, estimate) %>% 
               tidyr::pivot_wider(id_cols = taxon, names_from = term, values_from = estimate) %>% 
               tibble::column_to_rownames("taxon") %>% 
               as.matrix() %>% stats::dist(method = "euclidean") %>% 
               hclust(method = "ward.D2") 

taxa_order <- taxa_hclust$labels[taxa_hclust$order]

h1 <- correlations_df %>% mutate(p.FDR = p.adjust(p.value, method = "fdr")) %>% ggplot(aes(x = term, y = taxon)) +
      geom_raster(aes(fill = estimate)) +
      geom_point(data = function(x) filter(x, p.value < 0.05), shape = "asterisk") +
      geom_point(data = function(x) filter(x, p.FDR < 0.05), shape = "circle", size = 2) +
      scale_y_discrete(limits = taxa_order) +
      scale_fill_distiller(palette = "PiYG", limits = c(-1, 1)) + theme_bw() + 
      theme(axis.text.x = element_text(size = 9, angle = 45, hjust = 1),
            axis.text.y = element_text(size = 8), axis.ticks = element_blank(),
            legend.position = "left", legend.title = element_text(face = "bold", size = 8), 
            legend.text = element_text(size = 7), panel.border = element_blank(), panel.grid = element_blank()) + 
      labs(x = NULL, y = NULL, fill = "Spearman's\nRank\nCorrelation")

h1
 
library(ggdendro)

# Convert hclust to dendrogram and then to ggplot-compatible format
dendro_data <- ggdendro::dendro_data(as.dendrogram(taxa_hclust), type = "rectangle")

# Plot dendrogram
dendro_plot <- ggplot(segment(dendro_data)) + geom_segment(aes(x = y, y = x, xend = yend, yend = xend)) +
               scale_y_discrete(limits = taxa_order) + theme_bw() +
               theme(axis.text = element_blank(), axis.title = element_blank(), axis.ticks = element_blank(),
                     panel.border = element_blank(), panel.grid = element_blank())

# Combine with heatmap using patchwork
library(patchwork)
combined_plot <- h1 + dendro_plot + plot_layout(ncol = 2, widths = c(5, 1))
combined_plot
```

### Correlation biodegradation profile and the seasonal prokartotes (NOT USED)

```{r}
dataset <- clone(bac.mco.ssn)
dataset$sample_table$Season %<>% factor(., levels = c("Summer", "Autumn", "Winter", "Spring"))

# Initialize trans_env object and add environmental data
t1 <- trans_env$new(dataset = dataset, add_data = t(bdg.mco.fld$otu_table))

# use pH and bray-curtis distance
# add correlation statistics
t1$plot_scatterfit(x = "Sulfamethoxazole",
                   y = dataset$beta_diversity$bray[rownames(t1$data_env), rownames(t1$data_env)], 
                   type = "lm", group = "Season", 
                   point_size = 3, point_alpha = 0.1, 
                   label.x.npc = "center", label.y.npc = "bottom", 
                   x_axis_title = "Euclidean distance", 
                   y_axis_title = "Bray-Curtis distance") + theme_classic2()
```

### Correlation with prokaryotes and the environmnetal variables

```{r, Correlation heatmap of functions}
library("microViz")

# Clone the dataset and redefine Region levels
dataset <- clone(bac.mco.fld)

dataset$tax_table$Phylum <- gsub("p__", "", dataset$tax_table$Phylum)

psq <- file2meco::meco2phyloseq(dataset)
psq %>%  tax_transform("log", rank = "Phylum") %>% 
         cor_heatmap(taxa_side = "left",
                     vars = c("pH","Dissolved.O2","Electrical.conductivity","Water.temperature",
                              "Flow.rate","Ca","Cl","Cr","Cu","Mg","NH4.N","NO2.NO3.N","Ni","Pb",
                              "TSS","TOC","Total.N","Total.P","Zn"), 
                     cor = "spearman",
                     numbers = NULL, border = T,
                     #var_anno = varAnnotation(log10p = anno_var_box(function(x) log10(x + 1))),
                     #colors = heat_palette("Green-Orange", rev = TRUE, sym = TRUE))
                     colors = heat_palette(c("#F0E68C", "white", "#9370DB"), rev = TRUE, sym = TRUE))

# compute correlations, with p values, and store in a dataframe
correlations_df <- psq %>% 
  tax_model(trans = "log", 
            rank = "Phylum",
            variables = list("pH","Dissolved.O2","Electrical.conductivity","Water.temperature",
                             "Flow.rate","Ca","Cl","Cr","Cu","Mg","NH4.N","NO2.NO3.N","Ni","Pb",
                             "TSS","TOC","Total.N","Total.P","Zn"),
            type = microViz::cor_test, method = "spearman",
            return_psx = FALSE, verbose = FALSE) %>% tax_models2stats(rank = "Level.3")

# get nice looking ordering of correlation estimates using hclust
taxa_hclust <- correlations_df %>%  
               dplyr::select(term, taxon, estimate) %>% 
               tidyr::pivot_wider(id_cols = taxon, names_from = term, values_from = estimate) %>% 
               tibble::column_to_rownames("taxon") %>% 
               as.matrix() %>% stats::dist(method = "euclidean") %>% 
               hclust(method = "ward.D2") 
 
taxa_order <- taxa_hclust$labels[taxa_hclust$order]

h1 <- correlations_df %>% mutate(p.FDR = p.adjust(p.value, method = "fdr")) %>% ggplot(aes(x = term, y = taxon)) +
      geom_raster(aes(fill = estimate)) +
      geom_point(data = function(x) filter(x, p.value < 0.05), shape = "asterisk") +
      geom_point(data = function(x) filter(x, p.FDR < 0.05), shape = "circle", size = 2) +
      scale_y_discrete(limits = taxa_order) +
      scale_fill_distiller(palette = "PuOr", limits = c(-1, 1)) + theme_bw() + 
      theme(axis.text.x = element_text(size = 9, angle = 45, hjust = 1),
            axis.text.y = element_text(size = 8), axis.ticks = element_blank(),
            legend.position = "left", legend.title = element_text(face = "bold", size = 8), 
            legend.text = element_text(size = 7), panel.border = element_blank(), panel.grid = element_blank()) + 
      labs(x = NULL, y = NULL, fill = "Spearman's\nRank\nCorrelation")

h1

library(ggdendro)

# Convert hclust to dendrogram and then to ggplot-compatible format
dendro_data <- ggdendro::dendro_data(as.dendrogram(taxa_hclust), type = "rectangle")

# Plot dendrogram
dendro_plot <- ggplot(segment(dendro_data)) + geom_segment(aes(x = y, y = x, xend = yend, yend = xend)) +
               scale_y_discrete(limits = taxa_order) + theme_bw() +
               theme(axis.text = element_blank(), axis.title = element_blank(), axis.ticks = element_blank(),
                     panel.border = element_blank(), panel.grid = element_blank())

# Combine with heatmap using patchwork
library(patchwork)
combined_plot <- h1 + dendro_plot + plot_layout(ncol = 2, widths = c(5, 1))
combined_plot
```


### Multiple co-inertia analysis between trophic groups and biodegradation profile

```{r}
library("omicade4")

## Genus-level analysis
bac.mco.fld.gns <- bac.mco.fld$merge_taxa("Genus")
alg.mco.fld.gns <- alg.mco.fld$merge_taxa("Genus")
fun.mco.fld.gns <- fun.mco.fld$merge_taxa("Genus")
pro.mco.fld.gns <- pro.mco.fld$merge_taxa("Genus")
met.mco.fld.gns <- met.mco.fld$merge_taxa("Genus")
keg.mco.fld.gns <- keg.mco.fld
bdg.mco.fld.gns <- bdg.mco.fld

bac.psq.fld.gns <- file2meco::meco2phyloseq(bac.mco.fld.gns)
alg.psq.fld.gns <- file2meco::meco2phyloseq(alg.mco.fld.gns)
fun.psq.fld.gns <- file2meco::meco2phyloseq(fun.mco.fld.gns)
pro.psq.fld.gns <- file2meco::meco2phyloseq(pro.mco.fld.gns)
met.psq.fld.gns <- file2meco::meco2phyloseq(met.mco.fld.gns)
keg.psq.fld.gns <- file2meco::meco2phyloseq(keg.mco.fld.gns)
bdg.psq.fld.gns <- file2meco::meco2phyloseq(bdg.mco.fld.gns)

## Create dataframe from otu table

bac.cia <- as.data.frame(otu_table(bac.psq.fld.gns)); bac.cia <- labdsv::hellinger(bac.cia); dim(bac.cia)
alg.cia <- as.data.frame(otu_table(alg.psq.fld.gns)); alg.cia <- labdsv::hellinger(alg.cia); dim(alg.cia)
fun.cia <- as.data.frame(otu_table(fun.psq.fld.gns)); fun.cia <- labdsv::hellinger(fun.cia); dim(fun.cia)
pro.cia <- as.data.frame(otu_table(pro.psq.fld.gns)); pro.cia <- labdsv::hellinger(pro.cia); dim(pro.cia)
met.cia <- as.data.frame(otu_table(met.psq.fld.gns)); met.cia <- labdsv::hellinger(met.cia); dim(met.cia)
keg.cia <- as.data.frame(otu_table(keg.psq.fld.gns)); keg.cia <- labdsv::hellinger(keg.cia); dim(keg.cia)
bdg.cia <- as.data.frame(otu_table(bdg.psq.fld.gns)); bdg.cia <- labdsv::hellinger(bdg.cia); dim(bdg.cia)

library("omicade4")
sp.ct <- list(bac.cia, alg.cia, fun.cia, pro.cia, met.cia, keg.cia, bdg.cia) ## Merge datasets into one list
sapply(sp.ct, dim) ## Check the dimension of each dataset in the list

mcoin <- mcia(sp.ct, cia.nf = 10); class(mcoin)

mcoin$coa$Prokaryotes <- mcoin$coa$df1; mcoin$coa$df1 <- NULL
mcoin$coa$Algae <- mcoin$coa$df2; mcoin$coa$df2 <- NULL
mcoin$coa$Fungi <- mcoin$coa$df3; mcoin$coa$df3 <- NULL
mcoin$coa$Protists <- mcoin$coa$df4; mcoin$coa$df4 <- NULL
mcoin$coa$Metazoan <- mcoin$coa$df5; mcoin$coa$df5 <- NULL
mcoin$coa$Functions <- mcoin$coa$df6; mcoin$coa$df6 <- NULL
mcoin$coa$Biodegradation <- mcoin$coa$df7; mcoin$coa$df7 <- NULL

type <- t(bac.mco.fld$sample_table$Season); type

plot.mcia(mcoin, sample.lab = FALSE, phenovec = type, df.col = c("#4682B4", "#2E8B57", "#6A5ACD", "#DAA520", 
                                                                 "#DC143C", "#008B8B", "#A9A9A9"))
#plotVar(mcoin, var = NA, bg.var.col = 1,  var.lab=TRUE)

mcoin$mcoa$RV
```

```{r}
# RV coefficient
similarity_matrix <- matrix(c(1,0.3811052,0.6198456,0.4119582,0.5452583,0.65451088,0.32763949,
                              0.3811052,1,0.4861036,0.9144658,0.8504269,0.11492791,0.33155066,
                              0.6198456,0.4861036,1,0.4430552,0.4942627,0.35560293,0.32327972,
                              0.4119582,0.9144658,0.4430552,1,0.7871411,0.16030802,0.33144555,
                              0.5452583,0.8504269,0.4942627,0.7871411,1,0.24456503,0.39354957,
                              0.6545109,0.1149279,0.3556029,0.160308,0.244565,1,0.09755501,
                              0.3276395,0.3315507,0.3232797,0.3314456,0.3935496,0.09755501,1), nrow = 7, byrow = TRUE)

# Assign row and column names
rownames(similarity_matrix) <- colnames(similarity_matrix) <- c("Prokaryotes", "Algae", "Fungi", "Protists", 
                                                                "Metazoan", "Functions", "Biodegradation")

# Convert to data matrix object
rv.matrix <- as.matrix(similarity_matrix)

# Convert the matrix to a data frame
df <- melt(rv.matrix)

# Load required libraries
library(ggplot2)
library(reshape2)

# Convert matrix to long-format data frame
df <- melt(similarity_matrix) 

# Plot heatmap
c1 <- ggplot(df, aes(x = Var1, y = Var2, fill = value)) +
      geom_tile(color = "white") + 
      scale_fill_gradient2(low = "white", , mid = "#DCDCDC", high = "#696969",
                           midpoint = 0.75, limit = c(0, 1), space = "Lab", 
                           name = "RV coefficient") +
      geom_text(aes(label = round(value, 2)), size = 2.5) + theme_minimal() +
      theme(axis.title = element_blank(), legend.position = "top", legend.title = element_text(face = "bold"),
            axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
            panel.border = element_rect(colour = "white", fill = NA, linewidth = 1))
c1
```

### Procrustes and Mantel test between trophic groups and biodegradation profile

```{r}
library("vegan"); library("tidyr")

# Extract beta diversity matrices for each dataset
bac.bc <- bac.mco.fld$beta_diversity$wei_unifrac
alg.bc <- alg.mco.fld$beta_diversity$wei_unifrac
fun.bc <- fun.mco.fld$beta_diversity$wei_unifrac
pro.bc <- pro.mco.fld$beta_diversity$wei_unifrac
met.bc <- met.mco.fld$beta_diversity$wei_unifrac
keg.bc <- keg.mco.fld$beta_diversity$jaccard
bdg.bc <- bdg.mco.fld$beta_diversity$bray

# Hellinger transformation for PCA 
bac.bc.hel <- decostand(bac.bc, method = "hellinger"); bac.bc.hel.pca <- rda(bac.bc.hel)
alg.bc.hel <- decostand(alg.bc, method = "hellinger"); alg.bc.hel.pca <- rda(alg.bc.hel)
fun.bc.hel <- decostand(fun.bc, method = "hellinger"); fun.bc.hel.pca <- rda(fun.bc.hel)
pro.bc.hel <- decostand(pro.bc, method = "hellinger"); pro.bc.hel.pca <- rda(pro.bc.hel)
met.bc.hel <- decostand(met.bc, method = "hellinger"); met.bc.hel.pca <- rda(met.bc.hel)
keg.bc.hel <- decostand(keg.bc, method = "hellinger"); keg.bc.hel.pca <- rda(keg.bc.hel) 
bdg.bc.hel <- decostand(bdg.bc, method = "hellinger"); bdg.bc.hel.pca <- rda(bdg.bc.hel)

# Test between bdg and bac
bdg.bac.pro <- procrustes(X = bdg.bc.hel.pca, Y = bac.bc.hel.pca, symmetric = TRUE); summary(bdg.bac.pro)
bdg.bac.prt <- protest(X = bdg.bc.hel.pca, Y = bac.bc.hel.pca, scores = "sites", permutations = 999); bdg.bac.prt
plot(bdg.bac.pro, kind = 1, main = expression("Prokaryotes vs. Biodegradation profile"), ar.col = "gray85")
lines(bdg.bac.pro, type = "segments", col = "gray", lwd = 2, lty = 3)
points(bdg.bac.pro, display = c("target"), col = "black", cex = 1, pch = 21, bg = "#A9A9A9")
points(bdg.bac.pro, display = c("rotated"), col = "black", cex = 1, pch = 21, bg = "#4682B4")

# Test between bdg and alg
bdg.alg.pro <- procrustes(X = bdg.bc.hel.pca, Y = alg.bc.hel.pca, symmetric = TRUE)
summary(bdg.alg.pro)
bdg.alg.prt <- protest(X = bdg.bc.hel.pca, Y = alg.bc.hel.pca, scores = "sites", permutations = 999)
bdg.alg.prt
plot(bdg.alg.pro, kind = 1, main = expression("Algae vs. Biodegradation profile"), ar.col = "gray85")
lines(bdg.alg.pro, type = "segments", col = "gray", lwd = 2, lty = 3)
points(bdg.alg.pro, display = c("target"), col = "black", cex = 1, pch = 21, bg = "#A9A9A9")
points(bdg.alg.pro, display = c("rotated"), col = "black", cex = 1, pch = 21, bg = "#2E8B57")

# Test between bdg and fun
bdg.fun.pro <- procrustes(X = bdg.bc.hel.pca, Y = fun.bc.hel.pca, symmetric = TRUE)
summary(bdg.fun.pro)
bdg.fun.prt <- protest(X = bdg.bc.hel.pca, Y = fun.bc.hel.pca, scores = "sites", permutations = 999)
bdg.fun.prt
plot(bdg.fun.pro, kind = 1, main = expression("Fungi vs. Biodegradation profile"), ar.col = "gray85")
lines(bdg.fun.pro, type = "segments", col = "gray", lwd = 2, lty = 3)
points(bdg.fun.pro, display = c("target"), col = "black", cex = 1, pch = 21, bg = "#A9A9A9")
points(bdg.fun.pro, display = c("rotated"), col = "black", cex = 1, pch = 21, bg = "#6A5ACD")

# Test between bdg and pro
bdg.pro.pro <- procrustes(X = bdg.bc.hel.pca, Y = pro.bc.hel.pca, symmetric = TRUE)
summary(bdg.pro.pro)
bdg.pro.prt <- protest(X = bdg.bc.hel.pca, Y = pro.bc.hel.pca, scores = "sites", permutations = 999)
bdg.pro.prt
plot(bdg.pro.pro, kind = 1, main = expression("Protists vs. Biodegradation profile"), ar.col = "gray85")
lines(bdg.pro.pro, type = "segments", col = "gray", lwd = 2, lty = 3)
points(bdg.pro.pro, display = c("target"), col = "black", cex = 1, pch = 21, bg = "#A9A9A9")
points(bdg.pro.pro, display = c("rotated"), col = "black", cex = 1, pch = 21, bg = "#DAA520")

# Test between bdg and met
bdg.met.pro <- procrustes(X = bdg.bc.hel.pca, Y = met.bc.hel.pca, symmetric = TRUE)
summary(bdg.met.pro)
bdg.met.prt <- protest(X = bdg.bc.hel.pca, Y = met.bc.hel.pca, scores = "sites", permutations = 999)
bdg.met.prt
plot(bdg.met.pro, kind = 1, main = expression("Metazoan vs. Biodegradation profile"), ar.col = "gray85")
lines(bdg.met.pro, type = "segments", col = "gray", lwd = 2, lty = 3)
points(bdg.met.pro, display = c("target"), col = "black", cex = 1, pch = 21, bg = "#A9A9A9")
points(bdg.met.pro, display = c("rotated"), col = "black", cex = 1, pch = 21, bg = "#DC143C")

# Test between bdg and keg
bdg.keg.pro <- procrustes(X = bdg.bc.hel.pca, Y = keg.bc.hel.pca, symkegric = TRUE)
summary(bdg.keg.pro)
bdg.keg.prt <- protest(X = bdg.bc.hel.pca, Y = keg.bc.hel.pca, scores = "sites", permutations = 999)
bdg.keg.prt
plot(bdg.keg.pro, kind = 1, main = expression("Predicted functions vs. Biodegradation profile"), ar.col = "gray85")
lines(bdg.keg.pro, type = "segments", col = "gray", lwd = 2, lty = 3)
points(bdg.keg.pro, display = c("target"), col = "black", cex = 1, pch = 21, bg = "#A9A9A9")
points(bdg.keg.pro, display = c("rotated"), col = "black", cex = 1, pch = 21, bg = "#008B8B")

### Mantel statistic based on Pearson's product-moment correlation 

mantel(bdg.bc, bac.bc, method = "pearson", permutations = 999)
mantel(bdg.bc, alg.bc, method = "pearson", permutations = 999)
mantel(bdg.bc, fun.bc, method = "pearson", permutations = 999)
mantel(bdg.bc, pro.bc, method = "pearson", permutations = 999)
mantel(bdg.bc, met.bc, method = "pearson", permutations = 999)
mantel(bdg.bc, keg.bc, method = "pearson", permutations = 999)
```


### [NOT PERFORMED!!!] Partial least squares path modeling (PLS-PM) analysis

```{r, PLS-PM with Biodegradation}
# Load necessary library
library("plspm")
library("turner")
library("microbiome")
library("metagMisc")

bac.psq.fld <- file2meco::meco2phyloseq(bac.mco.fld)
alg.psq.fld <- file2meco::meco2phyloseq(alg.mco.fld)
fun.psq.fld <- file2meco::meco2phyloseq(fun.mco.fld)
pro.psq.fld <- file2meco::meco2phyloseq(pro.mco.fld)
met.psq.fld <- file2meco::meco2phyloseq(met.mco.fld)
bdg.psq.fld <- file2meco::meco2phyloseq(bdg.mco.fld)

taxa_names(bac.psq.fld) <- paste0("BAC", seq(ntaxa(bac.psq.fld)))
taxa_names(alg.psq.fld) <- paste0("ALG", seq(ntaxa(alg.psq.fld)))
taxa_names(fun.psq.fld) <- paste0("FUN", seq(ntaxa(fun.psq.fld)))
taxa_names(pro.psq.fld) <- paste0("PRO", seq(ntaxa(pro.psq.fld)))
taxa_names(met.psq.fld) <- paste0("MET", seq(ntaxa(met.psq.fld)))
taxa_names(bdg.psq.fld) <- paste0("BDG", seq(ntaxa(bdg.psq.fld)))

set.seed(123456) # for reproducibility, remove for normal use

# Create dataframe with datasets that only have genes that are found in at least 5% of samples.

#Based on UN-aggregated datasets
bie.df <- cbind(t(as.data.frame(abundances(phyloseq_filter_prevalence(bac.psq.fld, prev.trh = 0.05, abund.trh = NULL)))),
                t(as.data.frame(abundances(phyloseq_filter_prevalence(alg.psq.fld, prev.trh = 0.05, abund.trh = NULL)))),
                t(as.data.frame(abundances(phyloseq_filter_prevalence(fun.psq.fld, prev.trh = 0.05, abund.trh = NULL)))),
                t(as.data.frame(abundances(phyloseq_filter_prevalence(pro.psq.fld, prev.trh = 0.05, abund.trh = NULL)))),
                t(as.data.frame(abundances(phyloseq_filter_prevalence(met.psq.fld, prev.trh = 0.05, abund.trh = NULL)))),
                t(as.data.frame(abundances(phyloseq_filter_prevalence(bdg.psq.fld, prev.trh = 0, abund.trh = NULL)))))

dim(bie.df)

# define path matrix (inner model)
Metazoan <- c(0,0,0,0,0,0)
Protists <- c(1,0,0,0,0,0)
Fungi <- c(1,1,0,0,0,0)
Algae<- c(1,1,1,0,0,0)
Prokaryotes <- c(1,1,1,1,0,0)
Biodegradation <- c(1,1,1,1,1,0)

bie.path <- rbind(Metazoan, Protists, Fungi, Algae, Prokaryotes, Biodegradation)

# define list of blocks (outer model)
bie.blocks <- list(8297:8416,7214:8296,6382:7213,5880:6381,1:5879,8417:8512) # using UN-aggregated dataset.

# vector of modes (reflective indicators)
bie.modes <- rep("A", 6) 

# apply plspm with bootstrap validation
bie.pls <- plspm(bie.df, bie.path, bie.blocks, modes = bie.modes, boot.val = 1000)

# default print
print(bie.pls)

# summary of results
summary(bie.pls)

bie.pls$gof

# plot inner model results
plot(bie.pls, what = "inner",
     colpos = "#4169E1", colneg = "#F08080", arr.width = 0.3,
     box.prop = 1, box.size = 0.08, box.cex = 1, box.col = c("#A9A9A9", "#4682B4", "#2E8B57", "#6A5ACD", "#DAA520", "#DC143C"), lcol = "black",
     txt.col = "black", arr.pos = 0.5, cex.txt = 0.9)

# plot outer model loadings
#plot(bie.pls, what = "loadings")

# plot outer model weights
#plot(baltic.pls, what = "weights")
```

```{r, PLS-PM of the Trophic Groups}
# Load necessary library
library("plspm")
library("turner")
library("microbiome")
library("metagMisc")

bac.psq.fld <- file2meco::meco2phyloseq(bac.mco.fld)
alg.psq.fld <- file2meco::meco2phyloseq(alg.mco.fld)
fun.psq.fld <- file2meco::meco2phyloseq(fun.mco.fld)
pro.psq.fld <- file2meco::meco2phyloseq(pro.mco.fld)
met.psq.fld <- file2meco::meco2phyloseq(met.mco.fld)
bdg.psq.fld <- file2meco::meco2phyloseq(bdg.mco.fld)

taxa_names(bac.psq.fld) <- paste0("BAC", seq(ntaxa(bac.psq.fld)))
taxa_names(alg.psq.fld) <- paste0("ALG", seq(ntaxa(alg.psq.fld)))
taxa_names(fun.psq.fld) <- paste0("FUN", seq(ntaxa(fun.psq.fld)))
taxa_names(pro.psq.fld) <- paste0("PRO", seq(ntaxa(pro.psq.fld)))
taxa_names(met.psq.fld) <- paste0("MET", seq(ntaxa(met.psq.fld)))

set.seed(123456) # for reproducibility, remove for normal use

# Create dataframe with datasets that only have genes that are found in at least 5% of samples.

#Based on UN-aggregated datasets
bie.df <- cbind(t(as.data.frame(abundances(phyloseq_filter_prevalence(bac.psq.fld, prev.trh = 0.05, abund.trh = NULL)))),
                t(as.data.frame(abundances(phyloseq_filter_prevalence(alg.psq.fld, prev.trh = 0.05, abund.trh = NULL)))),
                t(as.data.frame(abundances(phyloseq_filter_prevalence(fun.psq.fld, prev.trh = 0.05, abund.trh = NULL)))),
                t(as.data.frame(abundances(phyloseq_filter_prevalence(pro.psq.fld, prev.trh = 0.05, abund.trh = NULL)))),
                t(as.data.frame(abundances(phyloseq_filter_prevalence(met.psq.fld, prev.trh = 0.05, abund.trh = NULL)))))

dim(bie.df)

# define path matrix (inner model)
Metazoan <- c(0,0,0,0,0)
Protists <- c(1,0,0,0,0)
Fungi <- c(1,1,0,0,0)
Algae<- c(1,1,1,0,0)
Prokaryotes <- c(1,1,1,1,0)

bie.path <- rbind(Metazoan, Protists, Fungi, Algae, Prokaryotes)

# define list of blocks (outer model)
bie.blocks <- list(8297:8416,7214:8296,6382:7213,5880:6381,1:5879) # using UN-aggregated dataset.

# vector of modes (reflective indicators)
bie.modes <- rep("A", 5) 

# apply plspm with bootstrap validation
bie.pls <- plspm(bie.df, bie.path, bie.blocks, modes = bie.modes, boot.val = 1000)

# default print
print(bie.pls)

# summary of results
summary(bie.pls)

bie.pls$gof

# plot inner model results
plot(bie.pls, what = "inner",
     colpos = "#4169E1", colneg = "#F08080", arr.width = 0.3,
     box.prop = 0.6, box.size = 0.1, box.cex = 1, box.col = c("#4682B4", "#2E8B57", "#6A5ACD", "#DAA520", "#DC143C"), lcol = "black",
     txt.col = "black", arr.pos = 0.5, cex.txt = 1)

# plot outer model loadings
#plot(bie.pls, what = "loadings")

# plot outer model weights
#plot(baltic.pls, what = "weights")
```

### Correlation between the environmental variables and the trophic groups

```{r, based on alpha diversity}
dataset <- clone(pro.mco.fld)
dataset$sample_table$River.Reach.Season %<>% factor(., levels = c("Knivstaån Up Winter", "Knivstaån Up Spring",
                                                                  "Knivstaån Up Summer", "Knivstaån Up Autumn", 
                                                                  "Knivstaån Dn Winter", "Knivstaån Dn Spring",
                                                                  "Knivstaån Dn Summer", "Knivstaån Dn Autumn", 
                                                                  "Vitsån Up Winter", "Vitsån Up Spring",
                                                                  "Vitsån Up Summer", "Vitsån Up Autumn", 
                                                                  "Vitsån Dn Winter", "Vitsån Dn Spring",
                                                                  "Vitsån Dn Summer", "Vitsån Dn Autumn"))

## Generate trans_nullmodel object use high abundance OTU with mean relative abundance > 0.001.
t2 <- trans_nullmodel$new(dataset, filter_thres = 0.001, env_cols = c(6:24))

## For nearest Taxon Index (NTI) and nearest Relative Index (NRI), please use cal_NTI and cal_NRI, respectively.
t2$cal_NTI(null.model = "taxa.labels", abundance.weighted = FALSE, runs = 999); t2$res_NTI; dataset$alpha_diversity[["NTI"]] <- t2$res_NTI$NTI ## Add matrix to alpha_diversity list.
t2$cal_NRI(null.model = "taxa.labels", abundance.weighted = FALSE, runs = 999); t2$res_NRI; dataset$alpha_diversity[["NRI"]] <- t2$res_NRI$NRI ## Add matrix to alpha_diversity list.

dataset$alpha_diversity

## Assess the relationship between environmental factors and alpha diversity, please use add_abund_table parameter in the cal_cor function. The correlation result is stored in object$res_cor ...
t3 <- trans_env$new(dataset = dataset, env_cols = c(6:24), complete_na = TRUE) ## env_cols is used to add the environmental data
t3$cal_cor(add_abund_table = dataset$alpha_diversity[c("Chao1", "Shannon", "Simpson", "PD", "NTI", "NRI")], p_adjust_type = "Env")
n1 <- t3$plot_cor(cluster_ggplot = "both", color_vector = c("#4169E1", "white", "#F08080"))
n1
```

```{r, based on beta diversity}
library("microeco"); library("magrittr")

dataset <- clone(bac.mco.fld)
dataset$sample_table[c(25:120)] <- apply(dataset$sample_table[c(25:120)], 2, function(x) ifelse(x < 0, 0, x)) # Transform negative biodegradation rates to 0.

## Calculate the auto-correlations among environmental variables.
t1 <- trans_env$new(dataset = dataset, env_cols = 6:24)
e1 <- quickcor(t1$data_env, type = "lower", cor.test = TRUE, show.diag = TRUE, cluster = TRUE, cluster.method = "mcquitty") +
               geom_circle2() + geom_cross() +
               #scale_fill_gradientn(limit = c(-1,1), colours = c("#A50026", "white", "#053061"))
               scale_fill_gradientn(limit = c(-1,1), colours = c("#4169E1", "white", "#F08080"))

# Define dataframes
d1 <- clone(bac.mco.fld)
d2 <- clone(euk.mco.fld)

# First perform Mantel test
t1 <- trans_env$new(dataset = d1, env_cols = 6:24); t1$cal_mantel(use_measure = "bray", partial_mantel = TRUE)
t2 <- trans_env$new(dataset = d2, env_cols = 6:24); t2$cal_mantel(use_measure = "bray", partial_mantel = TRUE)

# Extract a part of the results 
x1 <- data.frame(spec = "Prokaryotes", t1$res_mantel) %>% .[, c(1, 3, 6, 8)]
x2 <- data.frame(spec = "Eukaryotes", t2$res_mantel) %>% .[, c(1, 3, 6, 8)]

# Rename columns
colnames(x1) <- colnames(x2) <- c("spec", "env", "r", "p.value")

# Generate interval data
x1 %<>% dplyr::mutate(rd = cut(r, breaks = c(-Inf, 0.3, 0.6, Inf), labels = c("< 0.3", "0.3 - 0.6", ">= 0.6")), pd = cut(p.value, breaks = c(-Inf, 0.01, 0.05, Inf), labels = c("< 0.01", "0.01 - 0.05", ">= 0.05")))
x2 %<>% dplyr::mutate(rd = cut(r, breaks = c(-Inf, 0.3, 0.6, Inf), labels = c("< 0.3", "0.3 - 0.6", ">= 0.6")), pd = cut(p.value, breaks = c(-Inf, 0.01, 0.05, Inf), labels = c("< 0.01", "0.01 - 0.05", ">= 0.05")))

# Combine two tables
plot_table <- rbind(x1, x2)

## Require ggcor package (https://github.com/mj163163/ggcor-1)
g1 <- e1 + add_link(plot_table, ## For all tables.
                    #x1, ## For 1 plot for easier visualization.
                    mapping = aes(colour = pd, size = rd), diag.label = FALSE, curvature = 0) + 
                    scale_size_manual(values = c(0.25, 1, 2)) + 
                    scale_colour_manual(values = c("#DC143C", "#1B9E77", "#A2A2A288")) +
                    guides(size = guide_legend(title = "Mantel's R", override.aes = list(colour = "grey35"), order = 2),
                    colour = guide_legend(title = "Mantel's P", override.aes = list(size = 3), order = 1),
                    fill = guide_colorbar(title = "Pearson's R", order = 3))
g1
```

### Mediator analysis with MODIMA

```{r, MODIMA, environment-community-biodegradation}
library("vegan")
library("ade4")
library("energy")
library("permute")
library("matrixStats")
library("phyloseq")
library("tidyverse")
source("src/mediation/modima.R")

# change mediator input according to trophic group.
mediator.df <- as.data.frame(t(bac.mco.fld$otu_table))

exposure.df <- as.data.frame(bdg.mco.fld$sample_table[c(6:24)])
response.df <- as.data.frame(t(bdg.mco.fld$otu_table))

mediator.dist <- vegdist(mediator.df, method = "bray")
exposure.dist <- vegdist(exposure.df, method = "euclidean")
response.dist <- vegdist(response.df, method = "bray")

set.seed(12345)
modima.results <- modima(exposure = exposure.dist, 
                         mediator = mediator.dist, 
                         response = response.dist, 
                         nrep = 9999)

set.seed(12345)
ER_M_list <- pdcor.test(exposure.dist, response.dist, mediator.dist, R = 9999)
MR_E_list <- pdcor.test(mediator.dist, response.dist, exposure.dist, R = 9999)
beta_MR_E <- pdcor(mediator.dist, response.dist, exposure.dist) %>% as.numeric()
beta_ER <- bcdcor(exposure.dist, response.dist) %>% as.numeric()
ER_list <- dcor.test(exposure.dist, response.dist, R = 9999)
beta_EM <- bcdcor(exposure.dist, mediator.dist) %>% as.numeric()
EM_list <- dcor.test(exposure.dist, mediator.dist, R = 9999)

result.summary <- data.frame(Med_statistic = modima.results$statistic %>% as.numeric(),
                             Med_percentage = (beta_MR_E * beta_EM)/beta_ER * 100,
                             Med_P = modima.results$p.value,
                             beta_EM = beta_EM,
                             P_EM = EM_list$p.value,
                             beta_MR_E = beta_MR_E,
                             P_MR_E = MR_E_list$p.value,
                             beta_ER = beta_ER,
                             P_ER = ER_list$p.value,
                             name = "exposure_mediator_response_all")
modima.results
result.summary
#write_tsv(result.summary, "src/mediation/bac.modima.tsv")
```

```{r, MODIMA, environment-selected-kegg-biodegradation}
library("vegan")
library("ade4")
library("energy")
library("permute")
library("matrixStats")
library("phyloseq")
library("tidyverse")
source("src/mediation/modima.R")

# change mediator input according to functional level.
mediator.df <- as.data.frame(t(keg.mco.xen$otu_table))

exposure.df <- as.data.frame(bdg.mco.fld$sample_table[c(6:24)])
response.df <- as.data.frame(t(bdg.mco.fld$otu_table))

mediator.dist <- vegdist(mediator.df, method = "jaccard")
exposure.dist <- vegdist(exposure.df, method = "euclidean")
response.dist <- vegdist(response.df, method = "bray")

set.seed(12345)
modima.results <- modima(exposure = exposure.dist, 
                         mediator = mediator.dist, 
                         response = response.dist, 
                         nrep = 9999)

set.seed(12345)
ER_M_list <- pdcor.test(exposure.dist, response.dist, mediator.dist, R = 9999)
MR_E_list <- pdcor.test(mediator.dist, response.dist, exposure.dist, R = 9999)
beta_MR_E <- pdcor(mediator.dist, response.dist, exposure.dist) %>% as.numeric()
beta_ER <- bcdcor(exposure.dist, response.dist) %>% as.numeric()
ER_list <- dcor.test(exposure.dist, response.dist, R = 9999)
beta_EM <- bcdcor(exposure.dist, mediator.dist) %>% as.numeric()
EM_list <- dcor.test(exposure.dist, mediator.dist, R = 9999)

result.summary <- data.frame(Med_statistic = modima.results$statistic %>% as.numeric(),
                             Med_percentage = (beta_MR_E * beta_EM)/beta_ER * 100,
                             Med_P = modima.results$p.value,
                             beta_EM = beta_EM,
                             P_EM = EM_list$p.value,
                             beta_MR_E = beta_MR_E,
                             P_MR_E = MR_E_list$p.value,
                             beta_ER = beta_ER,
                             P_ER = ER_list$p.value,
                             name = "exposure_mediator_response_all")
modima.results
result.summary
#write_tsv(result.summary, "src/mediation/bac.modima.tsv")
```

```{r, LDM}
library(LDM) 
library(tidyverse)

tmp <- clone(bac.mco.ssn)
tmp$filter_taxa(rel_abund = 0.0001)
tmp

exposure.df <- as.data.frame(bdg.mco.fld$sample_table[c(6:24)])
outcome.df <- as.data.frame(t(bdg.mco.fld$otu_table))
mediator.df <- as.data.frame(t(tmp$otu_table))

colnames(exposure.df) <- sprintf("exp%02d", seq_along(exposure.df))
colnames(outcome.df) <- sprintf("out%03d", seq_along(outcome.df))
colnames(mediator.df) <- sprintf("med%05d", seq_along(mediator.df))

all(rownames(exposure.df) == rownames(outcome.df))  # Should be TRUE
all(rownames(exposure.df) == rownames(mediator.df)) # Should be TRUE

exposure.string <- paste(colnames(exposure.df), collapse = " + ")
outcome.string <- paste(colnames(outcome.df), collapse = " + ")

meta.df <- cbind(exposure.df, outcome.df)

## Mediation test
formula.str <- as.formula(paste("mediator.df ~", exposure.string, "+", outcome.string))
test.str <- as.formula(paste0("mediator.df ~"," (","exposure1 + exposure2", ") + ", "(","outcome1 + outcome2",")"))

res.ldm.med <- ldm(formula  = formula.str, data = meta.df, seed = 67817, n.cores = 1, test.mediation = TRUE) # parameter for requesting LDM-omni3
res.ldm.med$med.p.global.omni 
res.ldm.med$med.detected.otu.omni 
```

```{r, MODIMA, environment-prokaryotes-superclass}
library("vegan")
library("ade4")
library("energy")
library("permute")
library("matrixStats")
library("phyloseq")
library("tidyverse")
source("src/mediation/modima.R")

# change mediator input according to trophic group.
mediator.df <- as.data.frame(t(bac.mco.fld$otu_table))
exposure.df <- as.data.frame(bdg.mco.fld$sample_table[c(6:24)])

# Superclass options: "s__Benzenoids", "s__Lipids and lipid-like molecules", "s__Nucleosides, nucleotides, and analogues", "s__Organic acids and derivatives", "s__Organic nitrogen compounds", "s__Organic oxygen compounds", "s__Organoheterocyclic compounds", "s__Phenylpropanoids and polyketides"
tmp <- clone(bdg.mco.fld)
tmp$tax_table %<>% base::subset(Superclass == "s__Benzenoids")#; tmp$tidy_dataset()
tmp$cal_betadiv(unifrac = FALSE); class(tmp$beta_diversity) ## Create trans_beta object and measure parameter - dataset$beta_diversity

response.df <- as.data.frame(t(tmp$otu_table))

mediator.dist <- vegdist(mediator.df, method = "bray")
exposure.dist <- vegdist(exposure.df, method = "euclidean")
response.dist <- vegdist(response.df, method = "bray")

set.seed(12345)
modima.results <- modima(exposure = exposure.dist, 
                         mediator = mediator.dist, 
                         response = response.dist, 
                         nrep = 9999)

set.seed(12345)
ER_M_list <- pdcor.test(exposure.dist, response.dist, mediator.dist, R = 9999)
MR_E_list <- pdcor.test(mediator.dist, response.dist, exposure.dist, R = 9999)
beta_MR_E <- pdcor(mediator.dist, response.dist, exposure.dist) %>% as.numeric()
beta_ER <- bcdcor(exposure.dist, response.dist) %>% as.numeric()
ER_list <- dcor.test(exposure.dist, response.dist, R = 9999)
beta_EM <- bcdcor(exposure.dist, mediator.dist) %>% as.numeric()
EM_list <- dcor.test(exposure.dist, mediator.dist, R = 9999)

result.summary <- data.frame(Med_statistic = modima.results$statistic %>% as.numeric(),
                             Med_percentage = (beta_MR_E * beta_EM)/beta_ER * 100,
                             Med_P = modima.results$p.value,
                             beta_EM = beta_EM,
                             P_EM = EM_list$p.value,
                             beta_MR_E = beta_MR_E,
                             P_MR_E = MR_E_list$p.value,
                             beta_ER = beta_ER,
                             P_ER = ER_list$p.value,
                             name = "exposure_mediator_response_all")
modima.results
result.summary
#write_tsv(result.summary, "src/mediation/bac.modima.tsv")
```

```{r, MODIMA, trophic-prokaryotes-biodegradation}
library("vegan")
library("ade4")
library("energy")
library("permute")
library("matrixStats")
library("phyloseq")
library("tidyverse")
source("src/mediation/modima.R")

# change mediator input according to trophic group.
mediator.df <- as.data.frame(t(alg.mco.fld$otu_table))

exposure.df <- as.data.frame(bdg.mco.fld$sample_table[c(6:24)])
response.df <- as.data.frame(t(bac.mco.fld$otu_table))

mediator.dist <- vegdist(mediator.df, method = "bray")
exposure.dist <- vegdist(exposure.df, method = "euclidean")
response.dist <- vegdist(response.df, method = "bray")

set.seed(12345)
modima.results <- modima(exposure = exposure.dist, 
                         mediator = mediator.dist, 
                         response = response.dist, 
                         nrep = 9999)

set.seed(12345)
ER_M_list <- pdcor.test(exposure.dist, response.dist, mediator.dist, R = 9999)
MR_E_list <- pdcor.test(mediator.dist, response.dist, exposure.dist, R = 9999)
beta_MR_E <- pdcor(mediator.dist, response.dist, exposure.dist) %>% as.numeric()
beta_ER <- bcdcor(exposure.dist, response.dist) %>% as.numeric()
ER_list <- dcor.test(exposure.dist, response.dist, R = 9999)
beta_EM <- bcdcor(exposure.dist, mediator.dist) %>% as.numeric()
EM_list <- dcor.test(exposure.dist, mediator.dist, R = 9999)

result.summary <- data.frame(Med_statistic = modima.results$statistic %>% as.numeric(),
                             Med_percentage = (beta_MR_E * beta_EM)/beta_ER * 100,
                             Med_P = modima.results$p.value,
                             beta_EM = beta_EM,
                             P_EM = EM_list$p.value,
                             beta_MR_E = beta_MR_E,
                             P_MR_E = MR_E_list$p.value,
                             beta_ER = beta_ER,
                             P_ER = ER_list$p.value,
                             name = "exposure_mediator_response_all")
modima.results
result.summary
#write_tsv(result.summary, "src/mediation/bac.modima.tsv")
```

### Mediation analysis with SparseMCMM (NOT USED)

```{r}
library(vegan); library(SparseMCMM)

tmp <- clone(bac.mco.fld)
tmp$filter_taxa(rel_abund = 0.0001)
tmp

# Change mediator input according to trophic group.
exposure.df <- as.data.frame(bdg.mco.fld$sample_table[c(6:24)]) #Treatment <- SimulatedData$Treatment
mediator.df <- as.data.frame(t(tmp$otu_table)) #otu.com <- SimulatedData$otu.com
response.df <- as.data.frame(t(bdg.mco.fld$otu_table)) #outcome <- SimulatedData$outcome

# Calculate p-values based on 200 times of permutations in a single random data split
set.seed(1234)

#res <- SparseMCMM(Treatment, otu.com, outcome, n.split = 1, num.per=NULL
res <- SparseMCMM(exposure.df[, 1], mediator.df, response.df[, 1], n.split = 1, num.per = NULL)
```



### Network analaysis (NOT USED)

```{r}
dataset <- clone(bac.mco.fld)

library(WGCNA); library(igraph); library(rgexf)

# SPIEC-EASI (SParse InversE Covariance Estimation for Ecological Association Inference) approach of SpiecEasi R package (Kurtz et al. 2015) has two network construction approaches based on graph model, which relies on algorithms for sparse neighborhood and inverse covariance selection. See https://github.com/zdk123/SpiecEasi for the package installation. It is very slow for SpiecEasi_method = ‘glasso’ when there is a large number (such as hundreds to thousands) according to our test experience.

#t1 <- trans_network$new(dataset = dataset, cor_method = "spearman", use_WGCNA_pearson_spearman = TRUE, filter_thres = 0.0001)
t1 <- trans_network$new(dataset = dataset, cor_method = NULL, taxa_level = "OTU", filter_thres = 0.001)
# require SpiecEasi package installed https://github.com/zdk123/SpiecEasi
# also see SpiecEasi::spiec.easi for available model parameters
t1$cal_network(network_method = "SpiecEasi", SpiecEasi_method = "mb")

# invoke igraph cluster_fast_greedy function for this undirected network 
t1$cal_module(method = "cluster_fast_greedy")
# require rgexf package to be installed
t1$save_network(filepath = "src/network.gexf")

# calculate network attributes
t1$cal_network_attr()
t1$res_network_attr

# get node properties
t1$get_node_table(node_roles = TRUE)
# return t1$res_node_table

# get edge properties
t1$get_edge_table(); t1$res_edge_table 
t1$get_adjacency_matrix(); t1$res_adjacency_matrix

# add_label = TRUE can be used to directly add text label for points
t1$plot_taxa_roles(use_type = 1)

# plot node roles with phylum information
t1$plot_taxa_roles(use_type = 2)

t1$cal_eigen(); t1$res_eigen

# create trans_env object
t2 <- trans_env$new(dataset = dataset, add_data = (t(bdg.mco.fld$otu_table)))
# calculate correlations
t2$cal_cor(add_abund_table = t1$res_eigen)
# plot the correlation heatmap
t2$plot_cor() + coord_flip()# create trans_env object
```

### Sesonal taxa and biodegradation profile

```{r, Correlation heatmap of seasonal taxa and each compound}
library("microViz")

# Clone the dataset and redefine Region levels
dataset <- clone(bac.mco.ssn)
dataset$tax_table$Genus <- gsub("g__", "", dataset$tax_table$Genus)
dataset$sample_table <- cbind(dataset$sample_table, t(bdg.mco.fld$otu_table))

psq <- file2meco::meco2phyloseq(dataset)

# compute correlations, with p values, and store in a dataframe
correlations_df <- psq %>% 
  tax_model(trans = "log", 
            rank = "Genus",
            variables = list("`3-((2-Ethylhexyl)oxy)propane-1,2-diol`","`4-Chloro-3,5-dimethylphenol`","`4-(4-Nitrobenzyl)-pyridine`","`5-Methylbenzotriazole`",
                             "Abacavir","Alachlor","Amisulpride","Anastrozole","Acesulfame","Atenolol","Acetamiprid","Atrazine",
                             "Bezafibrate","Bisoprolol","Bromoxynil","Benzotriazole",
                             "Caffeine","Candesartan","Carbendazim","Carbamazepine","Cilastatin","`C12 Isethionate`","Citalopram","`Clofibric acid`","Climbazole","Chlorthalidone","Chlortoluron","Chlorothiazide","Cyclamate",
                             "Dicamba","Decylamine","Dibenzepin","Diclofenac","Dimethenamid","Diuron","Dodecylamine",
                             "Ethylparaben",
                             "Fluconazole","Flufenacet","Fenhexamid","Flecainide","Furosemide","Fluoxetine ",
                             "Gabapentin","Gemfibrozil",
                             "`Hydroxy bupropion`","Hydrochlorothiazide",
                             "Imidacloprid","Iprovalicarb","Irbesartan","Isoproturon",
                             "Ketoprofen",
                             "Lamotrigine","Levamisole","Lidocaine","Linezolid","Losartan","Levetiracetam",
                             "`Methyl 4-hydroxybenzoate`","MCPA","Mecoprop","`Mefenamic acid`","Metformin","Metolachlor","Metoprolol","Metoxuron",
                             "Naproxen","Neostigmine","`N,N-bis(2-hydroxyethyl)dodecanamide`","Nicotinamide","Novobiocin",
                             "Oxazepam","Oxprenolol",
                             "`Propyl 4-hydroxybenzoate`","Paracetamol","Pargyline","Propachlor","Propranolol","`p-Toluenesulfonic acid`",
                             "Ranitidine","Rufinamide",
                             "Sulfamethoxypyridazine","Sulfamethazine","Sulfadimethoxine","Sulfamethoxazole","Sotalol","Sulfathiazole",
                             "`Trinexapac-ethyl`","Tamsulosin","Terbutryn","Trimethoprim","Tramadol","`Triethyl citrate`",
                             "Valsartan","Venlafaxine",
                             "Zolpidem"),
            type = microViz::cor_test, method = "spearman",
            return_psx = FALSE, verbose = FALSE) %>% tax_models2stats(rank = "Genus")

# get nice looking ordering of correlation estimates using hclust
taxa_hclust <- correlations_df %>%  
               dplyr::select(term, taxon, estimate) %>% 
               tidyr::pivot_wider(id_cols = taxon, names_from = term, values_from = estimate) %>% 
               tibble::column_to_rownames("taxon") %>% 
               as.matrix() %>% stats::dist(method = "euclidean") %>% 
               hclust(method = "ward.D2") 

taxa_order <- taxa_hclust$labels[taxa_hclust$order]

h1 <- correlations_df %>% mutate(p.FDR = p.adjust(p.value, method = "fdr")) %>% ggplot(aes(x = term, y = taxon)) +
      geom_raster(aes(fill = estimate)) +
      geom_point(data = function(x) filter(x, p.value < 0.05), shape = "asterisk") +
      geom_point(data = function(x) filter(x, p.FDR < 0.05), shape = "circle", size = 2) +
      scale_y_discrete(limits = taxa_order) +
      scale_fill_distiller(palette = "PuOr", limits = c(-1, 1)) + theme_bw() + 
      theme(axis.text.x = element_text(size = 9, angle = 45, hjust = 1),
            axis.text.y = element_text(size = 8), axis.ticks = element_blank(),
            legend.position = "left", legend.title = element_text(face = "bold", size = 8), 
            legend.text = element_text(size = 7), panel.border = element_blank(), panel.grid = element_blank()) + 
      labs(x = NULL, y = NULL, fill = "Spearman's\nRank\nCorrelation")

h1
 
library(ggdendro)

# Convert hclust to dendrogram and then to ggplot-compatible format
dendro_data <- ggdendro::dendro_data(as.dendrogram(taxa_hclust), type = "rectangle")

# Plot dendrogram
dendro_plot <- ggplot(segment(dendro_data)) + geom_segment(aes(x = y, y = x, xend = yend, yend = xend)) +
               scale_y_discrete(limits = taxa_order) + theme_bw() +
               theme(axis.text = element_blank(), axis.title = element_blank(), axis.ticks = element_blank(),
                     panel.border = element_blank(), panel.grid = element_blank())

# Combine with heatmap using patchwork
library(patchwork)
combined_plot <- h1 + dendro_plot + plot_layout(ncol = 2, widths = c(5, 1))
combined_plot
```

```{r, Correlation heatmap of seasonal taxa and each superclass}
library("microViz")

# Clone the dataset and redefine Region levels
dataset <- clone(bac.mco.ssn)
dataset$tax_table$Phylum <- gsub("p__", "", dataset$tax_table$Phylum)

tmp <- clone(bdg.mco.fld)
tmp <- file2meco::meco2phyloseq(tmp)
tmp <- tax_glom(tmp, taxrank = "Superclass")
tmp <- file2meco::phyloseq2meco(tmp)
dataset$sample_table <- cbind(dataset$sample_table, t(tmp$otu_table))

# Rename the column in sample_table
names(dataset$sample_table)[names(dataset$sample_table) == "3-((2-Ethylhexyl)oxy)propane-1,2-diol"] <- "Lipids and lipid-like molecules"
names(dataset$sample_table)[names(dataset$sample_table) == "Abacavir"] <- "Nucleosides, nucleotides, and analogues"
names(dataset$sample_table)[names(dataset$sample_table) == "Cilastatin"] <- "Organic acids and derivatives"
names(dataset$sample_table)[names(dataset$sample_table) == "Citalopram"] <- "Benzenoids"
names(dataset$sample_table)[names(dataset$sample_table) == "Climbazole"] <- "Organoheterocyclic compounds"
names(dataset$sample_table)[names(dataset$sample_table) == "Hydroxy bupropion"] <- "Organic oxygen compounds"
names(dataset$sample_table)[names(dataset$sample_table) == "Novobiocin"] <- "Phenylpropanoids and polyketides"
names(dataset$sample_table)[names(dataset$sample_table) == "Ranitidine"] <- "Organic nitrogen compounds"

psq <- file2meco::meco2phyloseq(dataset)

# compute correlations, with p values, and store in a dataframe
correlations_df <- psq %>% 
  tax_model(trans = "standardize", 
            rank = "Phylum",
            variables = list("Benzenoids","`Lipids and lipid-like molecules`","`Nucleosides, nucleotides, and analogues`","`Organic acids and derivatives`",
                             "`Organic nitrogen compounds`","`Organic oxygen compounds`","`Organoheterocyclic compounds`","`Phenylpropanoids and polyketides`"),
            type = microViz::cor_test, method = "spearman",
            return_psx = FALSE, verbose = FALSE) %>% tax_models2stats(rank = "Phylum")

# get nice looking ordering of correlation estimates using hclust
taxa_hclust <- correlations_df %>%  
               dplyr::select(term, taxon, estimate) %>% 
               tidyr::pivot_wider(id_cols = taxon, names_from = term, values_from = estimate) %>% 
               tibble::column_to_rownames("taxon") %>% 
               as.matrix() %>% stats::dist(method = "euclidean") %>% 
               hclust(method = "ward.D2") 

taxa_order <- taxa_hclust$labels[taxa_hclust$order]

h2 <- correlations_df %>% mutate(p.FDR = p.adjust(p.value, method = "fdr")) %>% ggplot(aes(x = term, y = taxon)) +
      geom_raster(aes(fill = estimate)) +
      geom_point(data = function(x) filter(x, p.value < 0.05), shape = "asterisk") +
      geom_point(data = function(x) filter(x, p.FDR < 0.05), shape = "circle", size = 2) +
      scale_y_discrete(limits = taxa_order) +
      scale_fill_distiller(palette = "PuOr", limits = c(-1, 1)) + theme_bw() + 
      theme(axis.text.x = element_text(size = 9, angle = 45, hjust = 1),
            axis.text.y = element_text(size = 8), axis.ticks = element_blank(),
            legend.position = "left", legend.title = element_text(face = "bold", size = 8), 
            legend.text = element_text(size = 7), panel.border = element_blank(), panel.grid = element_blank()) + 
      labs(x = NULL, y = NULL, fill = "Spearman's\nRank\nCorrelation") + force_panelsizes(cols = unit(6, "cm")) 

h2 
 
library(ggdendro)

# Convert hclust to dendrogram and then to ggplot-compatible format
#dendro_data <- ggdendro::dendro_data(as.dendrogram(taxa_hclust), type = "triangle")

# Plot dendrogram
#dendro_plot <- ggplot(segment(dendro_data)) + geom_segment(aes(x = y, y = x, xend = yend, yend = xend)) +
#               scale_y_discrete(limits = taxa_order) + theme_bw() +
#               theme(axis.text = element_blank(), axis.title = element_blank(), axis.ticks = element_blank(),
#                     panel.border = element_blank(), panel.grid = element_blank())
#
# Combine with heatmap using patchwork
#library(patchwork)
#combined_plot <- h1 + dendro_plot + plot_layout(ncol = 2, widths = c(5, 1))
#combined_plot
```

```{r, Correlation heatmap of function and each superclass}
library("microViz")

# Clone the dataset and redefine Region levels
dataset <- clone(keg.mco.xen)

tmp <- clone(bdg.mco.fld)
tmp <- file2meco::meco2phyloseq(tmp)
tmp <- tax_glom(tmp, taxrank = "Superclass")
tmp <- file2meco::phyloseq2meco(tmp)
dataset$sample_table <- cbind(dataset$sample_table, t(tmp$otu_table))

# Rename the column in sample_table
names(dataset$sample_table)[names(dataset$sample_table) == "3-((2-Ethylhexyl)oxy)propane-1,2-diol"] <- "Lipids and lipid-like molecules"
names(dataset$sample_table)[names(dataset$sample_table) == "Abacavir"] <- "Nucleosides, nucleotides, and analogues"
names(dataset$sample_table)[names(dataset$sample_table) == "Cilastatin"] <- "Organic acids and derivatives"
names(dataset$sample_table)[names(dataset$sample_table) == "Citalopram"] <- "Benzenoids"
names(dataset$sample_table)[names(dataset$sample_table) == "Climbazole"] <- "Organoheterocyclic compounds"
names(dataset$sample_table)[names(dataset$sample_table) == "Hydroxy bupropion"] <- "Organic oxygen compounds"
names(dataset$sample_table)[names(dataset$sample_table) == "Novobiocin"] <- "Phenylpropanoids and polyketides"
names(dataset$sample_table)[names(dataset$sample_table) == "Ranitidine"] <- "Organic nitrogen compounds"

psq <- file2meco::meco2phyloseq(dataset)

#'arg' should be one of “total”, “max”, “frequency”, “normalize”, “range”, “rank”, “rrank”, “standardize”, “pa”, “chi.square”, “hellinger”, “log”, “clr”, “rclr”, “alr”

# compute correlations, with p values, and store in a dataframe
correlations_df <- psq %>% 
  tax_model(trans = "standardize", 
            rank = "Level.3",
            variables = list("Benzenoids","`Lipids and lipid-like molecules`",
                             "`Nucleosides, nucleotides, and analogues`","`Organic acids and derivatives`",
                             "`Organic nitrogen compounds`","`Organic oxygen compounds`",
                             "`Organoheterocyclic compounds`","`Phenylpropanoids and polyketides`"),
            type = microViz::cor_test, method = "spearman", 
            return_psx = FALSE, verbose = FALSE) %>% tax_models2stats(rank = "Level.3")

# get nice looking ordering of correlation estimates using hclust
taxa_hclust <- correlations_df %>%  
               dplyr::select(term, taxon, estimate) %>% 
               tidyr::pivot_wider(id_cols = taxon, names_from = term, values_from = estimate) %>% 
               tibble::column_to_rownames("taxon") %>% 
               as.matrix() %>% stats::dist(method = "euclidean") %>% 
               hclust(method = "ward.D2") 

taxa_order <- taxa_hclust$labels[taxa_hclust$order]

h3 <- correlations_df %>% mutate(p.FDR = p.adjust(p.value, method = "fdr")) %>% ggplot(aes(x = term, y = taxon)) +
      geom_raster(aes(fill = estimate)) +
      geom_point(data = function(x) filter(x, p.value < 0.05), shape = "asterisk") +
      geom_point(data = function(x) filter(x, p.FDR < 0.05), shape = "circle", size = 2) +
      scale_y_discrete(limits = taxa_order) +
      scale_fill_distiller(palette = "PuOr", limits = c(-1, 1)) + theme_bw() + 
      theme(axis.text.x = element_text(size = 9, angle = 45, hjust = 1),
            axis.text.y = element_text(size = 8), axis.ticks = element_blank(),
            legend.position = "left", legend.title = element_text(face = "bold", size = 8), 
            legend.text = element_text(size = 7), panel.border = element_blank(), panel.grid = element_blank()) + 
      labs(x = NULL, y = NULL, fill = "Spearman's\nRank\nCorrelation") + force_panelsizes(cols = unit(6, "cm")) 

h3
 
library(ggdendro)

# Convert hclust to dendrogram and then to ggplot-compatible format
#dendro_data <- ggdendro::dendro_data(as.dendrogram(taxa_hclust), type = "triangle")

# Plot dendrogram
#dendro_plot <- ggplot(segment(dendro_data)) + geom_segment(aes(x = y, y = x, xend = yend, yend = xend)) +
#               scale_y_discrete(limits = taxa_order) + theme_bw() +
#               theme(axis.text = element_blank(), axis.title = element_blank(), axis.ticks = element_blank(),
#                     panel.border = element_blank(), panel.grid = element_blank())
#
# Combine with heatmap using patchwork
#library(patchwork)
#combined_plot <- h1 + dendro_plot + plot_layout(ncol = 2, widths = c(5, 1))
#combined_plot
```

## ########################################################## ##
##                                                            ##
##                        THE END!                            ##
##                                                            ##
## ########################################################## ##